% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/deprecated.R, R/plotMA-methods.R
\docType{methods}
\name{plotMeanAverage}
\alias{plotMeanAverage}
\alias{plotMA}
\alias{plotMA,DESeqResults-method}
\title{MA Plot}
\usage{
\S4method{plotMA}{DESeqResults}(object, alpha, lfcThreshold = 0L,
  genes = NULL, gene2symbol = NULL, ntop = 0L, direction = c("both",
  "up", "down"), pointColor = "gray50", sigPointColor = c(upregulated =
  "purple", downregulated = "orange"), return = c("ggplot", "data.frame"))
}
\arguments{
\item{...}{Additional arguments.}

\item{object}{Object.}

\item{alpha}{Adjusted P value ("alpha") cutoff.}

\item{lfcThreshold}{Log fold change ratio (base 2) cutoff threshold.}

\item{genes}{Character vector of genes to include. These must match the
rownames of the object. It is best practice to use the stable gene
identifiers from Ensembl (e.g. "ENSG00000000003") and not the gene symbols.}

\item{gene2symbol}{\code{data.frame} containing gene-to-symbol mappings. Must
contain the columns \code{geneID} and \code{geneName}.}

\item{ntop}{Number of top genes to label.}

\item{direction}{Plot "\code{both}", "\code{up}", or "\code{down}" directions.}

\item{pointColor}{Default point color for the plot.}

\item{sigPointColor}{\code{character} vector containing color names for labeling
upregulated and downregulated genes. Also supports a character string for
labeling DEGs with the same color, regardless of direction.}

\item{return}{Object class to return. Uses \code{\link[=match.arg]{match.arg()}} internally and picks
the first item in the vector by default.}
}
\value{
\code{ggplot}.
}
\description{
An MA plot is an application of a Blandâ€“Altman plot for visual representation
of genomic data. The plot visualizes the differences between measurements
taken in two samples, by transforming the data onto M (log ratio) and A
(mean average) scales, then plotting these values.
}
\examples{
gene2symbol <- gene2symbol(bcb_small)

# DESeqResults ====
# Color DEGs in each direction separately
plotMA(
    object = res_small,
    sigPointColor = c(
        upregulated = "purple",
        downregulated = "orange"
    )
)

# Label DEGs with a single color
plotMA(res_small, sigPointColor = "purple")

# Directional support
plotMA(
    object = res_small,
    direction = "up",
    ntop = 5L,
    gene2symbol = gene2symbol
)
plotMA(
    object = res_small,
    direction = "down",
    ntop = 5L,
    gene2symbol = gene2symbol
)

# Label genes manually
plotMA(
    object = res_small,
    genes = head(rownames(res_small)),
    gene2symbol = gene2symbol
)
}
\seealso{
\code{\link[DESeq2:plotMA]{DESeq2::plotMA()}}.

Other Differential Expression Functions: \code{\link{alphaSummary}},
  \code{\link{contrastName}}, \code{\link{plotDEGHeatmap}},
  \code{\link{plotDEGPCA}}, \code{\link{plotVolcano}},
  \code{\link{resultsTables}}
}
\author{
Rory Kirchner, Michael Steinbaugh
}
