# Methods

RNA-seq counts were generated by [bcbio][] and [bcbioRNASeq][] using [Salmon][] [@Patro2017-pi]. Counts were imported into [R][] using [tximport][] [@Soneson2016-tu] and [DESeq2] [@Love2014-sq]. Gene annotations were obtained from [Ensembl][]. Plots were generated by [ggplot2][] [@Wickham2009-si]. Heatmaps were generated by [pheatmap][] [@Kolde2015-dy].

# R session information {.tabset}

```{r session_info}
markdownHeader("YAML params", level = 2)
print(params)

markdownHeader("`_setup.R`", level = 2)
readr::read_lines("_setup.R") %>% writeLines()

markdownHeader("`sessioninfo::session_info()`", level = 2)
sessioninfo::session_info()

markdownHeader("`utils::sessionInfo()`", level = 2)
utils::sessionInfo()
```

[7-Zip]: http://www.7-zip.org
[bcbio]: https://github.com/chapmanb/bcbio-nextgen
[bcbioRNASeq]: http://bioinformatics.sph.harvard.edu/bcbioRNASeq
[Bioconductor]: https://bioconductor.org
[clusterProfiler]: https://doi.org/doi:10.18129/B9.bioc.clusterProfiler
[DESeq2]: https://bioconductor.org/packages/release/bioc/html/DESeq2.html
[Ensembl]: https://www.ensembl.org
[Entrez]: https://www.ncbi.nlm.nih.gov/gquery
[Excel]: https://products.office.com/en-us/excel
[ggplot2]: http://ggplot2.org
[GO]: http://www.geneontology.org
[GSEA]: http://software.broadinstitute.org/gsea
[KEGG]: http://www.genome.jp/kegg
[LibreOffice]: https://www.libreoffice.org
[macOS]: https://www.apple.com/macos
[pathview]: https://doi.org/doi:10.18129/B9.bioc.pathview
[pheatmap]: https://cran.r-project.org/web/packages/pheatmap/index.html
[R]: https://www.r-project.org
[RStudio]: https://www.rstudio.com
[salmon]: https://combine-lab.github.io/salmon
[tximport]: https://bioconductor.org/packages/release/bioc/html/tximport.html
