---
title: "bcbioRNASeq"
date: "`r BiocStyle::doc_date()`"
output: BiocStyle::html_document
bibliography: ../inst/rmarkdown/shared/bibliography.bib
vignette: >
  %\VignetteIndexEntry{bcbioRNASeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
abstract: >
  RNA-seq analysis involves multiple steps from processing raw sequencing data
  to identifying, organizing, annotating, and reporting differentially expressed
  genes. bcbio is an open source, community-maintained framework providing
  automated and scalable RNA-seq methods for identifying gene abundance counts.
  We have developed bcbioRNASeq, a Bioconductor package that provides
  ready-to-render templates and wrapper functions to post-process bcbio output
  data. bcbioRNASeq automates the generation of high-level RNA-seq reports,
  including identification of differentially expressed genes, functional
  enrichment analysis and quality control analysis.
---

```{r setup, include=FALSE, message=FALSE}
ggplot2::theme_set(acidplots::acid_theme_light())
```

# Introduction

```{r}
library("bcbioRNASeq")
```

For a high-level overview of our [bcbio][] RNA-seq analysis pipeline, including detailed explanation of the `bcbioRNASeq` S4 class definition, first consult our [workflow paper][] published in *F1000 Research* [@Steinbaugh2018-rc]. This vignette is focused on more advanced usage and edge cases that a user may encounter when attempting to load a [bcbio][] dataset and perform downstream quality control analysis.

**Note:** if you use [bcbioRNASeq][] in published research, please include this citation:

```{r}
citation("bcbioRNASeq")
```

# Loading bcbio data

The `bcbioRNASeq()` constructor function is the main interface connecting [bcbio][] output data to interactive use in [R][]. It is highly customizable and supports a number of options for advanced use cases. Consult the documentation available at `help(topic = "bcbioRNASeq", package = "bcbioRNASeq")` for additional details.

```{r formals}
formals("bcbioRNASeq") %>% str()
```

## Upload directory

We have designed the constructor to work as simply as possible by default. The only required argument is `uploadDir`, the path to the [bcbio][] final upload directory specified with `upload:` in the YAML configuration. Refer to the [bcbio configuration](https://bcbio-nextgen.readthedocs.io/en/latest/contents/configuration.html) documentation for detailed information on how to set up a [bcbio][] run, which is outside the scope of this vignette.

For example, let's load up the example [bcbio][] dataset stored internally in the package.

```{r}
upload_dir <- system.file("extdata/bcbio", package = "bcbioRNASeq")
stopifnot(file.exists(upload_dir))
print(upload_dir)
```

[bcbio][] outputs RNA-seq data in a standardized directory structure, which is described in detail in our [workflow paper][].

```{r}
dir(path = upload_dir, full.names = FALSE, recursive = TRUE)
```

## Counts level

By default, [bcbioRNASeq][] imports counts at gene level, which are required for standard differential expression analysis (`level = "genes"`). For pseudo-aligned counts (e.g. [Salmon][], [Kallisto][], [Sailfish][]) [@Bray2016-me; @Patro2014-qz; @Patro2017-pi], [tximport][] [@Soneson2016-tu] is used internally to aggregate transcript-level counts to gene-level counts, and generates length-scaled transcripts per million (TPM) values. For aligned counts processed with [featureCounts][] [@Liao2014-js] (e.g. [STAR][], [HISAT2][]) [@Dobin2013-hr; @Dobin2016-qj; @Kim2015-rr], these values are already returned at gene level, and therefore not handled by [tximport][]. Once the gene-level counts are imported during the `bcbioRNASeq()` call, the [DESeq2][] package [@Love2014-sq] is then used to generate an internal `DESeqDataSet` from which we derive normalized and variance-stabilized counts.

```{r, message=FALSE}
bcb <- bcbioRNASeq(uploadDir = upload_dir, level = "genes")
print(bcb)
```

Alternatively, if you want to perform transcript-aware analysis, such as differential exon usage or splicing analysis, transcript-level counts can be obtained using `level = "transcripts"`. Note that when counts are loaded at transcript level, TPMs are generated with [tximport][] internally, but no additional normalizations or transformations normally calculated for gene-level counts with [DESeq2][] are generated.

```{r, message=FALSE}
bcb <- bcbioRNASeq(uploadDir = upload_dir, level = "transcripts")
print(bcb)
```

## Expression callers

Since [bcbio][] is flexible and supports a number of expression callers, we have provided advanced options in the `bcbioRNASeq()` constructor to support a variety of workflows using the `caller` argument.

[Salmon][], [Kallisto][], and [Sailfish][] counts are supported at either gene or transcript level. Internally, these are loaded using [tximport][].

```{r, message=FALSE}
salmon_genes <- bcbioRNASeq(uploadDir = upload_dir, caller = "salmon")
print(salmon_genes)
assayNames(salmon_genes)
```

```{r, message=FALSE}
salmon_tx <- bcbioRNASeq(uploadDir = upload_dir, level = "transcripts")
print(salmon_tx)
assayNames(salmon_tx)
```

[STAR][] and [HISAT2][] aligned counts processed with [featureCounts][] are also supported, but only at gene level.

```{r, message=FALSE}
star <- bcbioRNASeq(uploadDir = upload_dir, caller = "star")
print(star)
assayNames(star)
```

## Sample selection and metadata

If you'd like to load up only a subset of samples, this can be done easily using the `samples` argument. Note that the `character` vector declared here must match the `description` column specified in the sample metadata. For example, let's create a `bcbioRNASeq` object containing only the `group1_1` and `group1_2` samples.

```{r, message=FALSE}
bcb <- bcbioRNASeq(
    uploadDir = upload_dir,
    samples = c("group1_1", "group1_2")
)
print(bcb)
sampleNames(bcb)
```

Conversely, if you're working with a large dataset and you simply want to drop a few samples, this can be accomplished with the `censorSamples` argument. Note that the `samples` argument takes priority if both are declared. For example, let's drop the `group2_2` sample from our minimal dataset.

```{r, message=FALSE}
bcb <- bcbioRNASeq(
    uploadDir = upload_dir,
    censorSamples = "group2_2"
)
print(bcb)
sampleNames(bcb)
```

If you're working with a bcbio run that has incorrect or outdated metadata, the simplest way to fix this issue is to pass in new metadata from an external spreadsheet (CSV or Excel) using the `sampleMetadataFile` argument. Note that this can also be used to subset the bcbio dataset, similar to the `samples` argument (see above), based on the rows that are included in the spreadsheet.

```{r}
sample_metadata_file <- file.path(upload_dir, "sample_metadata.csv")
stopifnot(file.exists(sample_metadata_file))
readSampleData(sample_metadata_file)
```

```{r, message=FALSE}
bcb <- bcbioRNASeq(
    uploadDir = upload_dir,
    sampleMetadataFile = sample_metadata_file
)
print(bcb)
sampleData(bcb, clean = TRUE)
```

## Genome annotations

When analyzing a dataset against a well-annotated genome, we recommend importing the corresponding metadata using [AnnotationHub][] and [ensembldb][]. This functionality is natively supported in the `bcbioRNASeq()` constructor with using the `organism`, `ensemblRelease`, and `genomeBuild` arguments. This will return a `GRanges` object using the [GenomicRanges][] package [@Lawrence2013-vz], which contains coordinates and rich metadata for each gene or transcript. These annotations are accessible with the `rowRanges()` and `rowData()` functions defined in the [SummarizedExperiment][] package [@Huber2015-nw].

For example, with our internal [bcbio][] dataset, we're analyzing counts generated against the [Ensembl][] *Mus musculus* GRCm38 genome build (release 87). These parameters can be defined in the object load call to ensure that the annotations match up exactly with the genome used.

```{r, message=FALSE}
bcb <- bcbioRNASeq(
    uploadDir = upload_dir,
    level = "genes",
    organism = "Mus musculus",
    genomeBuild = "GRCm38",
    ensemblRelease = 87
)
print(bcb)
metadata(bcb)$rowRangesMetadata
rowRanges(bcb) %>% as.data.frame() %>% glimpse()
```

Alternatively, transcript-level annotations can also be obtained automatically using this method.

```{r, message=FALSE}
bcb <- bcbioRNASeq(
    uploadDir = upload_dir,
    level = "transcripts",
    organism = "Mus musculus",
    genomeBuild = "GRCm38",
    ensemblRelease = 87
)
print(bcb)
rowRanges(bcb) %>% as.data.frame() %>% glimpse()
```

When working with a dataset generated against a poorly-annotated or non-standard genome, we provide a fallback method for loading gene annotations from a [general feature format (GFF)][GFF] file with the `gffFile` argument. If possible, we recommend providing a general transfer format (GTF) file, which is identical to GFF version 2. GFFv3 is more complicated and non-standard, but [Ensembl][] GFFv3 files are also supported.

If your dataset contains transgenes (e.g. EGFP, TDTOMATO) or spike-ins (e.g. ERCCs), these features can be defined with the `transgeneNames` and `spikeNames` arguments, which will automatically populate the `rowRanges()` slot with placeholder metadata.

We recommend loading up data per genome in its own `bcbioRNASeq` object when possible, so that rich metadata can be imported easily. In the edge case where you need to look at multiple genomes simultaneously, set `organism = NULL`, and bcbioRNASeq will skip the gene annotation acquisition step.

```{r, message=FALSE}
bcb <- bcbioRNASeq(uploadDir = upload_dir, organism = NULL)
print(bcb)
```

## Variance stabilization

During the `bcbioRNASeq()` constructor call, variance stabilizaton of gene-level counts can be calculated automatically, and is recommended. This is performed internally by the [DESeq2][] package. These transformations will be slotted into `assays()` as `vst` and/or `rlog` matrices. The `vst` matrix is calculated by `DESeq2::varianceStabilizingTransformation()` and `rlog` by `DESeq2::rlog()`. For large datasets, `rlog` calculation can take a long time, so we are currently recommending calculation of only `vst` by default.

```{r, message=FALSE}
bcb <- bcbioRNASeq(
    uploadDir = upload_dir,
    vst = TRUE,
    rlog = TRUE
)
assayNames(bcb)
counts(bcb, normalized = "vst") %>% summary()
counts(bcb, normalized = "rlog") %>% summary()
rm(bcb)
```

# Use case dataset

To demonstrate the functionality and configuration of the package, we have taken an experiment from the [Gene Expression Omnibus (GEO)][GEO] public repository of expression data to use as an example use case. The RNA-seq data is from a study of acute kidney injury in a mouse model ([GSE65267][]) [@Craciun2016-ha]. The study aims to identify differentially expressed genes in progressive kidney fibrosis and contains samples from mouse kidneys at several time points (n = 3, per time point) after folic acid treatment. From this dataset, we are using a subset of the samples for our use case: before folic acid treatment, and 1, 3, 7 days after treatment.

For the vignette, we are loading a pre-computed version of the example `bcbioRNASeq` object used in our [workflow paper][].

```{r, message=FALSE}
loadRemoteData("https://github.com/hbc/bcbioRNASeq/raw/f1000v2/data/bcb.rda")
print(bcb)
```

# Sample metadata

For reference, let's take a look at the sample metadata. By comparison, using `colData()` will return all sample-level metadata, including our quality control metrics generated by bcbio. We recommend instead using `sampleData()` with the `clean = TRUE` argument in reports, which only returns `factor` columns of interest.

```{r}
sampleData(bcb, clean = TRUE)
```

# Interesting groups

Groups of interest to be used for coloring and sample grouping in the quality control plots can be defined in the `bcbioRNASeq` object using the `interestingGroups` argument in the `bcbioRNASeq()` constructor call. Assignment method support with the `interestingGroups()` function is also provided, which can modify the groups of interest after the object has been created.

The `interestingGroups` definition defaults to `sampleName`, which is automatically generated from the [bcbio][] `description` metadata for demultiplexed bulk RNA-seq samples. In this case, the samples will be grouped and colored uniquely.

Interesting groups must be defined using a `character` vector and refer to column names defined in the `colData()` slot of the object. Note that the [bcbioRNASeq][] packages uses lower camel case formatting for column names (e.g. "sampleName"), so the interesting groups should be defined using camel case, not snake (e.g. "sample_name") or dotted case (e.g. "sample.name").

This approach was inspired by the [DESeq2][] package, which uses the argument `intgroup` in some functions, such as `plotPCA()` for labeling groups of interest. We took this idea and ran with it for a number of our quality control functions, which are described in detail below.

```{r}
interestingGroups(bcb) <- c("treatment", "day")
interestingGroups(bcb)
```

# Quality control

Our [workflow paper][] describes our quality control process in detail. In addition, our **Quality Control** [R Markdown][] template contains notes detailing each metric. Here were are going into more technical detail regarding how to customize the appearance of the plots. Note that all quality control plotting functions inherit the `interestingGroups` defined inside the `bcbioRNASeq` object. You can also change this dynamially for each function call using the `interestingGroups` argument.

## Read counts

```{r}
plotTotalReads(bcb)
plotMappedReads(bcb)
```

## Mapping rates

Note that the overall mapping rate is relatively low per sample, while the exonic mapping rate is acceptable. High quality samples should have low intronic mapping rate, and high values are indicative of sample degradation and/or contamination.

```{r}
plotMappingRate(bcb)
plotExonicMappingRate(bcb)
plotIntronicMappingRate(bcb)
```

## rRNA mapping rate

Note that the samples here have a high rRNA mapping rate. This can be indicative of the polyA enrichment or ribo depletion protocol not having removed all ribosomal RNA (rRNA) transcripts. This will reduce the number of biologically meaningful reads in the experiment and is best avoided.

```{r}
plotRRNAMappingRate(bcb)
```

## 5'->3' bias

RNA-seq data can have specific biases at either the 5’ or 3’ end of sequenced fragments. It is common to see a small amount of bias, especially if polyA enrichment was performed, or if there is any sample degradation. If a large amount of bias is observed here, be sure to analyze the samples with a Bioanalyzer and check the RIN scores.

```{r}
plot5Prime3PrimeBias(bcb)
```

## Gene distributions

```{r}
plotGenesDetected(bcb)
```

```{r}
plotGeneSaturation(bcb, label = FALSE)
plotGeneSaturation(bcb, label = TRUE)
```

```{r}
plotCountsPerGene(bcb)
plotCountDensity(
    object = bcb,
    interestingGroups = "sampleName",
    style = "line"
)
```

## Dispersion

The following plot shows the dispersion by mean of normalized counts. We expect the dispersion to decrease as the mean of normalized counts increases.

```{r}
plotDispEsts(bcb)
```

## Variance stabilization

These plots show the standard deviation of normalized counts using `log2()`, `varianceStabilizingTransform()`, `rlog()`, and `tmm()` by `rank(mean)`. Note that all counts shown are log2 scale.

```{r}
plotMeanSD(bcb)
```

## Sample similarity

We can visualize sample similarity with principal component analysis (PCA) and hierarchical clustering of sample correlations. These functions support multiple normalization methods with the `normalized` argument. We recommend using `normalized = "vst"` by default.

Let's plot the PCA using our `vst` counts, the current recommended default.

```{r}
plotPCA(bcb, normalized = "vst", label = FALSE)
plotPCA(bcb, normalized = "vst", label = TRUE)
```

Let's visualize the `rlog` counts for comparison. The performance is similar overall, with some slight differences, but the grouping more or less remains the same.

```{r}
plotPCA(
    object = bcb,
    normalized = "rlog",
    label = FALSE
)
plotPCA(
    object = bcb,
    normalized = "rlog",
    label = TRUE
)
```

Generally, the `varianceStabilizationTransformation()` and `rlog()` functions provide similar performance, as seen here.

```{r}
plotCorrelationHeatmap(
    bcb,
    method = "pearson",
    normalized = "vst",
    color = viridis::viridis
)
plotCorrelationHeatmap(
    bcb,
    method = "spearman",
    normalized = "vst",
    color = viridis::plasma
)
```

```{r}
plotCorrelationHeatmap(
    object = bcb,
    method = "pearson",
    normalized = "rlog",
    color = viridis::viridis
)
plotCorrelationHeatmap(
    object = bcb,
    method = "spearman",
    normalized = "rlog",
    color = viridis::plasma
)
```

# Saving counts

The `counts()` function returns the abundance estimates generated by Salmon. Read counts for each sample in the dataset are aggregated into a matrix, in which columns correspond to samples and rows represent genes. Multiple normalized counts matrices are saved in the bcbioRNASeq object, and are accessible with `normalized` argument:

- `FALSE`: Raw counts (*default*).
- `TRUE`: [DESeq2][] normalized counts.
- `"tpm"`: Transcripts per million, calculated by [tximport][].
- `"tmm"`: Trimmed mean of M-values normalization method.
- `"vst"`: Variance stabilization transformation.
- `"rlog"`: Regularized log transformation.
- `"rle"`: Relative log expression transformation.

We recommend saving the raw counts, normalized counts, TPMs, and variance-stabilized counts to disk both in binary R Data and CSV formats.

```{r, message=FALSE}
raw <- counts(bcb, normalized = FALSE)
normalized <- counts(bcb, normalized = TRUE)
tpm <- counts(bcb, normalized = "tpm")
vst <- counts(bcb, normalized = "vst")
saveData(raw, normalized, tpm, vst, dir = ".")
writeCounts(raw, normalized, tpm, vst, dir = ".")
```

# Differential expression

[bcbioRNASeq][] integrates with [DESeq2][], which we recommend for differential expression. To prepare our dataset for analysis, we need to coerce the `bcbioRNASeq` object to a `DESeqDataSet`. Use the `as()` function or refer to the [DESeq2][] documentation for the `DESeqDataSet()` constructor.

```{r}
dds <- as(bcb, "DESeqDataSet")
print(dds)
```

Since both `bcbioRNASeq` and `DESeqDataSet` classes extend `RangedSummarizedExperiment`, internally we coerce the original `bcbioRNASeq` object to a `RangedSummarizedExperiment` and then use the `DESeqDataSet()` constructor, which requires a `SummarizedExperiment` with integer counts.

Here is the source code responsible for handing off the data to [DESeq2][].

```{r}
getMethod(
    f = "coerce",
    signature(
        from = "bcbioRNASeq",
        to = "DESeqDataSet"
    )
)
```

Now you can perform differential expression analysis with [DESeq2][]. The authors of that package have provided a number of detailed, well documented references online:

- [Analyzing RNA-seq data with DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).
- [RNA-Seq workflow: gene-level exploratory analysis and differential expression](http://dx.doi.org/10.12688/f1000research.7035.2).
- [Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences](https://f1000research.com/articles/4-1521/v2).

We also provide a parameterized [R Markdown][] template for standard [DESeq2][] differential expression, supporting analysis and visualization of multiple contrasts inside a single report. In our [workflow paper][], we also describe an example LRT analysis, and advanced edge-case templates are available inside our [hbcABC][] package, which is a current work in progress.

# Functional analysis

We provide an [R Markdown][] template for [gene set enrichment analysis (GSEA)][GSEA] and [Kyoto Encyclopedia of Genes and Genomes (KEGG)][KEGG] pathway enrichment analysis [@Kanehisa2000-hk; @Subramanian2005-zk], which leverages the functionality of the [clusterProfiler][] package [@Yu2012-ae]. This workflow is described in detail in our [workflow paper][], and is outside the scope of this advanced use vignette. For more information on functional analysis, consult the [clusterProfiler][] vignette, which is extremely detailed and thorough.

# R session information

```{r}
sessionInfo()
```

# References

The papers and software cited in our workflows are available as a [shared library](https://paperpile.com/shared/e1q8fn) on [Paperpile][].

[AnnotationHub]: https://doi.org/doi:10.18129/B9.bioc.AnnotationHub
[bcbio]: https://bcbio-nextgen.readthedocs.io
[bcbioRNASeq]: http://bioinformatics.sph.harvard.edu/bcbioRNASeq
[clusterProfiler]: https://doi.org/doi:10.18129/B9.bioc.clusterProfiler
[DESeq2]: https://doi.org/doi:10.18129/B9.bioc.DESeq2
[Ensembl]: https://www.ensembl.org
[ensembldb]: https://doi.org/doi:10.18129/B9.bioc.ensembldb
[featureCounts]: http://bioinf.wehi.edu.au/featureCounts
[GenomicRanges]: https://doi.org/doi:10.18129/B9.bioc.GenomicRanges
[GFF]: https://useast.ensembl.org/info/website/upload/gff.html
[GEO]: https://www.ncbi.nlm.nih.gov/geo
[GSE65267]: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE65267
[GSEA]: http://software.broadinstitute.org/gsea
[hbcABC]: https://github.com/hbc/hbcABC
[HISAT2]: https://ccb.jhu.edu/software/hisat2
[KEGG]: https://www.genome.jp/kegg
[Kallisto]: https://pachterlab.github.io/kallisto
[Paperpile]: https://paperpile.com
[R]: https://www.r-project.org
[R Markdown]: https://rmarkdown.rstudio.com
[Salmon]: https://combine-lab.github.io/salmon
[Sailfish]: http://www.cs.cmu.edu/~ckingsf/software/sailfish
[STAR]: https://github.com/alexdobin/STAR
[SummarizedExperiment]: https://doi.org/doi:10.18129/B9.bioc.SummarizedExperiment
[tximport]: https://doi.org/doi:10.18129/B9.bioc.tximport
[Workflow paper]: http://dx.doi.org/10.12688/f1000research.12093.2
