<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Changelog" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="bcbiornaseq-0-2-6-2018-07-09" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-2-6-2018-07-09" class="anchor"></a>bcbioRNASeq 0.2.6 (2018-07-09)</h1>
<div id="minor-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes" class="anchor"></a>Minor changes</h2>
<ul>
<li>Improved internal handling during <code>tximport()</code> call to handle transcript version mismatch with tx2gene data.frame. This can result if the bcbio pipeline is using an old genome build.</li>
<li>Updated internal code to use <code>aes()</code> instead of <code>aes_string()</code>, which uses tidyeval and quasiquotation.</li>
<li>
<code>plotGeneSaturation</code> now supports <code>label</code> argument, similar to <code><a href="../reference/plotPCA.html">plotPCA()</a></code>.</li>
<li>
<code><a href="../reference/plotGene.html">plotGene()</a></code>: reduced the number of <code>return</code> options to simply “facet” and “wide”. Previously, this also supported “grid”, “list”, and “markdown”, but these were removed because they are not frequently used.</li>
<li>Soft deprecated <code><a href="../reference/plotMA.html">plotMeanAverage()</a></code> in favor of <code><a href="../reference/plotMA.html">plotMA()</a></code>.</li>
<li>All function reexports have been consolidated in bcbioBase package.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-5-2018-06-21" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-2-5-2018-06-21" class="anchor"></a>bcbioRNASeq 0.2.5 (2018-06-21)</h1>
<div id="minor-changes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-1" class="anchor"></a>Minor changes</h2>
<ul>
<li>Enable support bcbio integration, by modifying <code><a href="../reference/bcbioRNASeq.html">bcbioRNASeq()</a></code> constructor to work with minimal bcbio test data.</li>
<li>Switched from internal usage of <code>aes_()</code> in favor of consistent usage of <code>aes_string()</code>. This will make the transition to [ggplot2][] v2.3.0 easier in a future update.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-4-2018-05-24" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-2-4-2018-05-24" class="anchor"></a>bcbioRNASeq 0.2.4 (2018-05-24)</h1>
<div id="major-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes" class="anchor"></a>Major changes</h2>
<ul>
<li>
<code><a href="../reference/aggregateReplicates.html">aggregateReplicates()</a></code> support has been added back. This function returns a <code>RangedSummarizedExperiment</code> instead of a <code>bcbioRNASeq</code> object, containing only an aggregate raw counts matrix in the <code>counts</code> slot of <code>assays()</code>.</li>
<li>The functional analysis [R Markdown][] template has been reworked to use <code>dds_file</code> and <code>organism</code> as new parameter arguments. We’ve reduced the number of parameters required here to run [clusterProfiler][].</li>
<li>Made <code>alphaSummary()</code> defunct for <code>bcbioRNASeq</code> object, in favor of <code>DESeqDataSet</code> only. This function is only useful when a proper design formula has been defined.</li>
</ul>
</div>
<div id="minor-changes-2" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-2" class="anchor"></a>Minor changes</h2>
<ul>
<li>
<code><a href="../reference/metrics.html">metrics()</a></code> now contains an informative error for datasets that were analyzed using the <code>fast-rnaseq</code> <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> pipeline.</li>
<li>
<code>DESeqDataSet</code> coercion from <code>bcbioRNASeq</code> object doesn’t attempt to run <code>DESeq()</code> command any more, which was unnecessary and improves speed.</li>
<li>
<code>bcbioSingleCell()</code> constructor now supports <code>censorSamples</code> parameter. This is useful for removing known poor quality samples upon loading.</li>
<li>[ggplot2][] color and fill palettes are now set <code>NULL</code> in the quality control functions. This behavior doesn’t change the appearance of the plot colors, which will still default to <code><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_hue">ggplot2::scale_colour_hue()</a></code> or <code><a href="http://www.rdocumentation.org/packages/ggplot2/topics/scale_hue">ggplot2::scale_fill_hue()</a></code>. The upcoming [ggplot2][] v2.3.0 update supports global options for color and fill palettes, so these parameters may be deprecated in a future release.</li>
<li>Reworked the internal code for <code>topTables()</code>.</li>
</ul>
</div>
<div id="infrastructure-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#infrastructure-changes" class="anchor"></a>Infrastructure changes</h2>
<ul>
<li>Added [macOS][] testing to [Travis CI][] build checks.</li>
<li>Fixed [clusterProfiler][] compilation error on [Travis CI][] by installing <code>libudunits2-dev</code> (Linux).</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-3-2018-05-10" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-2-3-2018-05-10" class="anchor"></a>bcbioRNASeq 0.2.3 (2018-05-10)</h1>
<div id="major-changes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-1" class="anchor"></a>Major changes</h2>
<ul>
<li>Now recommending variance stabilizing transformation (<code>vst</code>) over <code>rlog</code> counts by default in plots, where applicable.</li>
</ul>
</div>
<div id="minor-changes-3" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-3" class="anchor"></a>Minor changes</h2>
<ul>
<li>Tweaked Rory’s biotype plots in QC report to match formatting conventions in the package. These plots are now colored.</li>
<li>Added <code>plotDEGPCA()</code> to default differential expression R Markdown template.</li>
<li>
<code>colData()</code> factors are correctly releveled upon object subset with <code>[</code>. This helps avoid unwanted downstream errors when creating a <code>DESeqDataSet</code> and running differential expression with DESeq2.</li>
<li>Recommending <code>facet</code> return method by default for <code><a href="../reference/plotGene.html">plotGene()</a></code>. Updated the working example to reflect this.</li>
<li>
<code><a href="../reference/metrics.html">metrics()</a></code> now returns <code>interestingGroups</code> column.</li>
<li>
<code>sample</code> label has been removed from axis title for QC plot functions.</li>
<li>Now using shared ggplot2 convneience functions from bcbioBase 0.2.10: <code>bcbio_geom_abline()</code>, <code>bcbio_geom_label()</code>, and <code>bcbio_geom_label_repel()</code>. These are also used by bcbioSingleCell for improved graphical consistency.</li>
<li>Removed unused internal legacy ggplot2 code.</li>
<li>Increased DEGreport, DESeq2, and tximport dependency requirements.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-2-2018-04-26" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-2-2-2018-04-26" class="anchor"></a>bcbioRNASeq 0.2.2 (2018-04-26)</h1>
<div id="minor-changes-4" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-4" class="anchor"></a>Minor changes</h2>
<ul>
<li>Split out assertive imports so we can pin on [bioconda][].</li>
<li>Improved package documentation.</li>
<li>Improved label consistency in <code><a href="../reference/plotPCA.html">plotPCA()</a></code> functions to match <code><a href="../reference/plotMA.html">plotMeanAverage()</a></code> and <code>plotVolcano()</code>.</li>
<li>Improved automatic title labeling in <code>plotDEGPCA()</code>, matching the other DEG functions. Also added directionality to <code>plotDEGPCA()</code>.</li>
<li>Added DESeqDataSet method support to <code><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap()</a></code>, using the normalized counts.</li>
<li>
<code>reusltsTables()</code> now writes local files to <code>tempdir()</code> when Dropbox mode is enabled using <code>dropboxDir</code>.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-1-2018-04-24" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-2-1-2018-04-24" class="anchor"></a>bcbioRNASeq 0.2.1 (2018-04-24)</h1>
<p>Last set of code fixes before F1000v2 resubmission.</p>
<div id="major-changes-2" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-2" class="anchor"></a>Major changes</h2>
<ul>
<li>Added <code>rle</code> return support for <code><a href="../reference/counts.html">counts()</a></code>, which are calculated on the fly.</li>
<li>Added <code>transgeneNames</code> and <code>spikeNames</code> support to <code><a href="../reference/bcbioRNASeq.html">loadRNASeq()</a></code> function.</li>
<li>
<code><a href="../reference/bcbioRNASeq.html">loadRNASeq()</a></code> now supports <code>organism = NULL</code> again, for datasets with poorly annotated genomes.</li>
<li>Primary <code>assay()</code> containing raw counts is now named <code>counts</code> instead of <code>raw</code>, for consistency with other <code>SummarizedExperiment</code> objects (e.g. <code>DESeqDataSet</code>) and the bcbioSingleCell S4 class definition.</li>
<li>Improved internal code for <code><a href="../reference/plotGene.html">plotGene()</a></code> and <code>plotGenderMarkers()</code>.</li>
</ul>
</div>
<div id="minor-changes-5" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-5" class="anchor"></a>Minor changes</h2>
<ul>
<li>Improved AppVeyor CI support to test against bioc-devel using R 3.5.</li>
<li>Improved support and unit testing for <code><a href="../reference/updateObject.html">updateObject()</a></code> method.</li>
<li>DESeq2 normalized counts are always slotted in <code>assays()</code>, even when rlog and vst transformations are skipped.</li>
<li>Exporting <code>[[&lt;-</code>, <code>assays&lt;-</code>, <code>colData&lt;-</code>, <code>interestingGroups&lt;-</code>, and <code>metadata&lt;-</code> assignment methods, to avoid unwanted coercion to <code>SummarizedExperiment</code>. Objects extending <code>RangedSummarizedExperiment</code> shouldn’t be doing this, so we may need to file a bug report with Bioconductor or check our class definition in the package.</li>
<li>Now importing specific functions from S4Vectors and methods rather than importing everything.</li>
<li>Switched back to using <code>stop()</code>, <code>warning()</code> and <code>message()</code> rather than the alternate rlang functions <code>abort()</code>, <code>warn()</code>, and <code>inform()</code>.</li>
<li>Objects with invalid metadata now print which slots are invalid to the console.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-0-2018-03-22" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-2-0-2018-03-22" class="anchor"></a>bcbioRNASeq 0.2.0 (2018-03-22)</h1>
<div id="major-changes-3" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-3" class="anchor"></a>Major changes</h2>
<ul>
<li>
<code>bcbioRNASeq</code> S4 class object is now extending <code>RangedSummarizedExperiment</code> instead of <code>SummarizedExperiment</code>. Consequently, the row annotations are now stored in the <code>rowRanges</code> slot as <code>GRanges</code> class, instead of in the <code>rowData</code> slot as a <code>DataFrame</code>. The <code>rowData()</code> accessor still works and returns a data frame of gene/transcript annotations, but these are now coerced from the internally stored <code>GRanges</code>. The <code>GRanges</code> object is acquired automatically from [Ensembl][] using <code><a href="http://www.rdocumentation.org/packages/basejump/topics/deprecated">basejump::ensembl()</a></code>. By default, <code>GRanges</code> are acquired from [Ensembl][] using <a href="https://doi.org/doi:10.18129/B9.bioc.AnnotationHub">AnnotationHub</a> and [ensembldb][]. Legacy GRCh37 genome build is supported using the [EnsDb.Hsapiens.v75][] package.</li>
<li>
<code>assays()</code> now only slot matrices. We’ve moved the [tximport][] data from the now defunct <code>bcbio()</code> slot to assays. This includes the <code>lengths</code> matrix from [tximport][]. Additionally, we are optionally slotting [DESeq2][] variance-stabilized counts (“<code>rlog</code>”, <code>"vst"</code>). [DESeq2][] normalized counts and [edgeR][] TMM counts are calculated on the fly and no longer stored inside the <code>bcbioRNASeq</code> object.</li>
<li>
<code>colData()</code> now defaults to returning as <code>data.frame</code> instead of <code>DataFrame</code>, for easy piping to [tidyverse][] functions.</li>
<li>
<code>bcbio()</code> slot is now defunct.</li>
<li>FASTA spike-ins (e.g. EGFP, ERCCs) can be defined using the <code>isSpike</code> argument during the <code><a href="../reference/bcbioRNASeq.html">loadRNASeq()</a></code> data import step.</li>
<li>Melted counts are now scaled to log2 in the relevant quality control functions rather than using log10. This applies to <code>plotCountsPerGene()</code> and <code>plotCountDensity()</code>. Note that we are subsetting the nonzero genes as defined by the raw counts here.</li>
<li>Simplified internal <code>tximport()</code> code to no longer attempt to strip transcript versions. This is required for working with <em>C. elegans</em> transcripts.</li>
<li>Minimal working example dataset is now derived from GSE65267, which is also used in the F1000 paper.</li>
<li>Added <code><a href="../reference/coerce.html">as(object, "DESeqDataSet")</a></code> coercion method support for <code>bcbioRNASeq</code> class. This helps us set up the differential expression analysis easily.</li>
<li>
<code><a href="../reference/counts.html">counts()</a></code> function now returns [DESeq2][] normalized counts (<code>normalized =   TRUE</code>) and [edgeR][] TMM counts (<code>normalized = "tmm"</code>) on the fly, as suggested by the F1000 reviewers.</li>
<li>Design formula can no longer be slotted into <code>bcbioRNASeq</code> object, since we’re not stashing a <code>DESeqDataSet</code> any more.</li>
<li>Updated Functional Analysis R Markdown template.</li>
</ul>
</div>
<div id="minor-changes-6" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-6" class="anchor"></a>Minor changes</h2>
<ul>
<li>
<code>validObject()</code> is now required for all plotting functions. This check is also called in the R Markdown template. Legacy objects can be updated using <code><a href="../reference/updateObject.html">updateObject()</a></code>.</li>
<li>
<code><a href="../reference/metrics.html">metrics()</a></code> now returns columns sorted alphabetically.</li>
<li>Added <code>contrastName()</code> as a generic function.</li>
<li>
<code>plotDEGHeatmap()</code> and <code>plotDEGPCA()</code> generics no longer have <code>counts</code> defined in the signature. The <code>counts</code> argument is now only defined in the methods.</li>
<li>
<code><a href="../reference/prepareRNASeqTemplate.html">prepareRNASeqTemplate()</a></code> has been converted from a generic to a standard function.</li>
<li>Improved <code>metadata()</code> validity checks.</li>
<li>
<code><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap()</a></code> matrix method has been moved to <a href="http://steinbaugh.com/basejump">basejump</a> package, for improved consistency with the other heatmap code.</li>
<li>
<code>plotGenderMarkers()</code> internal code has been reworked to match <code><a href="../reference/plotGene.html">plotGene()</a></code>.</li>
<li>Default <code><a href="../reference/plotMA.html">plotMA()</a></code> appearance has changed, providing a line at the 0 y-intercept, similar to <code>DESeqDataSet</code> method.</li>
<li>Internal example datasets have been renamed (e.g. <code>bcb_small</code> instead of <code>bcb</code>).</li>
<li>Added AppVeyor CI support for code testing on Windows.</li>
<li>Made Travis CI checks stricter, added <code>BiocCheck()</code>.</li>
<li>Internal <code>.sampleDirs()</code> code is now exported in <a href="https://github.com/hbc/bcbioBase">bcbioBase</a> as a generic.</li>
<li>
<code>gene2symbol()</code> and <code>interestingGroups()</code> method support are now defined for <code>SummarizedExperiment</code> in the <a href="https://github.com/hbc/bcbioBase">bcbioBase</a> package.</li>
</ul>
</div>
<div id="updating-legacy-objects-v0-2-0" class="section level2">
<h2 class="hasAnchor">
<a href="#updating-legacy-objects-v0-2-0" class="anchor"></a>Updating legacy objects &lt; v0.2.0</h2>
<ul>
<li>Use <code><a href="../reference/updateObject.html">updateObject()</a></code> in combination with the <code>rowRanges</code> argument, which requires a <code>GRanges</code> object. <code>GRanges</code> can be obtained from [Ensembl][] using the <code><a href="http://www.rdocumentation.org/packages/basejump/topics/deprecated">basejump::ensembl()</a></code> function or the [ensembldb][] package.</li>
</ul>
</div>
<div id="deprecations" class="section level2">
<h2 class="hasAnchor">
<a href="#deprecations" class="anchor"></a>Deprecations</h2>
<ul>
<li>
<code>bcbio()</code> slot is now defunct, since we have moved all data into the <code>SummarizedExperiment</code> container.</li>
<li>Deprecated <code>plot5x3Bias()</code> in favor of <code>plot5Prime3PrimeBias()</code>. This is less confusing as to what this function plots.</li>
<li>
<code>flatFiles()</code> has been deprecated in favor of <code><a href="../reference/coerce.html">as(object, "list")</a></code> coercion method. See <a href="https://github.com/hbc/bcbioBase">bcbioBase</a> package for <code>SummarizedExperiment</code> method support.</li>
<li>Defunct: <code>design()</code>, <code><a href="../reference/deprecated.html">download()</a></code>, <code><a href="../reference/deprecated.html">meltLog10()</a></code>, <code><a href="../reference/deprecated.html">txi()</a></code>.</li>
<li>Legacy <code>bcbioRNADataSet</code> method support has been removed.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-1-8-2018-04-03" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-1-8-2018-04-03" class="anchor"></a>bcbioRNASeq 0.1.8 (2018-04-03)</h1>
<ul>
<li>Bug fix for <code>gene2symbol</code> argument not renaming rows in <code>plotDEGHeatmap()</code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-7-2018-02-28" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-1-7-2018-02-28" class="anchor"></a>bcbioRNASeq 0.1.7 (2018-02-28)</h1>
<ul>
<li>Bug fix for <code>[</code> subset method dropping metrics in metadata.</li>
<li>Simplified unit testing for Dropbox mode enabled in <code>resultsTables()</code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-6-2018-02-20" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-1-6-2018-02-20" class="anchor"></a>bcbioRNASeq 0.1.6 (2018-02-20)</h1>
<ul>
<li>Bug fix for gene-to-symbol mappings in <code>plotDEGHeatmap()</code>.</li>
<li>Added support for quickly plotting differentially expressed genes (DEG) in a PCA plot with <code>plotDEGPCA()</code>.</li>
<li>Added support for Dropbox shared links to <code>resultsTables()</code>, for use with the [Stem Cell Commons][] database.</li>
<li>Added assert checks internally for all functions.</li>
<li>Improved internal code for <code><a href="../reference/plotGene.html">plotGene()</a></code> and <code>plotGenderMarkers()</code> to run faster.</li>
<li>Deprecated data frame methods based on metrics for QC functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-5-2018-01-31" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-1-5-2018-01-31" class="anchor"></a>bcbioRNASeq 0.1.5 (2018-01-31)</h1>
<ul>
<li>Import shared dependency functions from bcbioBase instead of basejump.</li>
<li>Added method support for <code>selectSamples()</code>.</li>
<li>
<code>organism</code> and <code>genomeBuild</code> parameters are now user-definable in the main <code><a href="../reference/bcbioRNASeq.html">loadRNASeq()</a></code> import function.</li>
<li>Fixed gene subsetting method on S4 object, which handles genes using <code>intersect()</code> in the featureCounts matrix.</li>
<li>Removed internal <code><a href="../reference/aggregateReplicates.html">aggregateReplicates()</a></code> code. This needs to be reworked and added back in a future release.</li>
<li>Improve method for handling a missing normalized counts matrix in the assays slot. This can occur when the user opts to skip the CPU-intensive DESeq2 normalizations.</li>
<li>Improved internal code for the quality control functions. Improved the <code>if</code> statements to be more class specific.</li>
<li>Renamed <code><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap()</a></code> <code>transform</code> argument to <code>normalized</code>, for consistency with the <code><a href="../reference/counts.html">counts()</a></code> generic.</li>
<li>Added <code>title</code> support to plots, where applicable.</li>
<li>Updated internal code for <code>plotDEGHeatmap()</code>.</li>
<li>Updated internal marker handling code for <code>plotGenderMarkers()</code>.</li>
<li>
<code>resulsTables()</code> function now defaults to <code>summary = TRUE</code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-4-2018-11-27" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-1-4-2018-11-27" class="anchor"></a>bcbioRNASeq 0.1.4 (2018-11-27)</h1>
<ul>
<li>Migrated all basejump function imports to bcbioBase package.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-3-2017-12-03" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-1-3-2017-12-03" class="anchor"></a>bcbioRNASeq 0.1.3 (2017-12-03)</h1>
<ul>
<li>Combined examples (<code>bcb</code>, <code>dds</code>, <code>res</code>, etc.) into a single <code>examples</code> object. This helps avoid accidental use of example <code>bcb</code> in an analysis.</li>
<li>Moved ggplot imports from <code>internal-ggplot.R</code> to above each function.</li>
<li>Renamed <code>maxSamples</code> parameter in <code><a href="../reference/bcbioRNASeq.html">loadRNASeq()</a></code> to <code>transformationLimit</code>. If there are more samples than this limit, then the DESeq2 transformations will be skipped. In this case, <code>rlog</code> and <code>vst</code> will not be slotted into <code>assays()</code>.</li>
<li>Added a colData sanitization step in <code><a href="../reference/bcbioRNASeq.html">loadRNASeq()</a></code> to ensure rows are in the same order as the columns in the counts matrix. Otherwise, DESeq will report an error at the <code>DESeqDataSetFromTximport()</code> step. We’re also ensuring the factor levels get updated here.</li>
<li>Now using <code>glimpse()</code> instead of <code>str()</code> in examples, where applicable.</li>
<li>Added <code>colData&lt;-</code> assignment method support. This requires a <code>DataFrame</code> class object. Upon assignment, the internal colData at <code>bcbio(object,   "DESeqDataSet")</code>, <code>assays(object)[["rlog"]]</code> and <code>assays(object)[["vst"]]</code> are also updated to match.</li>
<li>Initial assignment support for <code>design</code>, which will update the internal DESeqDataSet.</li>
<li>Added method support for <code>gene2symbol()</code> generic, which will now return a 2 column <code>data.frame</code> with <code>ensgene</code> and <code>symbol</code> columns. This is helpful for downstream gene to symbol mapping operations.</li>
<li>Added working example for <code>interestingGroups&lt;-</code> in the documentation.</li>
<li>Added some code to improve factor releveling, where applicable. See <code>internal-meltLog10.R</code> for example.</li>
<li>Now explicitly defining the custom color palettes (e.g. <code><a href="http://www.rdocumentation.org/packages/viridis/topics/scale_viridis">viridis::scale_fill_viridis(discrete = TRUE)</a></code>. This makes it clearer to the user in the documentation where these palettes are located.</li>
<li>Improved axis label support in <code><a href="../reference/plotGene.html">plotGene()</a></code>.</li>
<li>
<code>plotHeatmap()</code> now uses internal <code>gene2symbol</code> mappings from stashed annotable, instead of always querying Ensembl. The user can define custom mappings with the <code>gene2symbol</code> argument, if desired.</li>
<li>
<code><a href="../reference/plotPCA.html">plotPCA()</a></code> now supports custom color palettes. The <code>shapes</code> parameter has been removed because it doesn’t work well and is limited to datasets with few samples. This behavior matches the PCA functionality in DESeq2.</li>
<li>Improved internal code for <code>plotVolcano()</code>. Added support for <code>gene2symbol</code> argument, like in <code>plotHeatmap()</code>. If left missing, the function will query Ensembl for the gene2symbol mappings. We’re now using <code>data</code> instead of <code>stats</code> as the main data source.</li>
<li>Improved legibility of subset method code.</li>
<li>Added some additional reexports, which are used for the package documentation and website.</li>
<li>Simplified legacy object coercion method code.</li>
<li>Updated Bioconductor installation method code. We’re now using the <code>dependencies</code> argument, which allows for automatic install of suggested packages along with imports.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-2-2017-11-08" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-1-2-2017-11-08" class="anchor"></a>bcbioRNASeq 0.1.2 (2017-11-08)</h1>
<ul>
<li>Updated package imports to match Bioconductor 3.6.</li>
<li>Added support for interesting groups assignment with <code>interestingGroups&lt;-</code>.</li>
<li>Renamed <code><a href="../reference/deprecated.html">plotGeneHeatmap()</a></code> to simply <code>plotHeatmap()</code>.</li>
<li>Added gender marker support for <em>Homo sapiens</em>.</li>
<li>Improved support for multiple interesting groups in quality control plots. Now interestingGroups is defined as a column in the metrics data.frame that is used to specify the plot color/fill. This matches the convention in the bcbioSingleCell 0.0.22 update.</li>
<li>Sample metadata columns are now consistently set as factors.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-1-2017-10-26" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-1-1-2017-10-26" class="anchor"></a>bcbioRNASeq 0.1.1 (2017-10-26)</h1>
<ul>
<li>Added support for coloring of multiple interesting groups in quality control plots.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-0-2017-10-23" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-1-0-2017-10-23" class="anchor"></a>bcbioRNASeq 0.1.0 (2017-10-23)</h1>
<ul>
<li>Updated version and author information to match the F1000 Research workflow.</li>
<li>Added an <code>f1000v1</code> branch containing the reproducible code used to generate the figures in our workflow.</li>
<li>Modified <code><a href="../reference/plotMA.html">plotMA()</a></code> to support vertical or horizontal layout return. Also added an argument to remove the color legend, which is typically not that informative.</li>
<li>Added custom color palette support to the quality control functions.</li>
<li>Upgrading from <code>bcbioRNADataSet</code> (&lt; 0.1.0) to <code>bcbioRNASeq</code> class object is now possible using <code><a href="../reference/coerce.html">as()</a></code> coercion method.</li>
<li>Object oriented methods are now restricted to use <code>bcbioRNASeq</code> object. Legacy <code>bcbioRNADataSet</code> objects must be upgraded to <code>bcbioRNASeq</code> class.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-28-2017-10-17" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-28-2017-10-17" class="anchor"></a>bcbioRNASeq 0.0.28 (2017-10-17)</h1>
<ul>
<li>Added support for output of unstructured data inside <code>bcbioRNASeq</code> S4 object using <code>flatFiles()</code> function.</li>
<li>Added <code>bcbioRNASeq</code> method support for <code>annotable()</code> generic.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-27-2017-10-10" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-27-2017-10-10" class="anchor"></a>bcbioRNASeq 0.0.27 (2017-10-10)</h1>
<ul>
<li>Renamed <code>bcbioRNADataSet</code> S4 class to <code>bcbioRNASeq</code>. This matches the naming conventions in the <a href="https://github.com/hbc/bcbioSingleCell">bcbioSingleCell</a> package.</li>
<li>Renamed <code><a href="../reference/deprecated.html">loadRNASeqRun()</a></code> to simply <code><a href="../reference/bcbioRNASeq.html">loadRNASeq()</a></code>.</li>
<li>Switched <code><a href="../reference/bcbioRNASeq.html">loadRNASeq()</a></code> from using S4 dispatch to a standard function.</li>
<li>Added a parameter argument to <code><a href="../reference/bcbioRNASeq.html">loadRNASeq()</a></code> that enables request of a specific Ensembl release version for gene annotations.</li>
<li>Renamed <code>interestingGroup</code> argument in quality control functions to <code>interestingGroups</code> for better consistency.</li>
<li>Improved handling of sample metrics in <code>plotPCACovariates()</code>.</li>
<li>Added functional analysis R Markdown template.</li>
<li>Offloaded some core functionality shared between [bcbioRNASeq][] and <a href="https://github.com/hbc/bcbioSingleCell">bcbioSingleCell</a> to the <a href="http://steinbaugh.com/basejump">basejump</a> package. This included some code to handle sample metadata YAML and file loading. This helps provide a consistent experience across both packages.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-26-2017-09-09" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-26-2017-09-09" class="anchor"></a>bcbioRNASeq 0.0.26 (2017-09-09)</h1>
<ul>
<li>Renamed package from bcbioRnaseq to bcbioRNASeq.</li>
<li>Improved website appearance.</li>
<li>Added [viridis][] color palette support to quality control functions.</li>
<li>Improved subset operations on <code>bcbioRNADataSet</code> object.</li>
<li>Fixed setup chunk loading of <code>bcbioRNADataSet</code> in differential expression [R Markdown][] template.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-25-2017-08-11" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-25-2017-08-11" class="anchor"></a>bcbioRNASeq 0.0.25 (2017-08-11)</h1>
<ul>
<li>Added S4 methods support for plots, allowing the user to use either <code>bcbioRNADataSet</code> or a metrics <code>data.frame</code> and manual <code>interesting_group</code> declaration for visualization.</li>
<li>Migrated function and variable names from <code>snake_case</code> to <code>camelCase</code>.</li>
<li>Offloaded small RNA functionality to a separate package named <a href="https://github.com/lpantano/bcbioSmallRna">bcbioSmallRNA</a>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-24-2017-07-13" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-24-2017-07-13" class="anchor"></a>bcbioRNASeq 0.0.24 (2017-07-13)</h1>
<ul>
<li>Reworked [R Markdown][] templates to improve YAML defaults and add more comments.</li>
<li>Modified default path variables in <code>setup.R</code> to use <code>*_dir</code> instead of <code>*_out</code>.</li>
<li>Updated NEWS file to use [Markdown][] syntax.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-23-2017-07-03" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-23-2017-07-03" class="anchor"></a>bcbioRNASeq 0.0.23 (2017-07-03)</h1>
<ul>
<li>Slotted <code>DESeqDataSet</code> using <code>design = formula(~1)</code> for quality control. This enables automatic generation of <code>rlog</code> and <code>vst</code> transformed counts.</li>
<li>Documentation fixes and website updates.</li>
<li>Renamed S4 class from <code>bcbioRnaDataSet</code> to <code>bcbioRNADataSet</code> (case sensitive).</li>
<li>Adjusted the number of exported functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-22-2017-06-21" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-22-2017-06-21" class="anchor"></a>bcbioRNASeq 0.0.22 (2017-06-21)</h1>
<ul>
<li>Added [testthat][] checking with [lintr][].</li>
<li>Initial setup of code coverage using [covr][].</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-21-2017-06-16" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-21-2017-06-16" class="anchor"></a>bcbioRNASeq 0.0.21 (2017-06-16)</h1>
<ul>
<li>Prepared draft of [F1000][] workflow document.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-20-2017-06-09" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-20-2017-06-09" class="anchor"></a>bcbioRNASeq 0.0.20 (2017-06-09)</h1>
<ul>
<li>Added [Travis-CI][] support for automatic rendering of quality control report.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-19-2017-06-07" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-19-2017-06-07" class="anchor"></a>bcbioRNASeq 0.0.19 (2017-06-07)</h1>
<ul>
<li>
<code>bcbioRnaDataSet</code> S4 definition updates.</li>
<li>Updates to <code>plot_pca()</code> and gene-level heatmaps.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-18-2017-05-24" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-18-2017-05-24" class="anchor"></a>bcbioRNASeq 0.0.18 (2017-05-24)</h1>
<ul>
<li>Simplified count pooling functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-17-2017-05-21" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-17-2017-05-21" class="anchor"></a>bcbioRNASeq 0.0.17 (2017-05-21)</h1>
<ul>
<li>Reduced number of exports and improved documentation.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-16-2017-05-18" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-16-2017-05-18" class="anchor"></a>bcbioRNASeq 0.0.16 (2017-05-18)</h1>
<ul>
<li>Draft migration of <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> run object into S4 <code>bcbioRnaDataSet</code>.</li>
<li>Created a new variant of <code>load_run()</code> that saves to S4 object instead of list.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-15-2017-05-15" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-15-2017-05-15" class="anchor"></a>bcbioRNASeq 0.0.15 (2017-05-15)</h1>
<ul>
<li>Reworked and re-organized internal functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-14-2017-05-10" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-14-2017-05-10" class="anchor"></a>bcbioRNASeq 0.0.14 (2017-05-10)</h1>
<ul>
<li>Defaulted to loading run using project summary YAML file.</li>
<li>Initial commit of [R Markdown][] templates (e.g. quality control).</li>
<li>Added support for dynamic file downloads from [HBC][] website.</li>
<li>Draft build of website using <code><a href="http://pkgdown.r-lib.org/reference/build_site.html">pkgdown::build_site()</a></code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-13-2017-05-08" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-13-2017-05-08" class="anchor"></a>bcbioRNASeq 0.0.13 (2017-05-08)</h1>
<ul>
<li>Improved [RDAVIDWebService][] utility functions to work with [dplyr][] 0.6.0.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-12-2017-05-01" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-12-2017-05-01" class="anchor"></a>bcbioRNASeq 0.0.12 (2017-05-01)</h1>
<ul>
<li>Reworked metadata and summary metrics functions to obtain information from <code>project-summary.yaml</code> saved in the final run directory.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-11-2017-04-27" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-11-2017-04-27" class="anchor"></a>bcbioRNASeq 0.0.11 (2017-04-27)</h1>
<ul>
<li>Reduced number of depdencies.</li>
<li>Initial commit of modified volcano plot from <a href="https://github.com/hbc/CHBUtils">CHBUtils</a> package.</li>
<li>Internal code updates for upcoming [dplyr][] 0.6.0/[tidyeval][] update.</li>
<li>Updated [Ensembl][] <a href="https://doi.org/doi:10.18129/B9.bioc.biomaRt">biomaRt</a> annotations to use live site, currently release 88.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-10-2017-04-19" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-10-2017-04-19" class="anchor"></a>bcbioRNASeq 0.0.10 (2017-04-19)</h1>
<ul>
<li>Renamed <code>import_*</code> functions to <code>read_*</code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-9-2017-04-13" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-9-2017-04-13" class="anchor"></a>bcbioRNASeq 0.0.9 (2017-04-13)</h1>
<ul>
<li>Consolidated NAMESPACE imports.</li>
<li>Defaulted to writing count matrices with gzip compression, to save disk space.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-8-2017-04-12" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-8-2017-04-12" class="anchor"></a>bcbioRNASeq 0.0.8 (2017-04-12)</h1>
<ul>
<li>Renamed internal parameters for better readability.</li>
<li>Improved documentation and consolidate functions by group.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-7-2017-04-10" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-7-2017-04-10" class="anchor"></a>bcbioRNASeq 0.0.7 (2017-04-10)</h1>
<ul>
<li>NAMESPACE simplification using <a href="http://steinbaugh.com/basejump">basejump</a> package.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-6-2017-04-07" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-6-2017-04-07" class="anchor"></a>bcbioRNASeq 0.0.6 (2017-04-07)</h1>
<ul>
<li>Reworked handling of plots and tables during knits.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-5-2017-04-06" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-5-2017-04-06" class="anchor"></a>bcbioRNASeq 0.0.5 (2017-04-06)</h1>
<ul>
<li>Initial commit of differential expression and gene set enrichment functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-4-2017-04-04" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-4-2017-04-04" class="anchor"></a>bcbioRNASeq 0.0.4 (2017-04-04)</h1>
<ul>
<li>Added <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> object integrity checks.</li>
<li>Improved detection and handling of lane split samples.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-3-2017-03-31" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-3-2017-03-31" class="anchor"></a>bcbioRNASeq 0.0.3 (2017-03-31)</h1>
<ul>
<li>Reworked functions to utilize <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> list object.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-2-2017-03-28" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-2-2017-03-28" class="anchor"></a>bcbioRNASeq 0.0.2 (2017-03-28)</h1>
<ul>
<li>Added plotting functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-1-2017-03-22" class="section level1">
<h1 class="page-header">
<a href="#bcbiornaseq-0-0-1-2017-03-22" class="anchor"></a>bcbioRNASeq 0.0.1 (2017-03-22)</h1>
<ul>
<li>Start of package development.</li>
<li>Initial draft release supporting automatic loading of <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> run data.</li>
</ul>
<p>[clusterProfiler]: [covr]: <a href="https://github.com/jimhester/covr" class="uri">https://github.com/jimhester/covr</a> [dplyr]: <a href="http://dplyr.tidyverse.org" class="uri">http://dplyr.tidyverse.org</a> [edgeR]: <a href="https://doi.org/doi:10.18129/B9.bioc.edgeR" class="uri">https://doi.org/doi:10.18129/B9.bioc.edgeR</a> [Ensembl]: <a href="http://www.ensembl.org" class="uri">http://www.ensembl.org</a> [ensembldb]: <a href="https://doi.org/doi:10.18129/B9.bioc.ensembldb" class="uri">https://doi.org/doi:10.18129/B9.bioc.ensembldb</a> [F1000]: <a href="https://f1000.com" class="uri">https://f1000.com</a> [HBC]: <a href="http://bioinformatics.sph.harvard.edu" class="uri">http://bioinformatics.sph.harvard.edu</a> [lintr]: <a href="https://github.com/jimhester/lintr" class="uri">https://github.com/jimhester/lintr</a> [macOS]: <a href="https://www.apple.com/macos" class="uri">https://www.apple.com/macos</a> [Markdown]: <a href="https://daringfireball.net/projects/markdown/syntax" class="uri">https://daringfireball.net/projects/markdown/syntax</a> [RDAVIDWebService]: <a href="https://doi.org/doi:10.18129/B9.bioc.RDAVIDWebService" class="uri">https://doi.org/doi:10.18129/B9.bioc.RDAVIDWebService</a> [R Markdown]: <a href="http://rmarkdown.rstudio.com" class="uri">http://rmarkdown.rstudio.com</a> [Stem Cell Commons]: <a href="http://stemcellcommons.org" class="uri">http://stemcellcommons.org</a> [testthat]: <a href="https://github.com/hadley/testthat" class="uri">https://github.com/hadley/testthat</a> [tidyeval]: <a href="http://dplyr.tidyverse.org/articles/programming.html" class="uri">http://dplyr.tidyverse.org/articles/programming.html</a> [tidyverse]: <a href="https://www.tidyverse.org/" class="uri">https://www.tidyverse.org/</a> [tximport]: <a href="https://doi.org/doi:10.18129/B9.bioc.tximport" class="uri">https://doi.org/doi:10.18129/B9.bioc.tximport</a> [Travis-CI]: <a href="https://travis-ci.org" class="uri">https://travis-ci.org</a> [viridis]: <a href="https://cran.r-project.org/web/packages/viridis/index.html" class="uri">https://cran.r-project.org/web/packages/viridis/index.html</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bcbiornaseq-0-2-6-2018-07-09">0.2.6</a></li>
        <li><a href="#bcbiornaseq-0-2-5-2018-06-21">0.2.5</a></li>
        <li><a href="#bcbiornaseq-0-2-4-2018-05-24">0.2.4</a></li>
        <li><a href="#bcbiornaseq-0-2-3-2018-05-10">0.2.3</a></li>
        <li><a href="#bcbiornaseq-0-2-2-2018-04-26">0.2.2</a></li>
        <li><a href="#bcbiornaseq-0-2-1-2018-04-24">0.2.1</a></li>
        <li><a href="#bcbiornaseq-0-2-0-2018-03-22">0.2.0</a></li>
        <li><a href="#bcbiornaseq-0-1-8-2018-04-03">0.1.8</a></li>
        <li><a href="#bcbiornaseq-0-1-7-2018-02-28">0.1.7</a></li>
        <li><a href="#bcbiornaseq-0-1-6-2018-02-20">0.1.6</a></li>
        <li><a href="#bcbiornaseq-0-1-5-2018-01-31">0.1.5</a></li>
        <li><a href="#bcbiornaseq-0-1-4-2018-11-27">0.1.4</a></li>
        <li><a href="#bcbiornaseq-0-1-3-2017-12-03">0.1.3</a></li>
        <li><a href="#bcbiornaseq-0-1-2-2017-11-08">0.1.2</a></li>
        <li><a href="#bcbiornaseq-0-1-1-2017-10-26">0.1.1</a></li>
        <li><a href="#bcbiornaseq-0-1-0-2017-10-23">0.1.0</a></li>
        <li><a href="#bcbiornaseq-0-0-28-2017-10-17">0.0.28</a></li>
        <li><a href="#bcbiornaseq-0-0-27-2017-10-10">0.0.27</a></li>
        <li><a href="#bcbiornaseq-0-0-26-2017-09-09">0.0.26</a></li>
        <li><a href="#bcbiornaseq-0-0-25-2017-08-11">0.0.25</a></li>
        <li><a href="#bcbiornaseq-0-0-24-2017-07-13">0.0.24</a></li>
        <li><a href="#bcbiornaseq-0-0-23-2017-07-03">0.0.23</a></li>
        <li><a href="#bcbiornaseq-0-0-22-2017-06-21">0.0.22</a></li>
        <li><a href="#bcbiornaseq-0-0-21-2017-06-16">0.0.21</a></li>
        <li><a href="#bcbiornaseq-0-0-20-2017-06-09">0.0.20</a></li>
        <li><a href="#bcbiornaseq-0-0-19-2017-06-07">0.0.19</a></li>
        <li><a href="#bcbiornaseq-0-0-18-2017-05-24">0.0.18</a></li>
        <li><a href="#bcbiornaseq-0-0-17-2017-05-21">0.0.17</a></li>
        <li><a href="#bcbiornaseq-0-0-16-2017-05-18">0.0.16</a></li>
        <li><a href="#bcbiornaseq-0-0-15-2017-05-15">0.0.15</a></li>
        <li><a href="#bcbiornaseq-0-0-14-2017-05-10">0.0.14</a></li>
        <li><a href="#bcbiornaseq-0-0-13-2017-05-08">0.0.13</a></li>
        <li><a href="#bcbiornaseq-0-0-12-2017-05-01">0.0.12</a></li>
        <li><a href="#bcbiornaseq-0-0-11-2017-04-27">0.0.11</a></li>
        <li><a href="#bcbiornaseq-0-0-10-2017-04-19">0.0.10</a></li>
        <li><a href="#bcbiornaseq-0-0-9-2017-04-13">0.0.9</a></li>
        <li><a href="#bcbiornaseq-0-0-8-2017-04-12">0.0.8</a></li>
        <li><a href="#bcbiornaseq-0-0-7-2017-04-10">0.0.7</a></li>
        <li><a href="#bcbiornaseq-0-0-6-2017-04-07">0.0.6</a></li>
        <li><a href="#bcbiornaseq-0-0-5-2017-04-06">0.0.5</a></li>
        <li><a href="#bcbiornaseq-0-0-4-2017-04-04">0.0.4</a></li>
        <li><a href="#bcbiornaseq-0-0-3-2017-03-31">0.0.3</a></li>
        <li><a href="#bcbiornaseq-0-0-2-2017-03-28">0.0.2</a></li>
        <li><a href="#bcbiornaseq-0-0-1-2017-03-22">0.0.1</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

