<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>All news • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRNASeq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">

  <div class="col-md-9">
    <div class="page-header">
      <h1>Change log <small>All releases</small></h1>
    </div>

    <div class="contents">
    <div id="bcbiornaseq-0-1-6-2018-02-16" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-6-2018-02-16" class="anchor"></a>bcbioRNASeq 0.1.6 (2018-02-16)</h2>
<ul>
<li>Bug fix for gene-to-symbol mappings in <code>plotDEGHeatmap()</code>.</li>
<li>Added support for quickly plotting differentially expressed genes (DEG) in a PCA plot with <code>plotDEGPCA()</code>.</li>
<li>Added support for Dropbox shared links to <code>resultsTables()</code>, for use with the <a href="http://stemcellcommons.org">Stem Cell Commons</a> database.</li>
<li>Added assert checks internally for all functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-5-2018-01-31" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-5-2018-01-31" class="anchor"></a>bcbioRNASeq 0.1.5 (2018-01-31)</h2>
<ul>
<li>Import shared dependency functions from bcbioBase instead of basejump.</li>
<li>Added method support for <code><a href="../reference/selectSamples.html">selectSamples()</a></code>.</li>
<li>
<code>organism</code> and <code>genomeBuild</code> parameters are now user-definable in the main <code><a href="../reference/loadRNASeq.html">loadRNASeq()</a></code> import function.</li>
<li>Fixed gene subsetting method on S4 object, which handles genes using <code>intersect()</code> in the featureCounts matrix.</li>
<li>Removed internal <code>aggregateReplicates()</code> code. This needs to be reworked and added back in a future release.</li>
<li>Improve method for handling a missing normalized counts matrix in the assays slot. This can occur when the user opts to skip the CPU-intensive DESeq2 normalizations.</li>
<li>Improved internal code for the quality control functions. Improved the <code>if</code> statements to be more class specific.</li>
<li>Renamed <code>plotCorrelationHeatmap()</code> <code>transform</code> argument to <code>normalized</code>, for consistency with the <code><a href="../reference/counts.html">counts()</a></code> generic.</li>
<li>Added <code>title</code> support to plots, where applicable.</li>
<li>Updated internal code for <code>plotDEGHeatmap()</code>.</li>
<li>Updated internal marker handling code for <code>plotGenderMarkers()</code>.</li>
<li>
<code>resulsTables()</code> function now defaults to <code>summary = TRUE</code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-4-2018-11-27" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-4-2018-11-27" class="anchor"></a>bcbioRNASeq 0.1.4 (2018-11-27)</h2>
<ul>
<li>Migrated all basejump function imports to bcbioBase package.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-3-2017-12-03" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-3-2017-12-03" class="anchor"></a>bcbioRNASeq 0.1.3 (2017-12-03)</h2>
<ul>
<li>Combined examples (<code>bcb</code>, <code>dds</code>, <code>res</code>, etc.) into a single <code>examples</code> object. This helps avoid accidental use of example <code>bcb</code> in an analysis.</li>
<li>Moved ggplot imports from <code>internal-ggplot.R</code> to above each function.</li>
<li>Renamed <code>maxSamples</code> parameter in <code><a href="../reference/loadRNASeq.html">loadRNASeq()</a></code> to <code>transformationLimit</code>. If there are more samples than this limit, then the DESeq2 transformations will be skipped. In this case, <code>rlog</code> and <code>vst</code> will not be slotted into <code>assays()</code>.</li>
<li>Added a colData sanitization step in <code><a href="../reference/loadRNASeq.html">loadRNASeq()</a></code> to ensure rows are in the same order as the columns in the counts matrix. Otherwise, DESeq will report an error at the <code>DESeqDataSetFromTximport()</code> step. We’re also ensuring the factor levels get updated here.</li>
<li>Now using <code><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse()</a></code> instead of <code>str()</code> in examples, where applicable.</li>
<li>Added <code>colData&lt;-</code> assignment method support. This requires a <code>DataFrame</code> class object. Upon assignment, the internal colData at <code><a href="../reference/bcbio.html">bcbio(object, "DESeqDataSet")</a></code>, <code>assays(object)[["rlog"]]</code> and <code>assays(object)[["vst"]]</code> are also updated to match.</li>
<li>Initial assignment support for <code>design</code>, which will update the internal DESeqDataSet.</li>
<li>Added method support for <code><a href="../reference/gene2symbol.html">gene2symbol()</a></code> generic, which will now return a 2 column <code>data.frame</code> with <code>ensgene</code> and <code>symbol</code> columns. This is helpful for downstream gene to symbol mapping operations.</li>
<li>Added working example for <code>interestingGroups&lt;-</code> in the documentation.</li>
<li>Added some code to improve factor releveling, where applicable. See <code>internal-meltLog10.R</code> for example.</li>
<li>Now explicitly defining the custom color palettes (e.g. <code><a href="http://www.rdocumentation.org/packages/viridis/topics/scale_viridis">viridis::scale_fill_viridis(discrete = TRUE)</a></code>. This makes it clearer to the user in the documentation where these palettes are located.</li>
<li>Improved axis label support in <code><a href="../reference/plotGene.html">plotGene()</a></code>.</li>
<li>
<code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code> now uses internal <code>gene2symbol</code> mappings from stashed annotable, instead of always querying Ensembl. The user can define custom mappings with the <code>gene2symbol</code> argument, if desired.</li>
<li>
<code><a href="../reference/plotPCA.html">plotPCA()</a></code> now supports custom color palettes. The <code>shapes</code> parameter has been removed because it doesn’t work well and is limited to datasets with few samples. This behavior matches the PCA functionality in DESeq2.</li>
<li>Improved internal code for <code>plotVolcano()</code>. Added support for <code>gene2symbol</code> argument, like in <code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code>. If left missing, the function will query Ensembl for the gene2symbol mappings. We’re now using <code>data</code> instead of <code>stats</code> as the main data source.</li>
<li>Improved legibility of subset method code.</li>
<li>Added some additional reexports, which are used for the package documentation and website.</li>
<li>Simplified legacy object coercion method code.</li>
<li>Updated Bioconductor installation method code. We’re now using the <code>dependencies</code> argument, which allows for automatic install of suggested packages along with imports.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-2-2017-11-08" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-2-2017-11-08" class="anchor"></a>bcbioRNASeq 0.1.2 (2017-11-08)</h2>
<ul>
<li>Updated package imports to match Bioconductor 3.6.</li>
<li>Added support for interesting groups assignment with <code>interestingGroups&lt;-</code>.</li>
<li>Renamed <code><a href="../reference/deprecated.html">plotGeneHeatmap()</a></code> to simply <code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code>.</li>
<li>Added gender marker support for <em>Homo sapiens</em>.</li>
<li>Improved support for multiple interesting groups in quality control plots. Now interestingGroups is defined as a column in the metrics data.frame that is used to specify the plot color/fill. This matches the convention in the bcbioSingleCell 0.0.22 update.</li>
<li>Sample metadata columns are now consistently set as factors.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-1-2017-10-26" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-1-2017-10-26" class="anchor"></a>bcbioRNASeq 0.1.1 (2017-10-26)</h2>
<ul>
<li>Added support for coloring of multiple interesting groups in quality control plots.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-0-2017-10-23" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-0-2017-10-23" class="anchor"></a>bcbioRNASeq 0.1.0 (2017-10-23)</h2>
<ul>
<li>Updated version and author information to match the F1000 Research workflow.</li>
<li>Added an <code>f1000v1</code> branch containing the reproducible code used to generate the figures in our workflow.</li>
<li>Modified <code><a href="../reference/plotMA.html">plotMA()</a></code> to support vertical or horizontal layout return. Also added an argument to remove the color legend, which is typically not that informative.</li>
<li>Added custom color palette support to the quality control functions.</li>
<li>Upgrading from <code>bcbioRNADataSet</code> (&lt; 0.1.0) to <code>bcbioRNASeq</code> class object is now possible using <code>as()</code> coercion method.</li>
<li>Object oriented methods are now restricted to use <code>bcbioRNASeq</code> object. Legacy <code>bcbioRNADataSet</code> objects must be upgraded to <code>bcbioRNASeq</code> class.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-28-2017-10-17" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-28-2017-10-17" class="anchor"></a>bcbioRNASeq 0.0.28 (2017-10-17)</h2>
<ul>
<li>Added support for output of unstructured data inside <code>bcbioRNASeq</code> S4 object using <code><a href="../reference/flatFiles.html">flatFiles()</a></code> function.</li>
<li>Added <code>bcbioRNASeq</code> method support for <code><a href="../reference/annotable.html">annotable()</a></code> generic.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-27-2017-10-10" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-27-2017-10-10" class="anchor"></a>bcbioRNASeq 0.0.27 (2017-10-10)</h2>
<ul>
<li>Renamed <code>bcbioRNADataSet</code> S4 class to <code>bcbioRNASeq</code>. This matches the naming conventions in the <a href="https://github.com/hbc/bcbioSingleCell">bcbioSingleCell</a> package.</li>
<li>Renamed <code><a href="../reference/deprecated.html">loadRNASeqRun()</a></code> to simply <code><a href="../reference/loadRNASeq.html">loadRNASeq()</a></code>.</li>
<li>Switched <code><a href="../reference/loadRNASeq.html">loadRNASeq()</a></code> from using S4 dispatch to a standard function.</li>
<li>Added a parameter argument to <code><a href="../reference/loadRNASeq.html">loadRNASeq()</a></code> that enables request of a specific Ensembl release version for gene annotations.</li>
<li>Renamed <code>interestingGroup</code> argument in quality control functions to <code>interestingGroups</code> for better consistency.</li>
<li>Improved handling of sample metrics in <code>plotPCACovariates()</code>.</li>
<li>Added functional analysis R Markdown template.</li>
<li>Offloaded some core functionality shared between [bcbioRNASeq][] and <a href="https://github.com/hbc/bcbioSingleCell">bcbioSingleCell</a> to the <a href="http://steinbaugh.com/basejump">basejump</a> package. This included some code to handle sample metadata YAML and file loading. This helps provide a consistent experience across both packages.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-26-2017-09-09" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-26-2017-09-09" class="anchor"></a>bcbioRNASeq 0.0.26 (2017-09-09)</h2>
<ul>
<li>Renamed package from bcbioRnaseq to bcbioRNASeq.</li>
<li>Improved website appearance.</li>
<li>Added <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palette support to quality control functions.</li>
<li>Improved subset operations on <code>bcbioRNADataSet</code> object.</li>
<li>Fixed setup chunk loading of <code>bcbioRNADataSet</code> in differential expression <a href="http://rmarkdown.rstudio.com">RMarkdown</a> template.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-25-2017-08-11" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-25-2017-08-11" class="anchor"></a>bcbioRNASeq 0.0.25 (2017-08-11)</h2>
<ul>
<li>Added S4 methods support for plots, allowing the user to use either <code>bcbioRNADataSet</code> or a metrics <code>data.frame</code> and manual <code>interesting_group</code> declaration for visualization.</li>
<li>Migrated function and variable names from <code>snake_case</code> to <code>camelCase</code>.</li>
<li>Offloaded small RNA functionality to a separate package named <a href="https://github.com/lpantano/bcbioSmallRna">bcbioSmallRNA</a>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-24-2017-07-13" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-24-2017-07-13" class="anchor"></a>bcbioRNASeq 0.0.24 (2017-07-13)</h2>
<ul>
<li>Reworked <a href="http://rmarkdown.rstudio.com">RMarkdown</a> templates to improve YAML defaults and add more comments.</li>
<li>Modified default path variables in <code>setup.R</code> to use <code>*_dir</code> instead of <code>*_out</code>.</li>
<li>Updated NEWS file to use <a href="https://daringfireball.net/projects/markdown/syntax">Markdown</a> syntax.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-23-2017-07-03" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-23-2017-07-03" class="anchor"></a>bcbioRNASeq 0.0.23 (2017-07-03)</h2>
<ul>
<li>Slotted <code>DESeqDataSet</code> using <code>design = formula(~1)</code> for quality control. This enables automatic generation of <code>rlog</code> and <code>vst</code> transformed counts.</li>
<li>Documentation fixes and website updates.</li>
<li>Renamed S4 class from <code>bcbioRnaDataSet</code> to <code>bcbioRNADataSet</code> (case sensitive).</li>
<li>Adjusted the number of exported functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-22-2017-06-21" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-22-2017-06-21" class="anchor"></a>bcbioRNASeq 0.0.22 (2017-06-21)</h2>
<ul>
<li>Added <a href="https://github.com/hadley/testthat">testthat</a> checking with <a href="https://github.com/jimhester/lintr">lintr</a>.</li>
<li>Initial setup of code coverage using <a href="https://github.com/jimhester/covr">covr</a>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-21-2017-06-16" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-21-2017-06-16" class="anchor"></a>bcbioRNASeq 0.0.21 (2017-06-16)</h2>
<ul>
<li>Prepared draft of <a href="https://f1000.com">F1000</a> workflow document.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-20-2017-06-09" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-20-2017-06-09" class="anchor"></a>bcbioRNASeq 0.0.20 (2017-06-09)</h2>
<ul>
<li>Added <a href="https://travis-ci.org">Travis-CI</a> support for automatic rendering of quality control report.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-19-2017-06-07" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-19-2017-06-07" class="anchor"></a>bcbioRNASeq 0.0.19 (2017-06-07)</h2>
<ul>
<li>
<code>bcbioRnaDataSet</code> S4 definition updates.</li>
<li>Updates to <code>plot_pca()</code> and gene-level heatmaps.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-18-2017-05-24" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-18-2017-05-24" class="anchor"></a>bcbioRNASeq 0.0.18 (2017-05-24)</h2>
<ul>
<li>Simplified count pooling functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-17-2017-05-21" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-17-2017-05-21" class="anchor"></a>bcbioRNASeq 0.0.17 (2017-05-21)</h2>
<ul>
<li>Reduced number of exports and improved documentation.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-16-2017-05-18" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-16-2017-05-18" class="anchor"></a>bcbioRNASeq 0.0.16 (2017-05-18)</h2>
<ul>
<li>Draft migration of <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> run object into S4 <code>bcbioRnaDataSet</code>.</li>
<li>Created a new variant of <code>load_run()</code> that saves to S4 object instead of list.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-15-2017-05-15" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-15-2017-05-15" class="anchor"></a>bcbioRNASeq 0.0.15 (2017-05-15)</h2>
<ul>
<li>Reworked and re-organized internal functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-14-2017-05-10" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-14-2017-05-10" class="anchor"></a>bcbioRNASeq 0.0.14 (2017-05-10)</h2>
<ul>
<li>Defaulted to loading run using project summary YAML file.</li>
<li>Initial commit of <a href="http://rmarkdown.rstudio.com">RMarkdown</a> templates (e.g. quality control).</li>
<li>Added support for dynamic file downloads from <a href="http://bioinformatics.sph.harvard.edu">HBC</a> website.</li>
<li>Draft build of website using <code><a href="http://pkgdown.r-lib.org/reference/build_site.html">pkgdown::build_site()</a></code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-13-2017-05-08" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-13-2017-05-08" class="anchor"></a>bcbioRNASeq 0.0.13 (2017-05-08)</h2>
<ul>
<li>Improved <a href="https://bioconductor.org/packages/release/bioc/html/RDAVIDWebService.html">RDAVIDWebService</a> utility functions to work with <a href="http://dplyr.tidyverse.org">dplyr</a> 0.6.0.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-12-2017-05-01" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-12-2017-05-01" class="anchor"></a>bcbioRNASeq 0.0.12 (2017-05-01)</h2>
<ul>
<li>Reworked metadata and summary metrics functions to obtain information from <code>project-summary.yaml</code> saved in the final run directory.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-11-2017-04-27" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-11-2017-04-27" class="anchor"></a>bcbioRNASeq 0.0.11 (2017-04-27)</h2>
<ul>
<li>Reduced number of depdencies.</li>
<li>Initial commit of modified volcano plot from <a href="https://github.com/hbc/CHBUtils">CHBUtils</a> package.</li>
<li>Internal code updates for upcoming <a href="http://dplyr.tidyverse.org">dplyr</a> 0.6.0/<a href="http://dplyr.tidyverse.org/articles/programming.html">tidyeval</a> update.</li>
<li>Updated <a href="http://www.ensembl.org">Ensembl</a> <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> annotations to use live site, currently release 88.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-10-2017-04-19" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-10-2017-04-19" class="anchor"></a>bcbioRNASeq 0.0.10 (2017-04-19)</h2>
<ul>
<li>Renamed <code>import_*</code> functions to <code>read_*</code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-9-2017-04-13" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-9-2017-04-13" class="anchor"></a>bcbioRNASeq 0.0.9 (2017-04-13)</h2>
<ul>
<li>Consolidated NAMESPACE imports.</li>
<li>Defaulted to writing count matrices with gzip compression, to save disk space.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-8-2017-04-12" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-8-2017-04-12" class="anchor"></a>bcbioRNASeq 0.0.8 (2017-04-12)</h2>
<ul>
<li>Renamed internal parameters for better readability.</li>
<li>Improved documentation and consolidate functions by group.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-7-2017-04-10" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-7-2017-04-10" class="anchor"></a>bcbioRNASeq 0.0.7 (2017-04-10)</h2>
<ul>
<li>NAMESPACE simplification using <a href="http://steinbaugh.com/basejump">basejump</a> package.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-6-2017-04-07" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-6-2017-04-07" class="anchor"></a>bcbioRNASeq 0.0.6 (2017-04-07)</h2>
<ul>
<li>Reworked handling of plots and tables during knits.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-5-2017-04-06" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-5-2017-04-06" class="anchor"></a>bcbioRNASeq 0.0.5 (2017-04-06)</h2>
<ul>
<li>Initial commit of differential expression and gene set enrichment functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-4-2017-04-04" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-4-2017-04-04" class="anchor"></a>bcbioRNASeq 0.0.4 (2017-04-04)</h2>
<ul>
<li>Added <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> object integrity checks.</li>
<li>Improved detection and handling of lane split samples.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-3-2017-03-31" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-3-2017-03-31" class="anchor"></a>bcbioRNASeq 0.0.3 (2017-03-31)</h2>
<ul>
<li>Reworked functions to utilize <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> list object.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-2-2017-03-28" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-2-2017-03-28" class="anchor"></a>bcbioRNASeq 0.0.2 (2017-03-28)</h2>
<ul>
<li>Added plotting functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-1-2017-03-22" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-0-1-2017-03-22" class="anchor"></a>bcbioRNASeq 0.0.1 (2017-03-22)</h2>
<ul>
<li>Start of package development.</li>
<li>Initial draft release supporting automatic loading of <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> run data.</li>
</ul>
</div>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bcbiornaseq-0-1-6-2018-02-16">0.1.6</a></li>
        <li><a href="#bcbiornaseq-0-1-5-2018-01-31">0.1.5</a></li>
        <li><a href="#bcbiornaseq-0-1-4-2018-11-27">0.1.4</a></li>
        <li><a href="#bcbiornaseq-0-1-3-2017-12-03">0.1.3</a></li>
        <li><a href="#bcbiornaseq-0-1-2-2017-11-08">0.1.2</a></li>
        <li><a href="#bcbiornaseq-0-1-1-2017-10-26">0.1.1</a></li>
        <li><a href="#bcbiornaseq-0-1-0-2017-10-23">0.1.0</a></li>
        <li><a href="#bcbiornaseq-0-0-28-2017-10-17">0.0.28</a></li>
        <li><a href="#bcbiornaseq-0-0-27-2017-10-10">0.0.27</a></li>
        <li><a href="#bcbiornaseq-0-0-26-2017-09-09">0.0.26</a></li>
        <li><a href="#bcbiornaseq-0-0-25-2017-08-11">0.0.25</a></li>
        <li><a href="#bcbiornaseq-0-0-24-2017-07-13">0.0.24</a></li>
        <li><a href="#bcbiornaseq-0-0-23-2017-07-03">0.0.23</a></li>
        <li><a href="#bcbiornaseq-0-0-22-2017-06-21">0.0.22</a></li>
        <li><a href="#bcbiornaseq-0-0-21-2017-06-16">0.0.21</a></li>
        <li><a href="#bcbiornaseq-0-0-20-2017-06-09">0.0.20</a></li>
        <li><a href="#bcbiornaseq-0-0-19-2017-06-07">0.0.19</a></li>
        <li><a href="#bcbiornaseq-0-0-18-2017-05-24">0.0.18</a></li>
        <li><a href="#bcbiornaseq-0-0-17-2017-05-21">0.0.17</a></li>
        <li><a href="#bcbiornaseq-0-0-16-2017-05-18">0.0.16</a></li>
        <li><a href="#bcbiornaseq-0-0-15-2017-05-15">0.0.15</a></li>
        <li><a href="#bcbiornaseq-0-0-14-2017-05-10">0.0.14</a></li>
        <li><a href="#bcbiornaseq-0-0-13-2017-05-08">0.0.13</a></li>
        <li><a href="#bcbiornaseq-0-0-12-2017-05-01">0.0.12</a></li>
        <li><a href="#bcbiornaseq-0-0-11-2017-04-27">0.0.11</a></li>
        <li><a href="#bcbiornaseq-0-0-10-2017-04-19">0.0.10</a></li>
        <li><a href="#bcbiornaseq-0-0-9-2017-04-13">0.0.9</a></li>
        <li><a href="#bcbiornaseq-0-0-8-2017-04-12">0.0.8</a></li>
        <li><a href="#bcbiornaseq-0-0-7-2017-04-10">0.0.7</a></li>
        <li><a href="#bcbiornaseq-0-0-6-2017-04-07">0.0.6</a></li>
        <li><a href="#bcbiornaseq-0-0-5-2017-04-06">0.0.5</a></li>
        <li><a href="#bcbiornaseq-0-0-4-2017-04-04">0.0.4</a></li>
        <li><a href="#bcbiornaseq-0-0-3-2017-03-31">0.0.3</a></li>
        <li><a href="#bcbiornaseq-0-0-2-2017-03-28">0.0.2</a></li>
        <li><a href="#bcbiornaseq-0-0-1-2017-03-22">0.0.1</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
