<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>All news • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRNASeq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">

  <div class="col-md-9">
    <div class="page-header">
      <h1>Change log <small>All releases</small></h1>
    </div>

    <div class="contents">
    <div id="bcbiornaseq-0-1-1" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-1-1" class="anchor"></a>bcbioRNASeq 0.1.1</h1>
<ul>
<li>Added support for coloring of multiple interesting groups in quality control plots.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-0" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-1-0" class="anchor"></a>bcbioRNASeq 0.1.0</h1>
<ul>
<li>Updated version and author information to match the F1000 Research workflow.</li>
<li>Added an <code>f1000v1</code> branch containing the reproducible code used to generate the figures in our workflow.</li>
<li>Modified <code>plotMA()</code> to support vertical or horizontal layout return. Also added an argument to remove the color legend, which is typically not that informative.</li>
<li>Added custom color palette support to the quality control functions.</li>
<li>Upgrading from <code>bcbioRNADataSet</code> (&lt; 0.1.0) to <code>bcbioRNASeq</code> class object is now possible using <code>as()</code> coercion method.</li>
<li>Object oriented methods are now restricted to use <code>bcbioRNASeq</code> object. Legacy <code>bcbioRNADataSet</code> objects must be upgraded to <code>bcbioRNASeq</code> class.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-28" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-28" class="anchor"></a>bcbioRNASeq 0.0.28</h1>
<ul>
<li>Added support for output of unstructured data inside <code>bcbioRNASeq</code> S4 object using <code>flatFiles()</code> function.</li>
<li>Added <code>bcbioRNASeq</code> method support for <code>annotable()</code> generic.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-27" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-27" class="anchor"></a>bcbioRNASeq 0.0.27</h1>
<ul>
<li>Renamed <code>bcbioRNADataSet</code> S4 class to <code>bcbioRNASeq</code>. This matches the naming conventions in the <a href="https://github.com/hbc/bcbioSingleCell">bcbioSingleCell</a> package.</li>
<li>Renamed <code>loadRNASeqRun()</code> to simply <code>loadRNASeq()</code>.</li>
<li>Switched <code>loadRNASeq()</code> from using S4 dispatch to a standard function.</li>
<li>Added a parameter argument to <code>loadRNASeq()</code> that enables request of a specific Ensembl release version for gene annotations.</li>
<li>Renamed <code>interestingGroup</code> argument in quality control functions to <code>interestingGroups</code> for better consistency.</li>
<li>Improved handling of sample metrics in <code>plotPCACovariates()</code>.</li>
<li>Added functional analysis R Markdown template.</li>
<li>Offloaded some core functionality shared between [bcbioRNASeq][] and <a href="https://github.com/hbc/bcbioSingleCell">bcbioSingleCell</a> to the <a href="http://steinbaugh.com/basejump">basejump</a> package. This included some code to handle sample metadata YAML and file loading. This helps provide a consistent experience across both packages.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-26" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-26" class="anchor"></a>bcbioRNASeq 0.0.26</h1>
<ul>
<li>Renamed package from bcbioRnaseq to bcbioRNASeq.</li>
<li>Improved website appearance.</li>
<li>Added <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palette support to quality control functions.</li>
<li>Improved subset operations on <code>bcbioRNADataSet</code> object.</li>
<li>Fixed setup chunk loading of <code>bcbioRNADataSet</code> in differential expression <a href="http://rmarkdown.rstudio.com">RMarkdown</a> template.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-25" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-25" class="anchor"></a>bcbioRNASeq 0.0.25</h1>
<ul>
<li>Added S4 methods support for plots, allowing the user to use either <code>bcbioRNADataSet</code> or a metrics <code>data.frame</code> and manual <code>interesting_group</code> declaration for visualization.</li>
<li>Migrated function and variable names from <code>snake_case</code> to <code>camelCase</code>.</li>
<li>Offloaded small RNA functionality to a separate package named <a href="https://github.com/lpantano/bcbioSmallRna">bcbioSmallRNA</a>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-24" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-24" class="anchor"></a>bcbioRNASeq 0.0.24</h1>
<ul>
<li>Reworked <a href="http://rmarkdown.rstudio.com">RMarkdown</a> templates to improve YAML defaults and add more comments.</li>
<li>Modified default path variables in <code>setup.R</code> to use <code>*_dir</code> instead of <code>*_out</code>.</li>
<li>Updated NEWS file to use <a href="https://daringfireball.net/projects/markdown/syntax">Markdown</a> syntax.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-23" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-23" class="anchor"></a>bcbioRNASeq 0.0.23</h1>
<ul>
<li>Slotted <code>DESeqDataSet</code> using <code>design = formula(~1)</code> for quality control. This enables automatic generation of <code>rlog</code> and <code>vst</code> transformed counts.</li>
<li>Documentation fixes and website updates.</li>
<li>Renamed S4 class from <code>bcbioRnaDataSet</code> to <code>bcbioRNADataSet</code> (case sensitive).</li>
<li>Adjusted the number of exported functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-22" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-22" class="anchor"></a>bcbioRNASeq 0.0.22</h1>
<ul>
<li>Added <a href="https://github.com/hadley/testthat">testthat</a> checking with <a href="https://github.com/jimhester/lintr">lintr</a>.</li>
<li>Initial setup of code coverage using <a href="https://github.com/jimhester/covr">covr</a>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-21" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-21" class="anchor"></a>bcbioRNASeq 0.0.21</h1>
<ul>
<li>Prepared draft of <a href="https://f1000.com">F1000</a> workflow document.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-20" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-20" class="anchor"></a>bcbioRNASeq 0.0.20</h1>
<ul>
<li>Added <a href="https://travis-ci.org">Travis-CI</a> support for automatic rendering of quality control report.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-19" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-19" class="anchor"></a>bcbioRNASeq 0.0.19</h1>
<ul>
<li>
<code>bcbioRnaDataSet</code> S4 definition updates.</li>
<li>Updates to <code>plot_pca()</code> and gene-level heatmaps.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-18" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-18" class="anchor"></a>bcbioRNASeq 0.0.18</h1>
<ul>
<li>Simplified count pooling functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-17" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-17" class="anchor"></a>bcbioRNASeq 0.0.17</h1>
<ul>
<li>Reduced number of exports and improved documentation.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-16" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-16" class="anchor"></a>bcbioRNASeq 0.0.16</h1>
<ul>
<li>Draft migration of <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> run object into S4 <code>bcbioRnaDataSet</code>.</li>
<li>Created a new variant of <code>load_run()</code> that saves to S4 object instead of list.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-15" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-15" class="anchor"></a>bcbioRNASeq 0.0.15</h1>
<ul>
<li>Reworked and re-organized internal functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-14" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-14" class="anchor"></a>bcbioRNASeq 0.0.14</h1>
<ul>
<li>Defaulted to loading run using project summary YAML file.</li>
<li>Initial commit of <a href="http://rmarkdown.rstudio.com">RMarkdown</a> templates (e.g. quality control).</li>
<li>Added support for dynamic file downloads from <a href="http://bioinformatics.sph.harvard.edu">HBC</a> website.</li>
<li>Draft build of website using <code><a href="http://hadley.github.io/pkgdown/reference/build_site.html">pkgdown::build_site()</a></code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-13" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-13" class="anchor"></a>bcbioRNASeq 0.0.13</h1>
<ul>
<li>Improved <a href="https://bioconductor.org/packages/release/bioc/html/RDAVIDWebService.html">RDAVIDWebService</a> utility functions to work with <a href="http://dplyr.tidyverse.org">dplyr</a> 0.6.0.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-12" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-12" class="anchor"></a>bcbioRNASeq 0.0.12</h1>
<ul>
<li>Reworked metadata and summary metrics functions to obtain information from <code>project-summary.yaml</code> saved in the final run directory.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-11" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-11" class="anchor"></a>bcbioRNASeq 0.0.11</h1>
<ul>
<li>Reduced number of depdencies.</li>
<li>Initial commit of modified volcano plot from <a href="https://github.com/hbc/CHBUtils">CHBUtils</a> package.</li>
<li>Internal code updates for upcoming <a href="http://dplyr.tidyverse.org">dplyr</a> 0.6.0/<a href="http://dplyr.tidyverse.org/articles/programming.html">tidyeval</a> update.</li>
<li>Updated <a href="http://www.ensembl.org">Ensembl</a> <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> annotations to use live site, currently release 88.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-10" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-10" class="anchor"></a>bcbioRNASeq 0.0.10</h1>
<ul>
<li>Renamed <code>import_*</code> functions to <code>read_*</code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-9" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-9" class="anchor"></a>bcbioRNASeq 0.0.9</h1>
<ul>
<li>Consolidated NAMESPACE imports.</li>
<li>Defaulted to writing count matrices with gzip compression, to save disk space.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-8" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-8" class="anchor"></a>bcbioRNASeq 0.0.8</h1>
<ul>
<li>Renamed internal parameters for better readability.</li>
<li>Improved documentation and consolidate functions by group.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-7" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-7" class="anchor"></a>bcbioRNASeq 0.0.7</h1>
<ul>
<li>NAMESPACE simplification using <a href="http://steinbaugh.com/basejump">basejump</a> package.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-6" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-6" class="anchor"></a>bcbioRNASeq 0.0.6</h1>
<ul>
<li>Reworked handling of plots and tables during knits.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-5" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-5" class="anchor"></a>bcbioRNASeq 0.0.5</h1>
<ul>
<li>Initial commit of differential expression and gene set enrichment functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-4" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-4" class="anchor"></a>bcbioRNASeq 0.0.4</h1>
<ul>
<li>Added <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> object integrity checks.</li>
<li>Improved detection and handling of lane split samples.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-3" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-3" class="anchor"></a>bcbioRNASeq 0.0.3</h1>
<ul>
<li>Reworked functions to utilize <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> list object.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-2" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-2" class="anchor"></a>bcbioRNASeq 0.0.2</h1>
<ul>
<li>Added plotting functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-0-1" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiornaseq-0-0-1" class="anchor"></a>bcbioRNASeq 0.0.1</h1>
<ul>
<li>Initial draft release supporting automatic loading of <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> run data.</li>
</ul>
</div>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bcbiornaseq-0-1-1">0.1.1</a></li>
        <li><a href="#bcbiornaseq-0-1-0">0.1.0</a></li>
        <li><a href="#bcbiornaseq-0-0-28">0.0.28</a></li>
        <li><a href="#bcbiornaseq-0-0-27">0.0.27</a></li>
        <li><a href="#bcbiornaseq-0-0-26">0.0.26</a></li>
        <li><a href="#bcbiornaseq-0-0-25">0.0.25</a></li>
        <li><a href="#bcbiornaseq-0-0-24">0.0.24</a></li>
        <li><a href="#bcbiornaseq-0-0-23">0.0.23</a></li>
        <li><a href="#bcbiornaseq-0-0-22">0.0.22</a></li>
        <li><a href="#bcbiornaseq-0-0-21">0.0.21</a></li>
        <li><a href="#bcbiornaseq-0-0-20">0.0.20</a></li>
        <li><a href="#bcbiornaseq-0-0-19">0.0.19</a></li>
        <li><a href="#bcbiornaseq-0-0-18">0.0.18</a></li>
        <li><a href="#bcbiornaseq-0-0-17">0.0.17</a></li>
        <li><a href="#bcbiornaseq-0-0-16">0.0.16</a></li>
        <li><a href="#bcbiornaseq-0-0-15">0.0.15</a></li>
        <li><a href="#bcbiornaseq-0-0-14">0.0.14</a></li>
        <li><a href="#bcbiornaseq-0-0-13">0.0.13</a></li>
        <li><a href="#bcbiornaseq-0-0-12">0.0.12</a></li>
        <li><a href="#bcbiornaseq-0-0-11">0.0.11</a></li>
        <li><a href="#bcbiornaseq-0-0-10">0.0.10</a></li>
        <li><a href="#bcbiornaseq-0-0-9">0.0.9</a></li>
        <li><a href="#bcbiornaseq-0-0-8">0.0.8</a></li>
        <li><a href="#bcbiornaseq-0-0-7">0.0.7</a></li>
        <li><a href="#bcbiornaseq-0-0-6">0.0.6</a></li>
        <li><a href="#bcbiornaseq-0-0-5">0.0.5</a></li>
        <li><a href="#bcbiornaseq-0-0-4">0.0.4</a></li>
        <li><a href="#bcbiornaseq-0-0-3">0.0.3</a></li>
        <li><a href="#bcbiornaseq-0-0-2">0.0.2</a></li>
        <li><a href="#bcbiornaseq-0-0-1">0.0.1</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
