<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.1 • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Version 0.1" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.21</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small>0.1</small></h1>
      
    </div>

    <div id="bcbiornaseq-0-1-0-2017-10-23" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-0-2017-10-23" class="anchor"></a>bcbioRNASeq 0.1.0 (2017-10-23)</h2>
<ul>
<li>Updated version and author information to match the F1000 Research workflow.</li>
<li>Added an <code>f1000v1</code> branch containing the reproducible code used to generate the figures in our workflow.</li>
<li>Modified <code>plotMA</code> to support vertical or horizontal layout return. Also added an argument to remove the color legend, which is typically not that informative.</li>
<li>Added custom color palette support to the quality control functions.</li>
<li>Upgrading from <code>bcbioRNADataSet</code> (&lt; 0.1.0) to <code>bcbioRNASeq</code> class object is now possible using <code>as</code> coercion method.</li>
<li>Object oriented methods are now restricted to use <code>bcbioRNASeq</code> object. Legacy <code>bcbioRNADataSet</code> objects must be upgraded to <code>bcbioRNASeq</code> class.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-1-2017-10-26" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-1-2017-10-26" class="anchor"></a>bcbioRNASeq 0.1.1 (2017-10-26)</h2>
<ul>
<li>Added support for coloring of multiple interesting groups in quality control plots.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-2-2017-11-08" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-2-2017-11-08" class="anchor"></a>bcbioRNASeq 0.1.2 (2017-11-08)</h2>
<ul>
<li>Updated package imports to match Bioconductor 3.6.</li>
<li>Added support for interesting groups assignment with <code>interestingGroups&lt;-</code>.</li>
<li>Renamed <code>plotGeneHeatmap</code> to simply <code>plotHeatmap</code>.</li>
<li>Added gender marker support for <em>Homo sapiens</em>.</li>
<li>Improved support for multiple interesting groups in quality control plots. Now interestingGroups is defined as a column in the metrics data.frame that is used to specify the plot color/fill. This matches the convention in the bcbioSingleCell 0.0.22 update.</li>
<li>Sample metadata columns are now consistently set as factors.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-3-2017-12-03" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-3-2017-12-03" class="anchor"></a>bcbioRNASeq 0.1.3 (2017-12-03)</h2>
<ul>
<li>Combined examples (<code>bcb</code>, <code>dds</code>, <code>res</code>, etc.) into a single <code>examples</code> object. This helps avoid accidental use of example <code>bcb</code> in an analysis.</li>
<li>Moved ggplot imports from <code>internal-ggplot.R</code> to above each function.</li>
<li>Renamed <code>maxSamples</code> parameter in <code>loadRNASeq</code> to <code>transformationLimit</code>. If there are more samples than this limit, then the DESeq2 transformations will be skipped. In this case, <code>rlog</code> and <code>vst</code> will not be slotted into <code>assays</code>.</li>
<li>Added a colData sanitization step in <code>loadRNASeq</code> to ensure rows are in the same order as the columns in the counts matrix. Otherwise, DESeq will report an error at the <code>DESeqDataSetFromTximport</code> step. We’re also ensuring the factor levels get updated here.</li>
<li>Now using <code>glimpse</code> instead of <code>str</code> in examples, where applicable.</li>
<li>Added <code>colData&lt;-</code> assignment method support. This requires a <code>DataFrame</code> class object. Upon assignment, the internal colData at <code>bcbio(object,   "DESeqDataSet")</code>, <code>assays(object)[["rlog"]]</code> and <code>assays(object)[["vst"]]</code> are also updated to match.</li>
<li>Initial assignment support for <code>design</code>, which will update the internal DESeqDataSet.</li>
<li>Added method support for <code>gene2symbol</code> generic, which will now return a 2 column <code>data.frame</code> with <code>ensgene</code> and <code>symbol</code> columns. This is helpful for downstream gene to symbol mapping operations.</li>
<li>Added working example for <code>interestingGroups&lt;-</code> in the documentation.</li>
<li>Added some code to improve factor releveling, where applicable. See <code>internal-meltLog10.R</code> for example.</li>
<li>Now explicitly defining the custom color palettes (e.g. <code><a href="https://rdrr.io/pkg/viridis/man/scale_viridis.html">viridis::scale_fill_viridis(discrete = TRUE)</a></code>. This makes it clearer to the user in the documentation where these palettes are located.</li>
<li>Improved axis label support in <code>plotGene</code>.</li>
<li>
<code>plotHeatmap</code> now uses internal <code>gene2symbol</code> mappings from stashed annotable, instead of always querying Ensembl. The user can define custom mappings with the <code>gene2symbol</code> argument, if desired.</li>
<li>
<code>plotPCA</code> now supports custom color palettes. The <code>shapes</code> parameter has been removed because it doesn’t work well and is limited to datasets with few samples. This behavior matches the PCA functionality in DESeq2.</li>
<li>Improved internal code for <code>plotVolcano</code>. Added support for <code>gene2symbol</code> argument, like in <code>plotHeatmap</code>. If left missing, the function will query Ensembl for the gene2symbol mappings. We’re now using <code>data</code> instead of <code>stats</code> as the main data source.</li>
<li>Improved legibility of subset method code.</li>
<li>Added some additional reexports, which are used for the package documentation and website.</li>
<li>Simplified legacy object coercion method code.</li>
<li>Updated Bioconductor installation method code. We’re now using the <code>dependencies</code> argument, which allows for automatic install of suggested packages along with imports.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-4-2018-11-27" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-4-2018-11-27" class="anchor"></a>bcbioRNASeq 0.1.4 (2018-11-27)</h2>
<ul>
<li>Migrated all basejump function imports to bcbioBase package.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-5-2018-01-31" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-5-2018-01-31" class="anchor"></a>bcbioRNASeq 0.1.5 (2018-01-31)</h2>
<ul>
<li>Import shared dependency functions from bcbioBase instead of basejump.</li>
<li>Added method support for <code>selectSamples</code>.</li>
<li>
<code>organism</code> and <code>genomeBuild</code> parameters are now user-definable in the main <code>loadRNASeq</code> import function.</li>
<li>Fixed gene subsetting method on S4 object, which handles genes using <code>intersect</code> in the featureCounts matrix.</li>
<li>Removed internal <code>aggregateReplicates</code> code. This needs to be reworked and added back in a future release.</li>
<li>Improve method for handling a missing normalized counts matrix in the assays slot. This can occur when the user opts to skip the CPU-intensive DESeq2 normalizations.</li>
<li>Improved internal code for the quality control functions. Improved the <code>if</code> statements to be more class specific.</li>
<li>Renamed <code>plotCorrelationHeatmap</code> <code>transform</code> argument to <code>normalized</code>, for consistency with the <code>counts</code> generic.</li>
<li>Added <code>title</code> support to plots, where applicable.</li>
<li>Updated internal code for <code>plotDEGHeatmap</code>.</li>
<li>Updated internal marker handling code for <code>plotGenderMarkers</code>.</li>
<li>
<code>resulsTables</code> function now defaults to <code>summary = TRUE</code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-6-2018-02-20" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-6-2018-02-20" class="anchor"></a>bcbioRNASeq 0.1.6 (2018-02-20)</h2>
<ul>
<li>Bug fix for gene-to-symbol mappings in <code>plotDEGHeatmap</code>.</li>
<li>Added support for quickly plotting differentially expressed genes (DEG) in a PCA plot with <code>plotDEGPCA</code>.</li>
<li>Added support for Dropbox shared links to <code>resultsTables</code>, for use with the <a href="http://stemcellcommons.org/">Stem Cell Commons</a> database.</li>
<li>Added assert checks internally for all functions.</li>
<li>Improved internal code for <code>plotGene</code> and <code>plotGenderMarkers</code> to run faster.</li>
<li>Deprecated data frame methods based on metrics for QC functions.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-7-2018-02-28" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-7-2018-02-28" class="anchor"></a>bcbioRNASeq 0.1.7 (2018-02-28)</h2>
<ul>
<li>Bug fix for <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> subset method dropping metrics in metadata.</li>
<li>Simplified unit testing for Dropbox mode enabled in <code>resultsTables</code>.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-1-8-2018-04-03" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-1-8-2018-04-03" class="anchor"></a>bcbioRNASeq 0.1.8 (2018-04-03)</h2>
<ul>
<li>Bug fix for <code>gene2symbol</code> argument not renaming rows in <code>plotDEGHeatmap</code>.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bcbiornaseq-0-1-0-2017-10-23">0.1.0</a></li>
        <li><a href="#bcbiornaseq-0-1-1-2017-10-26">0.1.1</a></li>
        <li><a href="#bcbiornaseq-0-1-2-2017-11-08">0.1.2</a></li>
        <li><a href="#bcbiornaseq-0-1-3-2017-12-03">0.1.3</a></li>
        <li><a href="#bcbiornaseq-0-1-4-2018-11-27">0.1.4</a></li>
        <li><a href="#bcbiornaseq-0-1-5-2018-01-31">0.1.5</a></li>
        <li><a href="#bcbiornaseq-0-1-6-2018-02-20">0.1.6</a></li>
        <li><a href="#bcbiornaseq-0-1-7-2018-02-28">0.1.7</a></li>
        <li><a href="#bcbiornaseq-0-1-8-2018-04-03">0.1.8</a></li>
      </ul>
    </div>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io/'>Lorena Pantano</a>, <a href='https://github.com/roryk/'>Rory Kirchner</a>, <a href='https://github.com/vbarrera/'>Victor Barrera</a>, <a href='https://github.com/chapmanb/'>Brad Chapman</a>, <a href='https://github.com/marypiper/'>Mary Piper</a>, <a href='https://github.com/mistrm82/'>Meeta Mistry</a>, <a href='https://github.com/rkhetani/'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford/'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann/'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson/'>John Hutchinson</a>, <a href='https://github.com/sjhosui/'>Shannan Ho Sui</a>, <a href='https://bioinformatics.sph.harvard.edu/'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


