---
title: "Small RNA-seq quality control"
author: ""
date: "`r BiocStyle::doc_date()`"
bibliography: bcbioRnaseq.bib
---

```{r setup, message=FALSE}
library(bcbioRnaseq)
library(isomiRs)
run <- load_run(
    "bcbio_yaml_file_in_final_folder",
    groups_of_interest = c("genotype", "treatment"),
    analysis = "srnaseq")

# [load_run()] automatically downloads knit defaults to `setup.R`.
# Source this file in each setup chunk to maintain report consistency.
source("setup.R")  # modify parameters before this line

final_path <- run$final_dir
res_folder <- file.path(final_path, "client")
dir.create(res_folder, recursive = TRUE, showWarnings = FALSE)

# colData metadata
metadata <- colData(run)

# Modify this to get your own formula, it should be a column in your metadata
formula <- formula(~ group)

# Turn this true to make DE ananlysis
isde <- FALSE
```

```{r header, child="header.Rmd"}
```



# Exploratory analysis

In this section we will see descriptive figures about quality of the data, 
reads with adapter, reads mapped to miRNAs, reads mapped to other small RNAs. 


## Size distribution

After adapter removal, we can plot the size distribution of the small RNAs.

```{r adapter}
plot_size_distribution(run)
```


## miRNA

### Total miRNA expression annotated with mirbase

```{r plot_mirna_counts, fig.height=9}
plot_mirna_counts(run)
```

```{r plot_srna_clusters}
plot_srna_clusters(run)
```

### isomiRs

```{r isomiRs, eval=ismirbase}
ids <- bcbio(run, "isomirs")
isoPlot(ids, type = "all", column = "group")
```



```{r footer, child="footer.Rmd"}
```
