---
title: "Differential expression"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bcbioRnaseq.bib
params:
    bcb: "data/bcb.rda"
    design: !r formula(~genotype)
    contrast: !r genotype = c("genotype", "mutant", "wildtype")
    alpha: 0.05
    lfc: 0
    data_dir: "data"
    counts_dir: "results/counts"
    de_dir: "results/differential_expression"
---

```{r setup, include=FALSE}
bcbioRnaseq::downloads()
source("setup.R")
library(DESeq2)

# bcbioRNADataSet
# Need to save bcb using `load_run()` prior to knit!
bcb_name <- load(params$bcb)
assign("bcb", get(bcb_name))
if (is.null(bcb)) stop("bcbioRNADataSet missing")

# Design formula
# help("design", "DESeq2")
design <- params$design

# Generate DESeqDataSet
dds <- DESeqDataSetFromTximport(
    txi = txi(bcb),
    colData = colData(bcb),
    design = design) %>%
    DESeq
save_data(dds, dir = data_dir)

# Contrast
# help("results", "DESeq2")
#   1. Design matrix parameter.
#   2. Numerator for LFC (expt).
#   3. Denominator for LFC (control).
contrast <- params$contrast

# Significance cutoffs
alpha <- params$alpha
lfc <- params$lfc

# Directory paths
data_dir <- params$data_dir
counts_dir <- params$counts_dir
de_dir <- params$de_dir
```

```{r header, child="header.Rmd"}
```



# Alpha level (FDR) cutoffs

Let's take a look at the number of genes we get with different false discovery rate (FDR) cutoffs. These tests subset *P* values that have been multiple test corrected using the Benjamini Hochberg (BH) method [@Benjamini:1995ws].

```{r alpha_summary, results="asis"}
alpha_summary(dds)
```



# Results

```{r res}
res_unshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
res_shrunken <- lfcShrink(
    dds = dds,
    contrast = contrast,
    res = res_unshrunken)

# Use shrunken LFC values by default
res <- res_shrunken
save_data(res, res_unshrunken, res_shrunken, dir = data_dir)
```

We performed the analysis using a BH adjusted *P* value cutoff of `r alpha` and a log fold-change (LFC) ratio cutoff of `r lfc`.



# Plots

## Mean average (MA)

An MA plot compares transformed counts on `M` (log ratio) and `A` (mean average) scales [@Yang:2002ty].

```{r plot_ma}
plot_ma(res, ylim = 3)
```


## Volcano

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log2) [@Cui:2003kh; @Li:2014fv]. Genes in the green box with text labels have an adjusted *P* value are likely to be the top candidate genes of interest.

```{r plot_volcano}
plot_volcano(bcb, res = res, lfc = lfc)
```


## Heatmap

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@WardJr:1963eu].

```{r plot_deg_heatmap}
plot_deg_heatmap(bcb, res = res, lfc = lfc)
```



# File downloads

The results are saved as gzip-compressed comma separated values (CSV). Gzip compression is natively supported on [macOS][] and Linux-based operating systems. If you're running Windows, we recommend installing [7-Zip][]. CSV files can be opened in [Excel][] or [RStudio][].


## Count matrices

- [`normalized_counts.csv.gz`](`r file.path(counts_dir, "normalized_counts.csv.gz")`): Use to evaluate individual genes and/or generate plots. These counts are normalized for the variation in sequencing depth across samples.
- [`tpm.csv.gz`](`r file.path(counts_dir, "tpm.csv.gz")`): Transcripts per million, scaled by length and also suitable for plotting.
- [`raw_counts.csv.gz`](`r file.path(counts_dir, "raw_counts.csv.gz")`): Only use to perform a new differential expression analysis. These counts will vary across samples due to differences in sequencing depth, and have not been normalized. Do not use this file for plotting genes.


## Differentially expressed genes (DEG)

DEG tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensembl_gene_id`: [Ensembl][] gene identifier.
- `base_mean`: Mean of the normalized counts per gene for all samples.
- `log2_fold_change`: log2 fold change.
- `lfc_se`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `external_gene_name`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `gene_biotype`: [Ensembl][] biotype (e.g. `protein_coding`).

```{r results_tables, results="asis"}
res_tbl <- results_tables(bcb, res = res, lfc = lfc, dir = de_dir)
save_data(res_tbl, dir = data_dir)
```



# Top tables

Only the top up- and down-regulated genes (arranged by log2 fold change) are shown.

```{r top_tables, results="asis"}
top_tables(res_tbl)
```



```{r footer, child="footer.Rmd"}
```
