<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>bcbio RNA-Seq • bcbioRNASeq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="bcbio RNA-Seq">
<meta property="og:description" content="Toolkit for bcbio RNA-seq analysis.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
<div id="bcbiornaseq" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bcbiornaseq" class="anchor"></a>bcbioRNASeq</h1></div>

<p><a href="https://www.r-project.org">R</a> package for <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> RNA-seq analysis.</p>
<div id="workflow-paper" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-paper" class="anchor"></a>Workflow paper</h2>
<p>Steinbaugh MJ, Pantano L, Kirchner RD, Barrera V, Chapman BA, Piper ME, Mistry M, Khetani RS, Rutherford KD, Hoffman O, Hutchinson JN, Ho Sui SJ. (2018). <a href="https://f1000research.com/articles/6-1976/v2">bcbioRNASeq: R package for bcbio RNA-seq analysis.</a> <em>F1000Research</em> 6:1976.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/citation">citation</a></span>(<span class="st">"bcbioRNASeq"</span>)</code></pre>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="bioconductor-method" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor-method" class="anchor"></a><a href="https://bioconductor.org">Bioconductor</a> method</h3>
<p>We recommend installing the package with <a href="https://cran.r-project.org/package=BiocManager">BiocManager</a>.</p>
<pre class="sourceCode r"><code class="sourceCode r">if (!<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"BiocManager"</span>)) {
    <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>)
}
BiocManager::<span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"remotes"</span>)
BiocManager::<span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"hbc/bcbioRNASeq"</span>)</code></pre>
<p>For <a href="https://www.r-project.org">R</a> &lt; 3.5, <a href="https://cran.r-project.org/package=BiocManager">BiocManager</a> is not supported. Use <code><a href="https://www.rdocumentation.org/packages/BiocInstaller/topics/biocLite">BiocInstaller::biocLite()</a></code> instead of <code><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">BiocManager::install()</a></code>. This requires sourcing the legacy <a href="https://bioconductor.org">Bioconductor</a> <code>biocLite.R</code> script.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># try http:// if https:// URLs are not supported</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/source">source</a></span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)</code></pre>
</div>
<div id="conda-method" class="section level3">
<h3 class="hasAnchor">
<a href="#conda-method" class="anchor"></a><a href="https://conda.io">conda</a> method</h3>
<p>Configure <a href="https://conda.io">conda</a> to use the <a href="https://bioconda.github.io">bioconda</a> channels.</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">conda</span> config --add channels defaults
<span class="kw">conda</span> config --add channels bioconda
<span class="kw">conda</span> config --add channels conda-forge</code></pre>
<p>To avoid version issues, your <code>.condarc</code> file should only contain the following channels, in this order:</p>
<pre><code>channels:
  - conda-forge
  - bioconda
  - defaults</code></pre>
<p>We recommend installing into a clean <a href="https://conda.io">conda</a> environment:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">conda</span> create --name r
<span class="kw">conda</span> activate r</code></pre>
<p>Launch <a href="https://www.r-project.org">R</a> and check that it is set up correctly with the <code><a href="https://www.rdocumentation.org/packages/base/topics/capabilities">capabilities()</a></code> function. Note that <code>X11 = TRUE</code> is required for graphical output, and requires X11 forwarding over SSH.</p>
<p>Now you’re ready to install <code>r-bcbiornaseq</code>.</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">conda</span> install -c bioconda r-bcbiornaseq</code></pre>
<p>Note that there is currently a bug with <a href="https://conda.io">conda</a> and <code>libgfortran</code>. You may need to install <code>libgfortran-ng</code> to get the bcbioRNASeq package to load in <a href="https://www.r-project.org">R</a>.</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">conda</span> install libgfortran-ng</code></pre>
</div>
</div>
<div id="load-bcbio-rna-seq-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-bcbio-rna-seq-data" class="anchor"></a>Load <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> RNA-seq data</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(bcbioRNASeq)
bcb &lt;-<span class="st"> </span><span class="kw"><a href="reference/bcbioRNASeq-class.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> <span class="st">"bcbio_rnaseq_run/final"</span>,
    <span class="dt">interestingGroups =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"genotype"</span>, <span class="st">"treatment"</span>),
    <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>
)
<span class="kw"><a href="https://www.rdocumentation.org/packages/brio/topics/saveData">saveData</a></span>(bcb, <span class="dt">dir =</span> <span class="st">"."</span>)</code></pre>
<p>This will return a <code>bcbioRNASeq</code> object, which is an extension of the <a href="https://bioconductor.org">Bioconductor</a> <a href="https://doi.org/doi:10.18129/B9.bioc.SummarizedExperiment">RangedSummarizedExperiment</a> container class. Consult the <code><a href="reference/bcbioRNASeq-class.html">bcbioRNASeq()</a></code> constructor function documentation for detailed information on the supported parameters:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/help">help</a></span>(<span class="dt">topic =</span> <span class="st">"bcbioRNASeq"</span>, <span class="dt">package =</span> <span class="st">"bcbioRNASeq"</span>)</code></pre>
<div id="sample-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-metadata" class="anchor"></a>Sample metadata</h3>
<p>When loading a <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> RNA-seq run, the sample metadata will be imported automatically from the <code>project-summary.yaml</code> file in the final upload directory. If you notice any typos in your metadata after completing the run, these can be corrected by editing the YAML file. Alternatively, you can pass in a sample metadata file into <code><a href="reference/bcbioRNASeq-class.html">bcbioRNASeq()</a></code> using the <code>sampleMetadataFile</code> argument.</p>
<div id="metadata-file-example" class="section level4">
<h4 class="hasAnchor">
<a href="#metadata-file-example" class="anchor"></a>Metadata file example</h4>
<p>The samples in the <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio</a> run must map to the <code>description</code> column. The values provided in <code>description</code> must be unique. These values will be sanitized into syntactically valid names (see <code><a href="https://www.rdocumentation.org/packages/utils/topics/help">help("make.names")</a></code>), and assigned as the column names of the <code>bcbioRNASeq</code> object. The original values are stored as the <code>sampleName</code> column in <code>colData()</code>, and are used for all plotting functions.</p>
<table class="table">
<thead><tr class="header">
<th align="left">description</th>
<th align="left">genotype</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">sample1</td>
<td align="left">wildtype</td>
</tr>
<tr class="even">
<td align="left">sample2</td>
<td align="left">knockout</td>
</tr>
<tr class="odd">
<td align="left">sample3</td>
<td align="left">wildtype</td>
</tr>
<tr class="even">
<td align="left">sample4</td>
<td align="left">knockout</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="r-markdown-templates" class="section level2">
<h2 class="hasAnchor">
<a href="#r-markdown-templates" class="anchor"></a><a href="http://rmarkdown.rstudio.com">R Markdown</a> templates</h2>
<p>The package provides multiple <a href="http://rmarkdown.rstudio.com">R Markdown</a> templates, including quality control, differential expression using <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a>, and functional enrichment analysis. These are available in <a href="https://www.rstudio.com">RStudio</a> at <code>File</code> -&gt; <code>New File</code> -&gt; <code>R Markdown...</code> -&gt; <code>From Template</code>.</p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/e1q8fn">shared library</a> on <a href="https://paperpile.com">Paperpile</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/hbc/bcbioRNASeq">https://​github.com/​hbc/​bcbioRNASeq</a>
</li>
<li>Report a bug at <br><a href="https://github.com/hbc/bcbioRNASeq/issues/">https://​github.com/​hbc/​bcbioRNASeq/​issues/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing bcbioRNASeq</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://steinbaugh.com/">Michael Steinbaugh</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2403-2221" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://lpantano.github.io/">Lorena Pantano</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-3859-3249" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/roryk/">Rory Kirchner</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4814-5885" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/vbarrera/">Victor Barrera</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-0590-4634" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/chapmanb/">Brad Chapman</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-3026-1856" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/marypiper/">Mary Piper</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2699-3840" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/mistrm82/">Meeta Mistry</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-5830-9479" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/rkhetani/">Radhika Khetani</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2430-2970" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/KayleighRutherford/">Kayleigh Rutherford</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-9699-2363" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/ohofmann/">Oliver Hofmann</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7738-1513" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/jnhutchinson/">John Hutchinson</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7804-7576" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://github.com/sjhosui/">Shannan Ho Sui</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6191-4709" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>
<a href="https://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core</a> <br><small class="roles"> Copyright holder, funder </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/hbc/bcbioRNASeq"><embed src="https://travis-ci.org/hbc/bcbioRNASeq.svg?branch=master"></embed></a></li>
<li><a href="https://ci.appveyor.com/project/mjsteinbaugh/bcbiornaseq/branch/master"><img src="https://ci.appveyor.com/api/projects/status/s0rrc28fwr0ua2wr/branch/master?svg=true" alt="AppVeyor CI"></a></li>
<li><a href="https://codecov.io/gh/hbc/bcbioRNASeq"><img src="https://codecov.io/gh/hbc/bcbioRNASeq/branch/master/graph/badge.svg" alt="Codecov"></a></li>
<li><a href="https://www.repostatus.org/#active"><img src="https://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active - The project has reached a stable, usable state and is being actively developed."></a></li>
<li><a href="https://anaconda.org/bioconda/r-bcbiornaseq"><img src="https://anaconda.org/bioconda/r-bcbiornaseq/badges/version.svg" alt="Anaconda-Server Badge"></a></li>
</ul>
</div>
</div>
</div>

      <footer><div class="copyright">
  <p>Developed by <a href="https://steinbaugh.com/">Michael Steinbaugh</a>, <a href="https://lpantano.github.io/">Lorena Pantano</a>, <a href="https://github.com/roryk/">Rory Kirchner</a>, <a href="https://github.com/vbarrera/">Victor Barrera</a>, <a href="https://github.com/chapmanb/">Brad Chapman</a>, <a href="https://github.com/marypiper/">Mary Piper</a>, <a href="https://github.com/mistrm82/">Meeta Mistry</a>, <a href="https://github.com/rkhetani/">Radhika Khetani</a>, <a href="https://github.com/KayleighRutherford/">Kayleigh Rutherford</a>, <a href="https://github.com/ohofmann/">Oliver Hofmann</a>, <a href="https://github.com/jnhutchinson/">John Hutchinson</a>, <a href="https://github.com/sjhosui/">Shannan Ho Sui</a>, <a href="https://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
