<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>bcbioRNASeq • bcbioRNASeq</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="bcbioRNASeq">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>bcbioRNASeq</h1>
            
            <h4 class="date">23 August 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/hbc/bcbioRNASeq/blob/master/vignettes/bcbioRNASeq.Rmd"><code>vignettes/bcbioRNASeq.Rmd</code></a></small>
      <div class="hidden name"><code>bcbioRNASeq.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>RNA-seq analysis involves multiple steps from processing raw sequencing data to identifying, organizing, annotating, and reporting differentially expressed genes. bcbio is an open source, community-maintained framework providing automated and scalable RNA-seq methods for identifying gene abundance counts. We have developed bcbioRNASeq, a Bioconductor package that provides ready-to-render templates and wrapper functions to post-process bcbio output data. bcbioRNASeq automates the generation of high-level RNA-seq reports, including identification of differentially expressed genes, functional enrichment analysis and quality control analysis.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>For a high-level overview of our <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> RNA-seq analysis pipeline, including detailed explanation of the <code>bcbioRNASeq</code> S4 class definition, first consult our <a href="http://dx.doi.org/10.12688/f1000research.12093.2">workflow paper</a> published in <em>F1000 Research</em> <span class="citation">(Steinbaugh et al. 2018)</span>. This vignette is focused on more advanced usage and edge cases that a user may encounter when attempting to load a <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> dataset and perform downstream quality control analysis.</p>
<p><strong>Note:</strong> if you use <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> in published research, please include this citation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">citation</span>(<span class="st">"bcbioRNASeq"</span>)</code></pre></div>
<pre><code>## 
## To cite bcbioRNASeq in publications use:
## 
##   Steinbaugh MJ, Pantano L, Kirchner RD, Barrera V, Chapman BA,
##   Piper ME, Mistry M, Khetani RS, Rutherford KD, Hoffman O,
##   Hutchinson JN, Ho Sui SJ. (2018). bcbioRNASeq: R package for
##   bcbio RNA-seq analysis. F1000Research, 6:1976. URL
##   https://f1000research.com/articles/6-1976/v2. DOI
##   10.12688/f1000research.12093.2.
## 
## A BibTeX entry for LaTeX users is
## 
##   @Article{,
##     title = {{bcbioRNASeq}: {R} package for bcbio {RNA-seq} analysis},
##     author = {Michael J. Steinbaugh and Lorena Pantano and Rory D. Kirchner and Victor Barrera and Brad A. Chapman and Mary E. Piper and Meeta Mistry and Radhika S. Khetani and Kayleigh D. Rutherford and Oliver Hofmann and John N. Hutchinson and Shannan J. Ho Sui},
##     journal = {F1000Research},
##     volume = {6},
##     number = {1976},
##     year = {2018},
##     url = {https://f1000research.com/articles/6-1976/v2},
##     doi = {10.12688/f1000research.12093.2},
##   }</code></pre>
</div>
<div id="loading-bcbio-data" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-bcbio-data" class="anchor"></a>Loading bcbio data</h1>
<p>The <code><a href="../reference/bcbioRNASeq.html">bcbioRNASeq()</a></code> constructor function is the main interface connecting <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> output data to interactive use in <a href="https://www.r-project.org">R</a>. It is highly customizable and supports a number of options for advanced use cases. Consult the documentation available at <code>help(topic = "bcbioRNASeq", package = "bcbioRNASeq")</code> for additional details.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">formals</span>(<span class="st">"bcbioRNASeq"</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">str</span>()</code></pre></div>
<pre><code>## Dotted pair list of 16
##  $ uploadDir         : symbol 
##  $ level             : language c("genes", "transcripts")
##  $ caller            : language c("salmon", "kallisto", "sailfish", "star", "hisat2")
##  $ interestingGroups : chr "sampleName"
##  $ samples           : NULL
##  $ censorSamples     : NULL
##  $ sampleMetadataFile: NULL
##  $ organism          : NULL
##  $ genomeBuild       : NULL
##  $ ensemblRelease    : NULL
##  $ gffFile           : NULL
##  $ transgeneNames    : NULL
##  $ spikeNames        : NULL
##  $ vst               : logi TRUE
##  $ rlog              : logi FALSE
##  $ ...               : symbol</code></pre>
<div id="upload-directory" class="section level2">
<h2 class="hasAnchor">
<a href="#upload-directory" class="anchor"></a>Upload directory</h2>
<p>We have designed the constructor to work as simply as possible by default. The only required argument is <code>uploadDir</code>, the path to the <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> final upload directory specified with <code>upload:</code> in the YAML configuration. Refer to the <a href="https://bcbio-nextgen.readthedocs.io/en/latest/contents/configuration.html">bcbio configuration</a> documentation for detailed information on how to set up a <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> run, which is outside the scope of this vignette.</p>
<p>For example, let’s load up the example <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> dataset stored internally in the package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">upload_dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/bcbio"</span>, <span class="dt">package =</span> <span class="st">"bcbioRNASeq"</span>)
<span class="kw">stopifnot</span>(<span class="kw">file.exists</span>(upload_dir))
<span class="kw">print</span>(upload_dir)</code></pre></div>
<pre><code>## [1] "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"</code></pre>
<p><a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> outputs RNA-seq data in a standardized directory structure, which is described in detail in our <a href="http://dx.doi.org/10.12688/f1000research.12093.2">workflow paper</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dir</span>(<span class="dt">path =</span> upload_dir, <span class="dt">full.names =</span> <span class="ot">FALSE</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>##  [1] "2017-05-23_rnaseq/bcbio-nextgen-commands.log"
##  [2] "2017-05-23_rnaseq/bcbio-nextgen.log"         
##  [3] "2017-05-23_rnaseq/combined.counts"           
##  [4] "2017-05-23_rnaseq/data_versions.csv"         
##  [5] "2017-05-23_rnaseq/programs.txt"              
##  [6] "2017-05-23_rnaseq/project-summary.yaml"      
##  [7] "2017-05-23_rnaseq/tx2gene.csv"               
##  [8] "group1_1/salmon/quant.sf"                    
##  [9] "group1_2/salmon/quant.sf"                    
## [10] "group2_1/salmon/quant.sf"                    
## [11] "group2_2/salmon/quant.sf"                    
## [12] "sample_metadata.csv"</code></pre>
</div>
<div id="counts-level" class="section level2">
<h2 class="hasAnchor">
<a href="#counts-level" class="anchor"></a>Counts level</h2>
<p>By default, <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> imports counts at gene level, which are required for standard differential expression analysis (<code>level = "genes"</code>). For pseudo-aligned counts (e.g. <a href="https://combine-lab.github.io/salmon">Salmon</a>, <a href="https://pachterlab.github.io/kallisto">Kallisto</a>, <a href="http://www.cs.cmu.edu/~ckingsf/software/sailfish">Sailfish</a>) <span class="citation">(Bray et al. 2016; Patro, Mount, and Kingsford 2014; Patro et al. 2017)</span>, <a href="https://doi.org/doi:10.18129/B9.bioc.tximport">tximport</a> <span class="citation">(Soneson, Love, and Robinson 2016)</span> is used internally to aggregate transcript-level counts to gene-level counts, and generates length-scaled transcripts per million (TPM) values. For aligned counts processed with <a href="http://bioinf.wehi.edu.au/featureCounts">featureCounts</a> <span class="citation">(Liao, Smyth, and Shi 2014)</span> (e.g. <a href="https://github.com/alexdobin/STAR">STAR</a>, <a href="https://ccb.jhu.edu/software/hisat2">HISAT2</a>) <span class="citation">(Dobin et al. 2013; Dobin and Gingeras 2016; Kim, Langmead, and Salzberg 2015)</span>, these values are already returned at gene level, and therefore not handled by <a href="https://doi.org/doi:10.18129/B9.bioc.tximport">tximport</a>. Once the gene-level counts are imported during the <code><a href="../reference/bcbioRNASeq.html">bcbioRNASeq()</a></code> call, the <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a> package <span class="citation">(Love, Huber, and Anders 2014)</span> is then used to generate an internal <code>DESeqDataSet</code> from which we derive normalized and variance-stabilized counts.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> upload_dir, <span class="dt">level =</span> <span class="st">"genes"</span>)
<span class="kw">print</span>(bcb)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 502 4 
## metadata(28): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(5): counts tpm length normalized vst
## rownames(502): ENSMUSG00000002459 ENSMUSG00000004768 ...
##   ENSMUSG00000104523 ENSMUSG00000105982
## rowData names(0):
## colnames(4): group1_1 group1_2 group2_1 group2_2
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "genes"
## Caller: "salmon"
## Organism: character(0)
## Interesting Groups: "sampleName"</code></pre>
<p>Alternatively, if you want to perform transcript-aware analysis, such as differential exon usage or splicing analysis, transcript-level counts can be obtained using <code>level = "transcripts"</code>. Note that when counts are loaded at transcript level, TPMs are generated with <a href="https://doi.org/doi:10.18129/B9.bioc.tximport">tximport</a> internally, but no additional normalizations or transformations normally calculated for gene-level counts with <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a> are generated.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> upload_dir, <span class="dt">level =</span> <span class="st">"transcripts"</span>)
<span class="kw">print</span>(bcb)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 996 4 
## metadata(28): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(3): counts tpm length
## rownames(996): ENSMUST00000193812 ENSMUST00000082908 ...
##   ENSMUST00000192656 ENSMUST00000193983
## rowData names(0):
## colnames(4): group1_1 group1_2 group2_1 group2_2
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "transcripts"
## Caller: "salmon"
## Organism: character(0)
## Interesting Groups: "sampleName"</code></pre>
</div>
<div id="expression-callers" class="section level2">
<h2 class="hasAnchor">
<a href="#expression-callers" class="anchor"></a>Expression callers</h2>
<p>Since <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> is flexible and supports a number of expression callers, we have provided advanced options in the <code><a href="../reference/bcbioRNASeq.html">bcbioRNASeq()</a></code> constructor to support a variety of workflows using the <code>caller</code> argument.</p>
<p><a href="https://combine-lab.github.io/salmon">Salmon</a>, <a href="https://pachterlab.github.io/kallisto">Kallisto</a>, and <a href="http://www.cs.cmu.edu/~ckingsf/software/sailfish">Sailfish</a> counts are supported at either gene or transcript level. Internally, these are loaded using <a href="https://doi.org/doi:10.18129/B9.bioc.tximport">tximport</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">salmon_genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> upload_dir, <span class="dt">caller =</span> <span class="st">"salmon"</span>)
<span class="kw">print</span>(salmon_genes)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 502 4 
## metadata(28): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(5): counts tpm length normalized vst
## rownames(502): ENSMUSG00000002459 ENSMUSG00000004768 ...
##   ENSMUSG00000104523 ENSMUSG00000105982
## rowData names(0):
## colnames(4): group1_1 group1_2 group2_1 group2_2
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "genes"
## Caller: "salmon"
## Organism: character(0)
## Interesting Groups: "sampleName"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assayNames</span>(salmon_genes)</code></pre></div>
<pre><code>## [1] "counts"     "tpm"        "length"     "normalized" "vst"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">salmon_tx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> upload_dir, <span class="dt">level =</span> <span class="st">"transcripts"</span>)
<span class="kw">print</span>(salmon_tx)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 996 4 
## metadata(28): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(3): counts tpm length
## rownames(996): ENSMUST00000193812 ENSMUST00000082908 ...
##   ENSMUST00000192656 ENSMUST00000193983
## rowData names(0):
## colnames(4): group1_1 group1_2 group2_1 group2_2
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "transcripts"
## Caller: "salmon"
## Organism: character(0)
## Interesting Groups: "sampleName"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assayNames</span>(salmon_tx)</code></pre></div>
<pre><code>## [1] "counts" "tpm"    "length"</code></pre>
<p><a href="https://github.com/alexdobin/STAR">STAR</a> and <a href="https://ccb.jhu.edu/software/hisat2">HISAT2</a> aligned counts processed with <a href="http://bioinf.wehi.edu.au/featureCounts">featureCounts</a> are also supported, but only at gene level.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">star &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> upload_dir, <span class="dt">caller =</span> <span class="st">"star"</span>)
<span class="kw">print</span>(star)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 999 4 
## metadata(27): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(3): counts normalized vst
## rownames(999): ENSMUSG00000102693 ENSMUSG00000064842 ...
##   ENSMUSG00000040865 ENSMUSG00000104991
## rowData names(0):
## colnames(4): group1_1 group1_2 group2_1 group2_2
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "genes"
## Caller: "star"
## Organism: character(0)
## Interesting Groups: "sampleName"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assayNames</span>(star)</code></pre></div>
<pre><code>## [1] "counts"     "normalized" "vst"</code></pre>
</div>
<div id="sample-selection-and-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-selection-and-metadata" class="anchor"></a>Sample selection and metadata</h2>
<p>If you’d like to load up only a subset of samples, this can be done easily using the <code>samples</code> argument. Note that the <code>character</code> vector declared here must match the <code>description</code> column specified in the sample metadata. For example, let’s create a <code>bcbioRNASeq</code> object containing only the <code>group1_1</code> and <code>group1_2</code> samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> upload_dir,
    <span class="dt">samples =</span> <span class="kw">c</span>(<span class="st">"group1_1"</span>, <span class="st">"group1_2"</span>)
)
<span class="kw">print</span>(bcb)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 502 2 
## metadata(28): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(5): counts tpm length normalized vst
## rownames(502): ENSMUSG00000002459 ENSMUSG00000004768 ...
##   ENSMUSG00000104523 ENSMUSG00000105982
## rowData names(0):
## colnames(2): group1_1 group1_2
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "genes"
## Caller: "salmon"
## Organism: character(0)
## Interesting Groups: "sampleName"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sampleNames</span>(bcb)</code></pre></div>
<pre><code>##   group1_1   group1_2 
## "group1_1" "group1_2"</code></pre>
<p>Conversely, if you’re working with a large dataset and you simply want to drop a few samples, this can be accomplished with the <code>censorSamples</code> argument. Note that the <code>samples</code> argument takes priority if both are declared. For example, let’s drop the <code>group2_2</code> sample from our minimal dataset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> upload_dir,
    <span class="dt">censorSamples =</span> <span class="st">"group2_2"</span>
)
<span class="kw">print</span>(bcb)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 502 3 
## metadata(28): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(5): counts tpm length normalized vst
## rownames(502): ENSMUSG00000002459 ENSMUSG00000004768 ...
##   ENSMUSG00000104523 ENSMUSG00000105982
## rowData names(0):
## colnames(3): group1_1 group1_2 group2_1
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "genes"
## Caller: "salmon"
## Organism: character(0)
## Interesting Groups: "sampleName"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sampleNames</span>(bcb)</code></pre></div>
<pre><code>##   group1_1   group1_2   group2_1 
## "group1_1" "group1_2" "group2_1"</code></pre>
<p>If you’re working with a bcbio run that has incorrect or outdated metadata, the simplest way to fix this issue is to pass in new metadata from an external spreadsheet (CSV or Excel) using the <code>sampleMetadataFile</code> argument. Note that this can also be used to subset the bcbio dataset, similar to the <code>samples</code> argument (see above), based on the rows that are included in the spreadsheet.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample_metadata_file &lt;-<span class="st"> </span><span class="kw">file.path</span>(upload_dir, <span class="st">"sample_metadata.csv"</span>)
<span class="kw">stopifnot</span>(<span class="kw">file.exists</span>(sample_metadata_file))
<span class="kw">readSampleData</span>(sample_metadata_file)</code></pre></div>
<pre><code>## Reading sample_metadata.csv</code></pre>
<pre><code>## Demultiplexed samples detected</code></pre>
<pre><code>##          sampleName             fileName description group
## group1_1   group1_1 group1_1_R1.fastq.gz    group1_1  ctrl
## group1_2   group1_2 group1_2_R1.fastq.gz    group1_2  ctrl
## group2_1   group2_1 group2_1_R1.fastq.gz    group2_1    ko
## group2_2   group2_2 group2_2_R1.fastq.gz    group2_2    ko</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> upload_dir,
    <span class="dt">sampleMetadataFile =</span> sample_metadata_file
)
<span class="kw">print</span>(bcb)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 502 4 
## metadata(28): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(5): counts tpm length normalized vst
## rownames(502): ENSMUSG00000002459 ENSMUSG00000004768 ...
##   ENSMUSG00000104523 ENSMUSG00000105982
## rowData names(0):
## colnames(4): group1_1 group1_2 group2_1 group2_2
## colData names(22): sampleName fileName ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "genes"
## Caller: "salmon"
## Organism: character(0)
## Interesting Groups: "sampleName"
## Metadata File: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio/sample_metadata.csv"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sampleData.html">sampleData</a></span>(bcb, <span class="dt">clean =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##          sampleName    group
##            &lt;factor&gt; &lt;factor&gt;
## group1_1   group1_1     ctrl
## group1_2   group1_2     ctrl
## group2_1   group2_1       ko
## group2_2   group2_2       ko</code></pre>
</div>
<div id="genome-annotations" class="section level2">
<h2 class="hasAnchor">
<a href="#genome-annotations" class="anchor"></a>Genome annotations</h2>
<p>When analyzing a dataset against a well-annotated genome, we recommend importing the corresponding metadata using <a href="https://doi.org/doi:10.18129/B9.bioc.AnnotationHub">AnnotationHub</a> and <a href="https://doi.org/doi:10.18129/B9.bioc.ensembldb">ensembldb</a>. This functionality is natively supported in the <code><a href="../reference/bcbioRNASeq.html">bcbioRNASeq()</a></code> constructor with using the <code>organism</code>, <code>ensemblRelease</code>, and <code>genomeBuild</code> arguments. This will return a <code>GRanges</code> object using the <a href="https://doi.org/doi:10.18129/B9.bioc.GenomicRanges">GenomicRanges</a> package <span class="citation">(Lawrence et al. 2013)</span>, which contains coordinates and rich metadata for each gene or transcript. These annotations are accessible with the <code>rowRanges()</code> and <code>rowData()</code> functions defined in the <a href="https://doi.org/doi:10.18129/B9.bioc.SummarizedExperiment">SummarizedExperiment</a> package <span class="citation">(Huber et al. 2015)</span>.</p>
<p>For example, with our internal <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> dataset, we’re analyzing counts generated against the <a href="https://www.ensembl.org">Ensembl</a> <em>Mus musculus</em> GRCm38 genome build (release 87). These parameters can be defined in the object load call to ensure that the annotations match up exactly with the genome used.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> upload_dir,
    <span class="dt">level =</span> <span class="st">"genes"</span>,
    <span class="dt">organism =</span> <span class="st">"Mus musculus"</span>,
    <span class="dt">genomeBuild =</span> <span class="st">"GRCm38"</span>,
    <span class="dt">ensemblRelease =</span> <span class="dv">87</span>
)
<span class="kw">print</span>(bcb)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 502 4 
## metadata(29): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(5): counts tpm length normalized vst
## rownames(502): ENSMUSG00000002459 ENSMUSG00000004768 ...
##   ENSMUSG00000104523 ENSMUSG00000105982
## rowData names(7): broadClass description ... geneName
##   seqCoordSystem
## colnames(4): group1_1 group1_2 group2_1 group2_2
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "genes"
## Caller: "salmon"
## Organism: "Mus musculus"
## Interesting Groups: "sampleName"
## AnnotationHub: "AH53222"
## Ensembl Release: "87"
## Genome Build: "GRCm38"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">metadata</span>(bcb)<span class="op">$</span>rowRangesMetadata</code></pre></div>
<pre><code>## # A tibble: 13 x 2
##    name               value                              
##    &lt;chr&gt;              &lt;chr&gt;                              
##  1 id                 AH53222                            
##  2 Db type            EnsDb                              
##  3 Type of Gene ID    Ensembl Gene ID                    
##  4 Supporting package ensembldb                          
##  5 Db created by      ensembldb package from Bioconductor
##  6 script_version     0.3.1                              
##  7 Creation time      Fri Jun  9 08:40:26 2017           
##  8 ensembl_version    87                                 
##  9 ensembl_host       localhost                          
## 10 Organism           mus_musculus                       
## 11 taxonomy_id        10090                              
## 12 genome_build       GRCm38                             
## 13 DBSCHEMAVERSION    2.1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rowRanges</span>(bcb) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">glimpse</span>()</code></pre></div>
<pre><code>## Observations: 502
## Variables: 12
## $ seqnames       &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,...
## $ start          &lt;int&gt; 4909576, 33719887, 13139105, 12692277, 18115191...
## $ end            &lt;int&gt; 5070285, 33742564, 13374083, 12861192, 18145902...
## $ width          &lt;int&gt; 160710, 22678, 234979, 168916, 30712, 39300, 18...
## $ strand         &lt;fct&gt; -, +, -, +, -, +, +, +, -, -, +, +, +, -, +, -,...
## $ broadClass     &lt;fct&gt; coding, coding, coding, coding, coding, coding,...
## $ description    &lt;fct&gt; regulator of G-protein signaling 20 [Source:MGI...
## $ entrezID       &lt;I(list)&gt;  58175,  19335,  17978, 240725,  78081,  83...
## $ geneBiotype    &lt;fct&gt; protein_coding, protein_coding, protein_coding,...
## $ geneID         &lt;chr&gt; "ENSMUSG00000002459", "ENSMUSG00000004768", "EN...
## $ geneName       &lt;fct&gt; Rgs20, Rab23, Ncoa2, Sulf1, Crisp4, Crispld1, G...
## $ seqCoordSystem &lt;fct&gt; chromosome, chromosome, chromosome, chromosome,...</code></pre>
<p>Alternatively, transcript-level annotations can also be obtained automatically using this method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> upload_dir,
    <span class="dt">level =</span> <span class="st">"transcripts"</span>,
    <span class="dt">organism =</span> <span class="st">"Mus musculus"</span>,
    <span class="dt">genomeBuild =</span> <span class="st">"GRCm38"</span>,
    <span class="dt">ensemblRelease =</span> <span class="dv">87</span>
)
<span class="kw">print</span>(bcb)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 996 4 
## metadata(29): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(3): counts tpm length
## rownames(996): ENSMUST00000193812 ENSMUST00000082908 ...
##   ENSMUST00000192656 ENSMUST00000193983
## rowData names(13): broadClass description ... transcriptName
##   transcriptSupportLevel
## colnames(4): group1_1 group1_2 group2_1 group2_2
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "transcripts"
## Caller: "salmon"
## Organism: "Mus musculus"
## Interesting Groups: "sampleName"
## AnnotationHub: "AH53222"
## Ensembl Release: "87"
## Genome Build: "GRCm38"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rowRanges</span>(bcb) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">glimpse</span>()</code></pre></div>
<pre><code>## Observations: 996
## Variables: 18
## $ seqnames               &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
## $ start                  &lt;int&gt; 3073253, 3102016, 3205901, 3206523, 321...
## $ end                    &lt;int&gt; 3074322, 3102125, 3216344, 3215632, 367...
## $ width                  &lt;int&gt; 1070, 110, 10444, 9110, 457017, 480, 28...
## $ strand                 &lt;fct&gt; +, +, -, -, -, +, -, -, -, +, -, +, -, ...
## $ broadClass             &lt;fct&gt; other, small, other, other, coding, pse...
## $ description            &lt;fct&gt; RIKEN cDNA 4933401J01 gene [Source:MGI ...
## $ entrezID               &lt;I(list)&gt;     NA,     NA, 497097, 497097, 497...
## $ geneBiotype            &lt;fct&gt; TEC, snRNA, protein_coding, protein_cod...
## $ geneID                 &lt;fct&gt; ENSMUSG00000102693, ENSMUSG00000064842,...
## $ geneName               &lt;fct&gt; 4933401J01Rik, Gm26206, Xkr4, Xkr4, Xkr...
## $ seqCoordSystem         &lt;fct&gt; chromosome, chromosome, chromosome, chr...
## $ transcriptBiotype      &lt;fct&gt; TEC, snRNA, processed_transcript, proce...
## $ transcriptCdsSeqEnd    &lt;int&gt; NA, NA, NA, NA, 3671348, NA, NA, NA, NA...
## $ transcriptCdsSeqStart  &lt;int&gt; NA, NA, NA, NA, 3216022, NA, NA, NA, NA...
## $ transcriptID           &lt;chr&gt; "ENSMUST00000193812", "ENSMUST000000829...
## $ transcriptName         &lt;chr&gt; "ENSMUST00000193812", "ENSMUST000000829...
## $ transcriptSupportLevel &lt;int&gt; NA, NA, 1, 1, 1, NA, NA, NA, NA, 3, NA,...</code></pre>
<p>When working with a dataset generated against a poorly-annotated or non-standard genome, we provide a fallback method for loading gene annotations from a <a href="https://useast.ensembl.org/info/website/upload/gff.html">general feature format (GFF)</a> file with the <code>gffFile</code> argument. If possible, we recommend providing a general transfer format (GTF) file, which is identical to GFF version 2. GFFv3 is more complicated and non-standard, but <a href="https://www.ensembl.org">Ensembl</a> GFFv3 files are also supported.</p>
<p>If your dataset contains transgenes (e.g. EGFP, TDTOMATO) or spike-ins (e.g. ERCCs), these features can be defined with the <code>transgeneNames</code> and <code>spikeNames</code> arguments, which will automatically populate the <code>rowRanges()</code> slot with placeholder metadata.</p>
<p>We recommend loading up data per genome in its own <code>bcbioRNASeq</code> object when possible, so that rich metadata can be imported easily. In the edge case where you need to look at multiple genomes simultaneously, set <code>organism = NULL</code>, and bcbioRNASeq will skip the gene annotation acquisition step.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> upload_dir, <span class="dt">organism =</span> <span class="ot">NULL</span>)
<span class="kw">print</span>(bcb)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.6
## class: RangedSummarizedExperiment 
## dim: 502 4 
## metadata(28): version level ... utilsSessionInfo
##   devtoolsSessionInfo
## assays(5): counts tpm length normalized vst
## rownames(502): ENSMUSG00000002459 ENSMUSG00000004768 ...
##   ENSMUSG00000104523 ENSMUSG00000105982
## rowData names(0):
## colnames(4): group1_1 group1_2 group2_1 group2_2
## colData names(21): sampleName description ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/Users/mike/R/library/3.5-bioc-release/library/bcbioRNASeq/extdata/bcbio"
## Upload Date: 2017-05-23
## R Load Date: 2018-08-23
## Level: "genes"
## Caller: "salmon"
## Organism: character(0)
## Interesting Groups: "sampleName"</code></pre>
</div>
<div id="variance-stabilization" class="section level2">
<h2 class="hasAnchor">
<a href="#variance-stabilization" class="anchor"></a>Variance stabilization</h2>
<p>During the <code><a href="../reference/bcbioRNASeq.html">bcbioRNASeq()</a></code> constructor call, variance stabilizaton of gene-level counts can be calculated automatically, and is recommended. This is performed internally by the <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a> package. These transformations will be slotted into <code>assays()</code> as <code>vst</code> and/or <code>rlog</code> matrices. The <code>vst</code> matrix is calculated by <code><a href="http://www.rdocumentation.org/packages/DESeq2/topics/varianceStabilizingTransformation">DESeq2::varianceStabilizingTransformation()</a></code> and <code>rlog</code> by <code><a href="http://www.rdocumentation.org/packages/DESeq2/topics/rlog">DESeq2::rlog()</a></code>. For large datasets, <code>rlog</code> calculation can take a long time, so we are currently recommending calculation of only <code>vst</code> by default.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> upload_dir,
    <span class="dt">vst =</span> <span class="ot">TRUE</span>,
    <span class="dt">rlog =</span> <span class="ot">TRUE</span>
)
<span class="kw">assayNames</span>(bcb)</code></pre></div>
<pre><code>## [1] "counts"     "tpm"        "length"     "normalized" "vst"       
## [6] "rlog"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/counts.html">counts</a></span>(bcb, <span class="dt">normalized =</span> <span class="st">"vst"</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summary</span>()</code></pre></div>
<pre><code>##     group1_1         group1_2         group2_1         group2_2     
##  Min.   : 5.708   Min.   : 5.708   Min.   : 5.708   Min.   : 5.708  
##  1st Qu.: 5.708   1st Qu.: 5.708   1st Qu.: 5.708   1st Qu.: 5.708  
##  Median : 5.708   Median : 5.708   Median : 5.708   Median : 5.708  
##  Mean   : 6.506   Mean   : 6.506   Mean   : 6.518   Mean   : 6.504  
##  3rd Qu.: 6.296   3rd Qu.: 6.260   3rd Qu.: 6.289   3rd Qu.: 6.261  
##  Max.   :17.081   Max.   :17.331   Max.   :17.334   Max.   :17.438</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/counts.html">counts</a></span>(bcb, <span class="dt">normalized =</span> <span class="st">"rlog"</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summary</span>()</code></pre></div>
<pre><code>##     group1_1         group1_2         group2_1         group2_2     
##  Min.   :-1.726   Min.   :-1.727   Min.   :-1.728   Min.   :-1.727  
##  1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.: 0.000  
##  Median : 0.000   Median : 0.000   Median : 0.000   Median : 0.000  
##  Mean   : 1.884   Mean   : 1.887   Mean   : 1.888   Mean   : 1.884  
##  3rd Qu.: 2.985   3rd Qu.: 2.969   3rd Qu.: 2.950   3rd Qu.: 2.979  
##  Max.   :17.130   Max.   :17.322   Max.   :17.325   Max.   :17.405</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rm</span>(bcb)</code></pre></div>
</div>
</div>
<div id="use-case-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#use-case-dataset" class="anchor"></a>Use case dataset</h1>
<p>To demonstrate the functionality and configuration of the package, we have taken an experiment from the <a href="https://www.ncbi.nlm.nih.gov/geo">Gene Expression Omnibus (GEO)</a> public repository of expression data to use as an example use case. The RNA-seq data is from a study of acute kidney injury in a mouse model (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE65267">GSE65267</a>) <span class="citation">(Craciun et al. 2016)</span>. The study aims to identify differentially expressed genes in progressive kidney fibrosis and contains samples from mouse kidneys at several time points (n = 3, per time point) after folic acid treatment. From this dataset, we are using a subset of the samples for our use case: before folic acid treatment, and 1, 3, 7 days after treatment.</p>
<p>For the vignette, we are loading a pre-computed version of the example <code>bcbioRNASeq</code> object used in our <a href="http://dx.doi.org/10.12688/f1000research.12093.2">workflow paper</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">loadRemoteData</span>(<span class="st">"https://github.com/hbc/bcbioRNASeq/raw/f1000v2/data/bcb.rda"</span>)
<span class="kw">print</span>(bcb)</code></pre></div>
<pre><code>## bcbioRNASeq 0.2.4
## class: RangedSummarizedExperiment 
## dim: 51652 12 
## metadata(30): version level ... devtoolsSessionInfo subset
## assays(6): counts tpm ... rlog vst
## rownames(51652): ENSMUSG00000000001 ENSMUSG00000000003 ...
##   ENSMUSG00000114967 ENSMUSG00000114968
## rowData names(7): geneID geneName ... entrezID broadClass
## colnames(12): control_rep1 control_rep2 ... fa_day7_rep2
##   fa_day7_rep3
## colData names(26): sampleName day ... x5x3Bias xGC
## ================================================================================
## Upload Dir: "/n/data1/cores/bcbio/bcbioRNASeq/F1000v2/GSE65267-merged/final"
## Upload Date: 2018-03-18
## R Load Date: 2018-06-01
## Level: "genes"
## Caller: "salmon"
## Organism: "Mus musculus"
## Interesting Groups: "day"
## AnnotationHub: "AH57770"
## Ensembl Release: "90"
## Genome Build: "GRCm38"</code></pre>
</div>
<div id="sample-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#sample-metadata" class="anchor"></a>Sample metadata</h1>
<p>For reference, let’s take a look at the sample metadata. By comparison, using <code>colData()</code> will return all sample-level metadata, including our quality control metrics generated by bcbio. We recommend instead using <code><a href="../reference/sampleData.html">sampleData()</a></code> with the <code>clean = TRUE</code> argument in reports, which only returns <code>factor</code> columns of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sampleData.html">sampleData</a></span>(bcb, <span class="dt">clean =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## DataFrame with 12 rows and 6 columns
##                sampleName      day replicate   strain   tissue  treatment
##                  &lt;factor&gt; &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;   &lt;factor&gt;
## control_rep1 control_rep1        0         1   Balb/c   kidney    control
## control_rep2 control_rep2        0         2   Balb/c   kidney    control
## control_rep3 control_rep3        0         3   Balb/c   kidney    control
## fa_day1_rep1 fa_day1_rep1        1         1   Balb/c   kidney folic acid
## fa_day1_rep2 fa_day1_rep2        1         2   Balb/c   kidney folic acid
## ...                   ...      ...       ...      ...      ...        ...
## fa_day3_rep2 fa_day3_rep2        3         2   Balb/c   kidney folic acid
## fa_day3_rep3 fa_day3_rep3        3         3   Balb/c   kidney folic acid
## fa_day7_rep1 fa_day7_rep1        7         1   Balb/c   kidney folic acid
## fa_day7_rep2 fa_day7_rep2        7         2   Balb/c   kidney folic acid
## fa_day7_rep3 fa_day7_rep3        7         3   Balb/c   kidney folic acid</code></pre>
</div>
<div id="interesting-groups" class="section level1">
<h1 class="hasAnchor">
<a href="#interesting-groups" class="anchor"></a>Interesting groups</h1>
<p>Groups of interest to be used for coloring and sample grouping in the quality control plots can be defined in the <code>bcbioRNASeq</code> object using the <code>interestingGroups</code> argument in the <code><a href="../reference/bcbioRNASeq.html">bcbioRNASeq()</a></code> constructor call. Assignment method support with the <code>interestingGroups()</code> function is also provided, which can modify the groups of interest after the object has been created.</p>
<p>The <code>interestingGroups</code> definition defaults to <code>sampleName</code>, which is automatically generated from the <a href="https://bcbio-nextgen.readthedocs.io">bcbio</a> <code>description</code> metadata for demultiplexed bulk RNA-seq samples. In this case, the samples will be grouped and colored uniquely.</p>
<p>Interesting groups must be defined using a <code>character</code> vector and refer to column names defined in the <code>colData()</code> slot of the object. Note that the <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> packages uses lower camel case formatting for column names (e.g. “sampleName”), so the interesting groups should be defined using camel case, not snake (e.g. “sample_name”) or dotted case (e.g. “sample.name”).</p>
<p>This approach was inspired by the <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a> package, which uses the argument <code>intgroup</code> in some functions, such as <code><a href="../reference/plotPCA.html">plotPCA()</a></code> for labeling groups of interest. We took this idea and ran with it for a number of our quality control functions, which are described in detail below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">interestingGroups</span>(bcb) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"treatment"</span>, <span class="st">"day"</span>)
<span class="kw">interestingGroups</span>(bcb)</code></pre></div>
<pre><code>## [1] "treatment" "day"</code></pre>
</div>
<div id="quality-control" class="section level1">
<h1 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h1>
<p>Our <a href="http://dx.doi.org/10.12688/f1000research.12093.2">workflow paper</a> describes our quality control process in detail. In addition, our <strong>Quality Control</strong> <a href="https://rmarkdown.rstudio.com">R Markdown</a> template contains notes detailing each metric. Here were are going into more technical detail regarding how to customize the appearance of the plots. Note that all quality control plotting functions inherit the <code>interestingGroups</code> defined inside the <code>bcbioRNASeq</code> object. You can also change this dynamially for each function call using the <code>interestingGroups</code> argument.</p>
<div id="read-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#read-counts" class="anchor"></a>Read counts</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotTotalReads.html">plotTotalReads</a></span>(bcb)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotMappedReads.html">plotMappedReads</a></span>(bcb)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
</div>
<div id="mapping-rates" class="section level2">
<h2 class="hasAnchor">
<a href="#mapping-rates" class="anchor"></a>Mapping rates</h2>
<p>Note that the overall mapping rate is relatively low per sample, while the exonic mapping rate is acceptable. High quality samples should have low intronic mapping rate, and high values are indicative of sample degradation and/or contamination.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotMappingRate.html">plotMappingRate</a></span>(bcb)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotExonicMappingRate.html">plotExonicMappingRate</a></span>(bcb)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-21-2.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotIntronicMappingRate.html">plotIntronicMappingRate</a></span>(bcb)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-21-3.png" width="700"></p>
</div>
<div id="rrna-mapping-rate" class="section level2">
<h2 class="hasAnchor">
<a href="#rrna-mapping-rate" class="anchor"></a>rRNA mapping rate</h2>
<p>Note that the samples here have a high rRNA mapping rate. This can be indicative of the polyA enrichment or ribo depletion protocol not having removed all ribosomal RNA (rRNA) transcripts. This will reduce the number of biologically meaningful reads in the experiment and is best avoided.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotRRNAMappingRate.html">plotRRNAMappingRate</a></span>(bcb)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
<div id="bias" class="section level2">
<h2 class="hasAnchor">
<a href="#bias" class="anchor"></a>5’-&gt;3’ bias</h2>
<p>RNA-seq data can have specific biases at either the 5’ or 3’ end of sequenced fragments. It is common to see a small amount of bias, especially if polyA enrichment was performed, or if there is any sample degradation. If a large amount of bias is observed here, be sure to analyze the samples with a Bioanalyzer and check the RIN scores.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot5Prime3PrimeBias.html">plot5Prime3PrimeBias</a></span>(bcb)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
<div id="gene-distributions" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-distributions" class="anchor"></a>Gene distributions</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGenesDetected.html">plotGenesDetected</a></span>(bcb)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneSaturation.html">plotGeneSaturation</a></span>(bcb, <span class="dt">label =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneSaturation.html">plotGeneSaturation</a></span>(bcb, <span class="dt">label =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-25-2.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCountsPerGene.html">plotCountsPerGene</a></span>(bcb)</code></pre></div>
<pre><code>## 38142 non-zero genes detected</code></pre>
<pre><code>## Applying trimmed mean of M-values (TMM) normalization</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCountDensity.html">plotCountDensity</a></span>(
    <span class="dt">object =</span> bcb,
    <span class="dt">interestingGroups =</span> <span class="st">"sampleName"</span>,
    <span class="dt">style =</span> <span class="st">"line"</span>
)</code></pre></div>
<pre><code>## 38142 non-zero genes detected
## Applying trimmed mean of M-values (TMM) normalization</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-26-2.png" width="700"></p>
</div>
<div id="dispersion" class="section level2">
<h2 class="hasAnchor">
<a href="#dispersion" class="anchor"></a>Dispersion</h2>
<p>The following plot shows the dispersion by mean of normalized counts. We expect the dispersion to decrease as the mean of normalized counts increases.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotDispEsts.html">plotDispEsts</a></span>(bcb)</code></pre></div>
<pre><code>## Coercing bcbioRNASeq to DESeqDataSet with DESeq2 1.20.0</code></pre>
<pre><code>## converting counts to integer mode</code></pre>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<pre><code>## -- replacing outliers and refitting for 3103 genes
## -- DESeq argument 'minReplicatesForReplace' = 7 
## -- original counts are preserved in counts(dds)</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## fitting model and testing</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
</div>
<div id="variance-stabilization-1" class="section level2">
<h2 class="hasAnchor">
<a href="#variance-stabilization-1" class="anchor"></a>Variance stabilization</h2>
<p>These plots show the standard deviation of normalized counts using <code>log2()</code>, <code>varianceStabilizingTransform()</code>, <code>rlog()</code>, and <code><a href="../reference/tmm.html">tmm()</a></code> by <code>rank(mean)</code>. Note that all counts shown are log2 scale.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotMeanSD.html">plotMeanSD</a></span>(bcb)</code></pre></div>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
</div>
<div id="sample-similarity" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-similarity" class="anchor"></a>Sample similarity</h2>
<p>We can visualize sample similarity with principal component analysis (PCA) and hierarchical clustering of sample correlations. These functions support multiple normalization methods with the <code>normalized</code> argument. We recommend using <code>normalized = "vst"</code> by default.</p>
<p>Let’s plot the PCA using our <code>vst</code> counts, the current recommended default.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotPCA.html">plotPCA</a></span>(bcb, <span class="dt">normalized =</span> <span class="st">"vst"</span>, <span class="dt">label =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## Using vst counts</code></pre>
<pre><code>## Plotting PCA using 500 genes</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotPCA.html">plotPCA</a></span>(bcb, <span class="dt">normalized =</span> <span class="st">"vst"</span>, <span class="dt">label =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## Using vst counts
## Plotting PCA using 500 genes</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-29-2.png" width="700"></p>
<p>Let’s visualize the <code>rlog</code> counts for comparison. The performance is similar overall, with some slight differences, but the grouping more or less remains the same.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotPCA.html">plotPCA</a></span>(
    <span class="dt">object =</span> bcb,
    <span class="dt">normalized =</span> <span class="st">"rlog"</span>,
    <span class="dt">label =</span> <span class="ot">FALSE</span>
)</code></pre></div>
<pre><code>## Using rlog counts</code></pre>
<pre><code>## Plotting PCA using 500 genes</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotPCA.html">plotPCA</a></span>(
    <span class="dt">object =</span> bcb,
    <span class="dt">normalized =</span> <span class="st">"rlog"</span>,
    <span class="dt">label =</span> <span class="ot">TRUE</span>
)</code></pre></div>
<pre><code>## Using rlog counts
## Plotting PCA using 500 genes</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-30-2.png" width="700"></p>
<p>Generally, the <code>varianceStabilizationTransformation()</code> and <code>rlog()</code> functions provide similar performance, as seen here.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap</a></span>(
    bcb,
    <span class="dt">method =</span> <span class="st">"pearson"</span>,
    <span class="dt">normalized =</span> <span class="st">"vst"</span>,
    <span class="dt">color =</span> viridis<span class="op">::</span>viridis
)</code></pre></div>
<pre><code>## Using vst counts</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap</a></span>(
    bcb,
    <span class="dt">method =</span> <span class="st">"spearman"</span>,
    <span class="dt">normalized =</span> <span class="st">"vst"</span>,
    <span class="dt">color =</span> viridis<span class="op">::</span>plasma
)</code></pre></div>
<pre><code>## Using vst counts</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-31-2.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap</a></span>(
    <span class="dt">object =</span> bcb,
    <span class="dt">method =</span> <span class="st">"pearson"</span>,
    <span class="dt">normalized =</span> <span class="st">"rlog"</span>,
    <span class="dt">color =</span> viridis<span class="op">::</span>viridis
)</code></pre></div>
<pre><code>## Using rlog counts</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap</a></span>(
    <span class="dt">object =</span> bcb,
    <span class="dt">method =</span> <span class="st">"spearman"</span>,
    <span class="dt">normalized =</span> <span class="st">"rlog"</span>,
    <span class="dt">color =</span> viridis<span class="op">::</span>plasma
)</code></pre></div>
<pre><code>## Using rlog counts</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-32-2.png" width="700"></p>
</div>
</div>
<div id="saving-counts" class="section level1">
<h1 class="hasAnchor">
<a href="#saving-counts" class="anchor"></a>Saving counts</h1>
<p>The <code><a href="../reference/counts.html">counts()</a></code> function returns the abundance estimates generated by Salmon. Read counts for each sample in the dataset are aggregated into a matrix, in which columns correspond to samples and rows represent genes. Multiple normalized counts matrices are saved in the bcbioRNASeq object, and are accessible with <code>normalized</code> argument:</p>
<ul>
<li>
<code>FALSE</code>: Raw counts (<em>default</em>).</li>
<li>
<code>TRUE</code>: <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a> normalized counts.</li>
<li>
<code>"tpm"</code>: Transcripts per million, calculated by <a href="https://doi.org/doi:10.18129/B9.bioc.tximport">tximport</a>.</li>
<li>
<code>"tmm"</code>: Trimmed mean of M-values normalization method.</li>
<li>
<code>"vst"</code>: Variance stabilization transformation.</li>
<li>
<code>"rlog"</code>: Regularized log transformation.</li>
<li>
<code>"rle"</code>: Relative log expression transformation.</li>
</ul>
<p>We recommend saving the raw counts, normalized counts, TPMs, and variance-stabilized counts to disk both in binary R Data and CSV formats.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">raw &lt;-<span class="st"> </span><span class="kw"><a href="../reference/counts.html">counts</a></span>(bcb, <span class="dt">normalized =</span> <span class="ot">FALSE</span>)
normalized &lt;-<span class="st"> </span><span class="kw"><a href="../reference/counts.html">counts</a></span>(bcb, <span class="dt">normalized =</span> <span class="ot">TRUE</span>)
tpm &lt;-<span class="st"> </span><span class="kw"><a href="../reference/counts.html">counts</a></span>(bcb, <span class="dt">normalized =</span> <span class="st">"tpm"</span>)
vst &lt;-<span class="st"> </span><span class="kw"><a href="../reference/counts.html">counts</a></span>(bcb, <span class="dt">normalized =</span> <span class="st">"vst"</span>)
<span class="kw">saveData</span>(raw, normalized, tpm, vst, <span class="dt">dir =</span> <span class="st">"."</span>)
<span class="kw">writeCounts</span>(raw, normalized, tpm, vst, <span class="dt">dir =</span> <span class="st">"."</span>)</code></pre></div>
</div>
<div id="differential-expression" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h1>
<p><a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> integrates with <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a>, which we recommend for differential expression. To prepare our dataset for analysis, we need to coerce the <code>bcbioRNASeq</code> object to a <code>DESeqDataSet</code>. Use the <code><a href="../reference/coerce.html">as()</a></code> function or refer to the <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a> documentation for the <code>DESeqDataSet()</code> constructor.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dds &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coerce.html">as</a></span>(bcb, <span class="st">"DESeqDataSet"</span>)</code></pre></div>
<pre><code>## Coercing bcbioRNASeq to DESeqDataSet with DESeq2 1.20.0</code></pre>
<pre><code>## converting counts to integer mode</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(dds)</code></pre></div>
<pre><code>## class: DESeqDataSet 
## dim: 51652 12 
## metadata(30): version level ... devtoolsSessionInfo subset
## assays(6): counts tpm ... rlog vst
## rownames(51652): ENSMUSG00000000001 ENSMUSG00000000003 ...
##   ENSMUSG00000114967 ENSMUSG00000114968
## rowData names(7): geneID geneName ... entrezID broadClass
## colnames(12): control_rep1 control_rep2 ... fa_day7_rep2
##   fa_day7_rep3
## colData names(26): sampleName day ... x5x3Bias xGC</code></pre>
<p>Since both <code>bcbioRNASeq</code> and <code>DESeqDataSet</code> classes extend <code>RangedSummarizedExperiment</code>, internally we coerce the original <code>bcbioRNASeq</code> object to a <code>RangedSummarizedExperiment</code> and then use the <code>DESeqDataSet()</code> constructor, which requires a <code>SummarizedExperiment</code> with integer counts.</p>
<p>Here is the source code responsible for handing off the data to <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getMethod</span>(
    <span class="dt">f =</span> <span class="st">"coerce"</span>,
    <span class="kw">signature</span>(
        <span class="dt">from =</span> <span class="st">"bcbioRNASeq"</span>,
        <span class="dt">to =</span> <span class="st">"DESeqDataSet"</span>
    )
)</code></pre></div>
<pre><code>## Method Definition:
## 
## function (from, to = "DESeqDataSet", strict = TRUE) 
## {
##     validObject(from)
##     if (metadata(from)[["level"]] != "genes") {
##         stop("Gene-level counts are required")
##     }
##     message(paste("Coercing bcbioRNASeq to DESeqDataSet with DESeq2", 
##         packageVersion("DESeq2")))
##     rse &lt;- as(from, "RangedSummarizedExperiment")
##     assay(rse) &lt;- round(assay(rse), digits = 0L)
##     to &lt;- DESeqDataSet(se = rse, design = ~1L)
##     interestingGroups(to) &lt;- interestingGroups(from)
##     validObject(to)
##     to
## }
## &lt;bytecode: 0x7fad478ab0a8&gt;
## &lt;environment: namespace:bcbioRNASeq&gt;
## 
## Signatures:
##         from          to            
## target  "bcbioRNASeq" "DESeqDataSet"
## defined "bcbioRNASeq" "DESeqDataSet"</code></pre>
<p>Now you can perform differential expression analysis with <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a>. The authors of that package have provided a number of detailed, well documented references online:</p>
<ul>
<li>
<a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">Analyzing RNA-seq data with DESeq2</a>.</li>
<li>
<a href="http://dx.doi.org/10.12688/f1000research.7035.2">RNA-Seq workflow: gene-level exploratory analysis and differential expression</a>.</li>
<li>
<a href="https://f1000research.com/articles/4-1521/v2">Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences</a>.</li>
</ul>
<p>We also provide a parameterized <a href="https://rmarkdown.rstudio.com">R Markdown</a> template for standard <a href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">DESeq2</a> differential expression, supporting analysis and visualization of multiple contrasts inside a single report. In our <a href="http://dx.doi.org/10.12688/f1000research.12093.2">workflow paper</a>, we also describe an example LRT analysis, and advanced edge-case templates are available inside our <a href="https://github.com/hbc/hbcABC">hbcABC</a> package, which is a current work in progress.</p>
</div>
<div id="functional-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#functional-analysis" class="anchor"></a>Functional analysis</h1>
<p>We provide an <a href="https://rmarkdown.rstudio.com">R Markdown</a> template for <a href="http://software.broadinstitute.org/gsea">gene set enrichment analysis (GSEA)</a> and <a href="https://www.genome.jp/kegg">Kyoto Encyclopedia of Genes and Genomes (KEGG)</a> pathway enrichment analysis <span class="citation">(Kanehisa and Goto 2000; Subramanian et al. 2005)</span>, which leverages the functionality of the <a href="https://doi.org/doi:10.18129/B9.bioc.clusterProfiler">clusterProfiler</a> package <span class="citation">(Yu et al. 2012)</span>. This workflow is described in detail in our <a href="http://dx.doi.org/10.12688/f1000research.12093.2">workflow paper</a>, and is outside the scope of this advanced use vignette. For more information on functional analysis, consult the <a href="https://doi.org/doi:10.18129/B9.bioc.clusterProfiler">clusterProfiler</a> vignette, which is extremely detailed and thorough.</p>
</div>
<div id="r-session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#r-session-information" class="anchor"></a>R session information</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.5.1 (2018-07-02)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.6
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] hexbin_1.27.2               bindrcpp_0.2.2             
##  [3] bcbioRNASeq_0.2.6           DESeq2_1.20.0              
##  [5] bcbioBase_0.4.1             basejump_0.7.2             
##  [7] SummarizedExperiment_1.10.1 DelayedArray_0.6.5         
##  [9] BiocParallel_1.14.2         matrixStats_0.54.0         
## [11] Biobase_2.40.0              GenomicRanges_1.32.6       
## [13] GenomeInfoDb_1.16.0         IRanges_2.14.10            
## [15] S4Vectors_0.18.3            BiocGenerics_0.26.0        
## [17] BiocStyle_2.8.2            
## 
## loaded via a namespace (and not attached):
##   [1] utf8_1.1.4                    R.utils_2.6.0                
##   [3] tidyselect_0.2.4              RSQLite_2.1.1                
##   [5] AnnotationDbi_1.42.1          htmlwidgets_1.2              
##   [7] grid_3.5.1                    munsell_0.5.0                
##   [9] preprocessCore_1.42.0         codetools_0.2-15             
##  [11] withr_2.1.2                   colorspace_1.3-2             
##  [13] BiocInstaller_1.30.0          knitr_1.20                   
##  [15] rstudioapi_0.7                assertive.base_0.0-7         
##  [17] rdrop2_0.8.1                  labeling_0.3                 
##  [19] tximport_1.8.0                GenomeInfoDbData_1.1.0       
##  [21] mnormt_1.5-5                  bit64_0.9-7                  
##  [23] pheatmap_1.0.10               rprojroot_1.3-2              
##  [25] Matrix.utils_0.9.7            xfun_0.3                     
##  [27] R6_2.2.2                      assertive.sets_0.0-3         
##  [29] locfit_1.5-9.1                AnnotationFilter_1.4.0       
##  [31] bitops_1.0-6                  reshape_0.8.7                
##  [33] assertthat_0.2.0              promises_1.0.1               
##  [35] scales_1.0.0                  nnet_7.3-12                  
##  [37] gtable_0.2.0                  affy_1.58.0                  
##  [39] ensembldb_2.4.1               rlang_0.2.2                  
##  [41] clisymbols_1.2.0              MatrixModels_0.4-1           
##  [43] genefilter_1.62.0             GlobalOptions_0.1.0          
##  [45] splines_3.5.1                 rtracklayer_1.40.5           
##  [47] lazyeval_0.2.1                acepack_1.4.1                
##  [49] checkmate_1.8.5               yaml_2.2.0                   
##  [51] reshape2_1.4.3                GenomicFeatures_1.32.2       
##  [53] backports_1.1.2               httpuv_1.4.5                 
##  [55] Hmisc_4.1-1                   tools_3.5.1                  
##  [57] bookdown_0.7                  psych_1.8.4                  
##  [59] logging_0.7-103               ggplot2_3.0.0                
##  [61] affyio_1.50.0                 assertive.strings_0.0-3      
##  [63] RColorBrewer_1.1-2            sessioninfo_1.0.0            
##  [65] Rcpp_0.12.18                  plyr_1.8.4                   
##  [67] base64enc_0.1-3               progress_1.2.0               
##  [69] zlibbioc_1.26.0               purrr_0.2.5                  
##  [71] RCurl_1.95-4.11               prettyunits_1.0.2            
##  [73] rpart_4.1-13                  viridis_0.5.1                
##  [75] pbapply_1.3-4                 GetoptLong_0.1.7             
##  [77] cowplot_0.9.3                 haven_1.1.2                  
##  [79] grr_0.9.5                     ggrepel_0.8.0                
##  [81] cluster_2.0.7-1               fs_1.2.5                     
##  [83] magrittr_1.5                  data.table_1.11.4            
##  [85] assertive.data_0.0-1          openxlsx_4.1.0               
##  [87] SparseM_1.77                  circlize_0.4.4               
##  [89] ProtGenerics_1.12.0           hms_0.4.2                    
##  [91] mime_0.5                      evaluate_0.11                
##  [93] xtable_1.8-2                  XML_3.98-1.16                
##  [95] rio_0.5.10                    readxl_1.1.0                 
##  [97] gridExtra_2.3                 shape_1.4.4                  
##  [99] compiler_3.5.1                biomaRt_2.36.1               
## [101] tibble_1.4.2                  crayon_1.3.4                 
## [103] R.oo_1.22.0                   htmltools_0.3.6              
## [105] later_0.7.3                   Formula_1.2-3                
## [107] tidyr_0.8.1                   geneplotter_1.58.0           
## [109] DBI_1.0.0                     assertive.files_0.0-2        
## [111] ComplexHeatmap_1.18.1         MASS_7.3-50                  
## [113] assertive.numbers_0.0-2       Matrix_1.2-14                
## [115] readr_1.1.1                   cli_1.0.0                    
## [117] vsn_3.48.1                    assertive.types_0.0-3        
## [119] R.methodsS3_1.7.1             bindr_0.1.1                  
## [121] forcats_0.3.0                 pkgconfig_2.0.2              
## [123] pkgdown_1.1.0                 GenomicAlignments_1.16.0     
## [125] foreign_0.8-71                xml2_1.2.0                   
## [127] roxygen2_6.1.0                annotate_1.58.0              
## [129] XVector_0.20.0                stringr_1.3.1                
## [131] digest_0.6.16                 ConsensusClusterPlus_1.44.0  
## [133] assertive.code_0.0-1          Biostrings_2.48.0            
## [135] cellranger_1.1.0              rmarkdown_1.10               
## [137] htmlTable_1.12                edgeR_3.22.3                 
## [139] curl_3.2                      shiny_1.1.0                  
## [141] Rsamtools_1.32.3              commonmark_1.5               
## [143] quantreg_5.36                 rjson_0.2.20                 
## [145] nlme_3.1-137                  jsonlite_1.5                 
## [147] DEGreport_1.16.0              viridisLite_0.3.0            
## [149] fansi_0.3.0                   desc_1.2.0                   
## [151] limma_3.36.2                  pillar_1.3.0                 
## [153] lattice_0.20-35               Nozzle.R1_1.1-1              
## [155] httr_1.3.1                    survival_2.42-6              
## [157] interactiveDisplayBase_1.18.0 glue_1.3.0                   
## [159] zip_1.0.0                     bit_1.1-14                   
## [161] assertive.properties_0.0-4    stringi_1.2.4                
## [163] blob_1.1.1                    AnnotationHub_2.12.0         
## [165] latticeExtra_0.6-28           memoise_1.1.0                
## [167] dplyr_0.7.6</code></pre>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/e1q8fn">shared library</a> on <a href="https://paperpile.com">Paperpile</a>.</p>
<div id="refs" class="references">
<div id="ref-Bray2016-me">
<p>Bray, Nicolas L, Harold Pimentel, Páll Melsted, and Lior Pachter. 2016. “Near-Optimal Probabilistic RNA-seq Quantification.” <em>Nat. Biotechnol.</em> 34 (5): 525–27. doi:<a href="https://doi.org/10.1038/nbt.3519">10.1038/nbt.3519</a>.</p>
</div>
<div id="ref-Craciun2016-ha">
<p>Craciun, Florin L, Vanesa Bijol, Amrendra K Ajay, Poornima Rao, Ramya K Kumar, John Hutchinson, Oliver Hofmann, et al. 2016. “RNA Sequencing Identifies Novel Translational Biomarkers of Kidney Fibrosis.” <em>J. Am. Soc. Nephrol.</em> 27 (6): 1702–13. doi:<a href="https://doi.org/10.1681/ASN.2015020225">10.1681/ASN.2015020225</a>.</p>
</div>
<div id="ref-Dobin2016-qj">
<p>Dobin, Alexander, and Thomas R Gingeras. 2016. “Optimizing RNA-Seq Mapping with STAR.” <em>Methods Mol. Biol.</em> 1415: 245–62. doi:<a href="https://doi.org/10.1007/978-1-4939-3572-7%5C_13">10.1007/978-1-4939-3572-7\_13</a>.</p>
</div>
<div id="ref-Dobin2013-hr">
<p>Dobin, Alexander, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. 2013. “STAR: Ultrafast Universal RNA-seq Aligner.” <em>Bioinformatics</em> 29 (1): 15–21. doi:<a href="https://doi.org/10.1093/bioinformatics/bts635">10.1093/bioinformatics/bts635</a>.</p>
</div>
<div id="ref-Huber2015-nw">
<p>Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” <em>Nat. Methods</em> 12 (2): 115–21. doi:<a href="https://doi.org/10.1038/nmeth.3252">10.1038/nmeth.3252</a>.</p>
</div>
<div id="ref-Kanehisa2000-hk">
<p>Kanehisa, M, and S Goto. 2000. “KEGG: Kyoto Encyclopedia of Genes and Genomes.” <em>Nucleic Acids Res.</em> 28 (1): 27–30. <a href="https://www.ncbi.nlm.nih.gov/pubmed/10592173" class="uri">https://www.ncbi.nlm.nih.gov/pubmed/10592173</a>.</p>
</div>
<div id="ref-Kim2015-rr">
<p>Kim, Daehwan, Ben Langmead, and Steven L Salzberg. 2015. “HISAT: A Fast Spliced Aligner with Low Memory Requirements.” <em>Nat. Methods</em> 12 (4): 357–60. doi:<a href="https://doi.org/10.1038/nmeth.3317">10.1038/nmeth.3317</a>.</p>
</div>
<div id="ref-Lawrence2013-vz">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8): e1003118. doi:<a href="https://doi.org/10.1371/journal.pcbi.1003118">10.1371/journal.pcbi.1003118</a>.</p>
</div>
<div id="ref-Liao2014-js">
<p>Liao, Yang, Gordon K Smyth, and Wei Shi. 2014. “FeatureCounts: An Efficient General Purpose Program for Assigning Sequence Reads to Genomic Features.” <em>Bioinformatics</em> 30 (7): 923–30. doi:<a href="https://doi.org/10.1093/bioinformatics/btt656">10.1093/bioinformatics/btt656</a>.</p>
</div>
<div id="ref-Love2014-sq">
<p>Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for RNA-seq Data with DESeq2.” <em>Genome Biol.</em> 15 (12): 550. doi:<a href="https://doi.org/10.1186/s13059-014-0550-8">10.1186/s13059-014-0550-8</a>.</p>
</div>
<div id="ref-Patro2017-pi">
<p>Patro, Rob, Geet Duggal, Michael I Love, Rafael A Irizarry, and Carl Kingsford. 2017. “Salmon Provides Fast and Bias-Aware Quantification of Transcript Expression.” <em>Nat. Methods</em> 14 (4): 417–19. doi:<a href="https://doi.org/10.1038/nmeth.4197">10.1038/nmeth.4197</a>.</p>
</div>
<div id="ref-Patro2014-qz">
<p>Patro, Rob, Stephen M Mount, and Carl Kingsford. 2014. “Sailfish Enables Alignment-Free Isoform Quantification from RNA-seq Reads Using Lightweight Algorithms.” <em>Nat. Biotechnol.</em> 32 (5): 462–64. doi:<a href="https://doi.org/10.1038/nbt.2862">10.1038/nbt.2862</a>.</p>
</div>
<div id="ref-Soneson2016-tu">
<p>Soneson, Charlotte, Michael I Love, and Mark D Robinson. 2016. “Differential Analyses for RNA-seq: Transcript-Level Estimates Improve Gene-Level Inferences.” <em>F1000Res.</em> 4 (December). doi:<a href="https://doi.org/10.12688/f1000research.7563.1">10.12688/f1000research.7563.1</a>.</p>
</div>
<div id="ref-Steinbaugh2018-rc">
<p>Steinbaugh, Michael J, Lorena Pantano, Rory D Kirchner, Victor Barrera, Brad A Chapman, Mary E Piper, Meeta Mistry, et al. 2018. “bcbioRNASeq: R Package for Bcbio RNA-seq Analysis.” <em>F1000Res.</em> 6 (June). doi:<a href="https://doi.org/10.12688/f1000research.12093.2">10.12688/f1000research.12093.2</a>.</p>
</div>
<div id="ref-Subramanian2005-zk">
<p>Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, et al. 2005. “Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.” <em>Proc. Natl. Acad. Sci. U. S. A.</em> 102 (43): 15545–50. doi:<a href="https://doi.org/10.1073/pnas.0506580102">10.1073/pnas.0506580102</a>.</p>
</div>
<div id="ref-Yu2012-ae">
<p>Yu, Guangchuang, Li-Gen Wang, Yanyan Han, and Qing-Yu He. 2012. “ClusterProfiler: An R Package for Comparing Biological Themes Among Gene Clusters.” <em>OMICS</em> 16 (5): 284–87. doi:<a href="https://doi.org/10.1089/omi.2011.0118">10.1089/omi.2011.0118</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#loading-bcbio-data">Loading bcbio data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#upload-directory">Upload directory</a></li>
      <li><a href="#counts-level">Counts level</a></li>
      <li><a href="#expression-callers">Expression callers</a></li>
      <li><a href="#sample-selection-and-metadata">Sample selection and metadata</a></li>
      <li><a href="#genome-annotations">Genome annotations</a></li>
      <li><a href="#variance-stabilization">Variance stabilization</a></li>
      </ul>
</li>
      <li><a href="#use-case-dataset">Use case dataset</a></li>
      <li><a href="#sample-metadata">Sample metadata</a></li>
      <li><a href="#interesting-groups">Interesting groups</a></li>
      <li>
<a href="#quality-control">Quality control</a><ul class="nav nav-pills nav-stacked">
<li><a href="#read-counts">Read counts</a></li>
      <li><a href="#mapping-rates">Mapping rates</a></li>
      <li><a href="#rrna-mapping-rate">rRNA mapping rate</a></li>
      <li><a href="#bias">5’-&gt;3’ bias</a></li>
      <li><a href="#gene-distributions">Gene distributions</a></li>
      <li><a href="#dispersion">Dispersion</a></li>
      <li><a href="#variance-stabilization-1">Variance stabilization</a></li>
      <li><a href="#sample-similarity">Sample similarity</a></li>
      </ul>
</li>
      <li><a href="#saving-counts">Saving counts</a></li>
      <li><a href="#differential-expression">Differential expression</a></li>
      <li><a href="#functional-analysis">Functional analysis</a></li>
      <li><a href="#r-session-information">R session information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://steinbaugh.com">Michael Steinbaugh</a>, <a href="https://lpantano.github.io">Lorena Pantano</a>, <a href="https://github.com/roryk">Rory Kirchner</a>, <a href="https://github.com/vbarrera">Victor Barrera</a>, <a href="https://github.com/chapmanb">Brad Chapman</a>, <a href="https://github.com/marypiper">Mary Piper</a>, <a href="https://github.com/mistrm82">Meeta Mistry</a>, <a href="https://github.com/rkhetani">Radhika Khetani</a>, <a href="https://github.com/KayleighRutherford">Kayleigh Rutherford</a>, <a href="https://github.com/ohofmann">Oliver Hofmann</a>, <a href="https://github.com/jnhutchinson">John Hutchinson</a>, <a href="https://github.com/sjhosui">Shannan Ho Sui</a>, <a href="http://bioinformatics.sph.harvard.edu">Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
