<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Force an Object to Belong to a Class — as • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Force an Object to Belong to a Class — as" />

<meta property="og:description" content="Coerce an object to a given class." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Force an Object to Belong to a Class</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/as-methods.R'><code>R/as-methods.R</code></a></small>
    <div class="hidden name"><code>as.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Coerce an object to a given class.</p>
    
    </div>

        
    <h2 class="hasAnchor" id="bcbiornaseq-to-deseqdataset"><a class="anchor" href="#bcbiornaseq-to-deseqdataset"></a>bcbioRNASeq to DESeqDataSet</h2>

    
    <ol>
<li><p>Coerces to <code>RangedSummarizedExperiment</code>.</p></li>
<li><p>Rounds raw counts to <code>integer matrix</code>.</p></li>
<li><p>Subsets <code>colData</code> to include only clean factor columns.
See <code><a href='https://www.rdocumentation.org/packages/basejump/topics/sampleData'>sampleData()</a></code> for more details.</p></li>
<li><p>Simplifies <code>metadata</code> to include only relevant information and updates
<code>sessionInfo</code> slot.</p></li>
</ol>
    <p>Note that gene-level counts are required. Alternatively
<code><a href='https://www.rdocumentation.org/packages/tximport/topics/tximport'>tximport::summarizeToGene()</a></code> can be called to convert transcript-level
counts to gene-level. By default, we're using length-scaled TPM, so a
corresponding average transcript length matrix isn't necessary. The average
transcript length matrix is only necessary when raw counts matrix isn't
scaled during tximport call (see <code>countsFromAbundance</code> in tximport).</p>
    
    <h2 class="hasAnchor" id="bcbiornaseq-to-deseqtransform"><a class="anchor" href="#bcbiornaseq-to-deseqtransform"></a>bcbioRNASeq to DESeqTransform</h2>

    
    <ol>
<li><p>Coerces to <code>DESeqDataSet</code>.</p></li>
<li><p>Calls <code><a href='https://www.rdocumentation.org/packages/DESeq2/topics/DESeq'>DESeq2::DESeq()</a></code>.</p></li>
<li><p>Calls <code><a href='https://www.rdocumentation.org/packages/DESeq2/topics/varianceStabilizingTransformation'>DESeq2::varianceStabilizingTransformation()</a></code>.</p></li>
</ol>
    
    <h2 class="hasAnchor" id="description"><a class="anchor" href="#description"></a>Description</h2>

    
    <p><code>as(object)</code>
      returns the version of this object coerced to be the given
      <code>Class</code>.  When used in the replacement form on the left of
      an assignment, the portion of the object corresponding to
      <code>Class</code> is replaced by <code>value</code>.</p>
<p>The operation of <code>as()</code> in either form depends on the
      definition of coerce methods.  Methods are defined automatically
      when the two classes are related by inheritance; that is, when
      one of the classes is a subclass of the other.</p>
<p>Coerce methods are also predefined for basic classes (including all
      the types of vectors, functions and a few others).</p>
<p>Beyond these two sources of methods, further methods are defined
      by calls to the <code>setAs</code> function.  See that
      documentation also for details of how coerce methods work. Use
      <code>showMethods(coerce)</code> for a list of all currently defined methods, as in the
      example below.</p>
    
    <h2 class="hasAnchor" id="basic-coercion-methods"><a class="anchor" href="#basic-coercion-methods"></a>Basic Coercion Methods</h2>

    
    <p>Methods are pre-defined for coercing any object to one of the basic
  datatypes.  For example, <code>as(x, "numeric")</code> uses the existing
  <code>as.numeric</code> function.  These and all other existing methods
  can be listed as shown in the example.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Chambers, John M. (2016)
 <em>Extending R</em>,
  Chapman &amp; Hall.
(Chapters 9 and 10.)</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code><a href='https://www.rdocumentation.org/packages/tximport/topics/tximport'>tximport::tximport()</a></code></p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/DESeq2/topics/DESeqDataSet'>DESeq2::DESeqDataSetFromTximport()</a></code>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>bcb</span>)

<span class='co'>## bcbioRNASeq to DESeqDataSet ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>as</span>(<span class='no'>bcb</span>, <span class='st'>"DESeqDataSet"</span>)</div><div class='output co'>#&gt; <span class='message'>Generating DESeqDataSet with DESeq2 1.22.2.</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/names'>names</a></span>(<span class='kw pkg'>S4Vectors</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class'>mcols</a></span>(<span class='no'>x</span>))</div><div class='output co'>#&gt; [1] "broadClass"     "description"    "entrezID"       "geneBiotype"   
#&gt; [5] "geneID"         "geneName"       "seqCoordSystem"</div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/class'>class</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "DESeqDataSet"
#&gt; attr(,"package")
#&gt; [1] "DESeq2"</div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: DESeqDataSet 
#&gt; dim: 500 6 
#&gt; metadata(11): version caller ... date sessionInfo
#&gt; assays(1): counts
#&gt; rownames(500): ENSMUSG00000000088 ENSMUSG00000000326 ...
#&gt;   ENSMUSG00000114367 ENSMUSG00000114458
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(9): qualityFormat sampleName ... tissue treatment</div><div class='input'>
<span class='co'>## bcbioRNASeq to RangedSummarizedExperiment ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>as</span>(<span class='no'>bcb</span>, <span class='st'>"RangedSummarizedExperiment"</span>)
<span class='fu'>slotNames</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "rowRanges"       "colData"         "assays"          "NAMES"          
#&gt; [5] "elementMetadata" "metadata"       </div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 500 6 
#&gt; metadata(28): version level ... subset previousVersion
#&gt; assays(6): counts tpm ... vst fpkm
#&gt; rownames(500): ENSMUSG00000000088 ENSMUSG00000000326 ...
#&gt;   ENSMUSG00000114367 ENSMUSG00000114458
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(25): averageInsertSize averageReadLength ... tissue
#&gt;   treatment</div><div class='input'>
<span class='co'>## bcbioRNASeq to SummarizedExperiment ====</span>
<span class='co'>## Coerce to RangedSummarizedExperiment first.</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>as</span>(<span class='no'>bcb</span>, <span class='st'>"RangedSummarizedExperiment"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>as</span>(<span class='no'>x</span>, <span class='st'>"SummarizedExperiment"</span>)
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/class'>class</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "SummarizedExperiment"
#&gt; attr(,"package")
#&gt; [1] "SummarizedExperiment"</div><div class='input'><span class='fu'>slotNames</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "colData"         "assays"          "NAMES"           "elementMetadata"
#&gt; [5] "metadata"       </div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: SummarizedExperiment 
#&gt; dim: 500 6 
#&gt; metadata(28): version level ... subset previousVersion
#&gt; assays(6): counts tpm ... vst fpkm
#&gt; rownames(500): ENSMUSG00000000088 ENSMUSG00000000326 ...
#&gt;   ENSMUSG00000114367 ENSMUSG00000114458
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(25): averageInsertSize averageReadLength ... tissue
#&gt;   treatment</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#bcbiornaseq-to-deseqdataset">bcbioRNASeq to DESeqDataSet</a></li>

      <li><a href="#bcbiornaseq-to-deseqtransform">bcbioRNASeq to DESeqTransform</a></li>

      <li><a href="#description">Description</a></li>

      <li><a href="#basic-coercion-methods">Basic Coercion Methods</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Michael Steinbaugh</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io/'>Lorena Pantano</a>, <a href='https://github.com/roryk/'>Rory Kirchner</a>, <a href='https://github.com/vbarrera/'>Victor Barrera</a>, <a href='https://github.com/chapmanb/'>Brad Chapman</a>, <a href='https://github.com/marypiper/'>Mary Piper</a>, <a href='https://github.com/mistrm82/'>Meeta Mistry</a>, <a href='https://github.com/rkhetani/'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford/'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann/'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson/'>John Hutchinson</a>, <a href='https://github.com/sjhosui/'>Shannan Ho Sui</a>, <a href='https://bioinformatics.sph.harvard.edu/'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

