<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>bcbioRNASeq Object Class — bcbioRNASeq-class • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="bcbioRNASeq Object Class — bcbioRNASeq-class" />

<meta property="og:description" content="S4 class containing RNA-seq counts and metadata generated by
bcbio." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>bcbioRNASeq Object Class</h1>
    <small>Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/AllClasses.R'><code>R/AllClasses.R</code></a></small>
    <p class="hidden name"><code>bcbioRNASeq-class.Rd</code></p>
    </div>

    
    <p>S4 class containing RNA-seq counts and metadata generated by
<a href='https://bcbio-nextgen.readthedocs.org/'>bcbio</a>.</p>
    

        
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>bcbioRNASeq</code> is a subclass of <code>SummarizedExperiment</code> designed to store
an RNA-seq analysis. This class contains raw read counts and transcripts per
million (TPM) generated by <code>tximport()</code>.</p>
<p>The <code>metadata</code> accessor contains:</p><ul>
<li><p>Sample summary statistics (a.k.a. metrics).</p></li>
<li><p>Ensembl annotations.</p></li>
<li><p>Server run paths.</p></li>
<li><p>R local environment information, including <code>sessionInfo()</code>.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    
    <dl class='dl-horizontal'>
<dt><code>bcbio</code></dt><dd><p>SimpleList containing additional bcbio run data with dimensions
that don't match the count matrix. This is currently used to store STAR
featureCounts aligned counts.</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>SummarizedExperiment::SummarizedExperiment</a>.</p></li>
<li><p><code>.S4methods(class = "bcbioRNASeq")</code>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>uploadDir</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata/bcbio"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>)

<span class='co'># We're suppressing warnings about missing log files in our minimal</span>
<span class='co'># dataset, which end up in the Travis CI logs. This isn't recommended</span>
<span class='co'># for a typical bcbio run.</span>
<span class='no'>bcb</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='loadRNASeq.html'>loadRNASeq</a></span>(<span class='no'>uploadDir</span>)</div><div class='output co'>#&gt; <span class='message'>2017-05-23_rnaseq</span></div><div class='output co'>#&gt; <span class='message'>4 samples detected</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml</span></div><div class='output co'>#&gt; <span class='message'>Detecting organism from genome build</span></div><div class='output co'>#&gt; <span class='message'>Genome: Mus musculus (mm10)</span></div><div class='output co'>#&gt; <span class='message'>Loading Ensembl annotations from AnnotationHub</span>
#&gt; <span class='message'>2017-10-27</span></div><div class='output co'>#&gt; <span class='message'>EnsDB AH57770: Mus musculus Ensembl 90</span></div><div class='output co'>#&gt; <span class='message'>Obtaining transcript-to-gene mappings</span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  enstxp = col_character(),</span>
#&gt; <span class='message'>  ensgene = col_character()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>Reading sample metrics</span></div><div class='output co'>#&gt; <span class='message'>Reading bcbio run information</span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  genome = col_character(),</span>
#&gt; <span class='message'>  resource = col_character(),</span>
#&gt; <span class='message'>  version = col_datetime(format = "")</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>bcbio-nextgen.log.txt</span></div><div class='output co'>#&gt; <span class='message'>bcbio-nextgen-commands.log.txt</span></div><div class='output co'>#&gt; <span class='message'>Reading salmon counts using tximport</span></div><div class='output co'>#&gt; <span class='message'>1 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>2 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>3 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>4 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>summarizing abundance</span></div><div class='output co'>#&gt; <span class='message'>summarizing counts</span></div><div class='output co'>#&gt; <span class='message'>summarizing length</span></div><div class='output co'>#&gt; <span class='message'>Performing trimmed mean of M-values (TMM) normalization</span></div><div class='output co'>#&gt; <span class='message'>Generating internal DESeqDataSet</span></div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Performing rlog transformation</span></div><div class='output co'>#&gt; <span class='message'>Performing variance stabilizing transformation</span></div><div class='output co'>#&gt; <span class='message'>Reading STAR featureCounts aligned counts</span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  id = col_character(),</span>
#&gt; <span class='message'>  group1_1 = col_integer(),</span>
#&gt; <span class='message'>  group1_2 = col_integer(),</span>
#&gt; <span class='message'>  group2_1 = col_integer(),</span>
#&gt; <span class='message'>  group2_2 = col_integer()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Unannotated genes detected in assay (0.594%)</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>bcb</span>)</div><div class='output co'>#&gt; class: bcbioRNASeq 
#&gt; dim: 505 4 
#&gt; metadata(28): version uploadDir ... devtoolsSessionInfo
#&gt;   unannotatedGenes
#&gt; assays(6): raw normalized ... rlog vst
#&gt; rownames(505): ENSMUSG00000002459 ENSMUSG00000004768 ...
#&gt;   ENSMUSG00000105982 ENSMUSG00000109048
#&gt; rowData names(11): ensgene symbol ... seqCoordSystem entrez
#&gt; colnames(4): group1_1 group1_2 group2_1 group2_2
#&gt; colData names(4): sampleID sampleName description group</div><div class='input'>
<span class='co'># Check validity</span>
<span class='fu'>validObject</span>(<span class='no'>bcb</span>)</div><div class='output co'>#&gt; [1] TRUE</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#details">Details</a></li>

      <li><a href="#slots">Slots</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Lorena Pantano, Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

