<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title><code>bcbioRNASeq</code> Object and Constructor — bcbioRNASeq • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="<code>bcbioRNASeq</code> Object and Constructor — bcbioRNASeq" />

<meta property="og:description" content="bcbioRNASeq is an S4 class that extends RangedSummarizedExperiment, and
is designed to store a bcbio RNA-seq
analysis." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1><code>bcbioRNASeq</code> Object and Constructor</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/AllClasses.R'><code>R/AllClasses.R</code></a>, <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/deprecated.R'><code>R/deprecated.R</code></a></small>
    <div class="hidden name"><code>bcbioRNASeq.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p><code>bcbioRNASeq</code> is an S4 class that extends <code>RangedSummarizedExperiment</code>, and
is designed to store a <a href='https://bcbio-nextgen.readthedocs.org'>bcbio</a> RNA-seq
analysis.</p>
    
    </div>

    <pre class="usage"><span class='fu'>bcbioRNASeq</span>(<span class='no'>uploadDir</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>caller</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"salmon"</span>, <span class='st'>"kallisto"</span>, <span class='st'>"sailfish"</span>), <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>samples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>censorSamples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>, <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>spikeNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gffFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>transformationLimit</span> <span class='kw'>=</span> <span class='fl'>50L</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uploadDir</th>
      <td><p>Path to final upload directory. This path is set when
running "<code>bcbio_nextgen -w template</code>".</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p>Import counts as "<code>genes</code>" (default) or "<code>transcripts</code>".</p></td>
    </tr>
    <tr>
      <th>caller</th>
      <td><p>Expression caller. Supports "<code>salmon</code>" (default), "<code>kallisto</code>",
or "<code>sailfish</code>".</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p>Organism name. Use the full Latin name (e.g. "Homo sapiens"),
since this will be input downstream to AnnotationHub and ensembldb, unless
<code>gffFile</code> is set. If left <code>NULL</code> (<em>not recommended</em>), the function call
will skip loading gene/transcript-level annotations into <code>rowRanges()</code>.
This can be useful for poorly annotation genomes or experiments involving
multiple genomes.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p><em>Optional.</em> Custom metadata file containing
sample information. Otherwise defaults to sample metadata saved in the YAML
file. Remote URLs are supported. Typically this can be left unset.</p></td>
    </tr>
    <tr>
      <th>samples</th>
      <td><p><em>Optional.</em> Specify a subset of samples to load. The names
must match the <code>description</code> specified in the bcbio YAML metadata. If a
<code>sampleMetadataFile</code> is provided, that will take priority for sample
selection. Typically this can be left unset.</p></td>
    </tr>
    <tr>
      <th>censorSamples</th>
      <td><p><em>Optional.</em> Samples to exclude from the analysis.</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p>Character vector denoting groups of interest that
define the samples. If left unset, defaults to <code>sampleName</code>.</p></td>
    </tr>
    <tr>
      <th>ensemblRelease</th>
      <td><p><em>Optional.</em> Ensembl release version. If unset,
defaults to current release, and does not typically need to be
user-defined. Passed to AnnotationHub for <code>EnsDb</code> annotation matching,
unless <code>gffFile</code> is set.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><em>Optional.</em> Ensembl genome build name (e.g. "GRCh38").
This will be passed to AnnotationHub for <code>EnsDb</code> annotation matching,
unless <code>gffFile</code> is set.</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code> vector indicating which <code>assay()</code> rows
denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code> vector indicating which <code>assay()</code> rows denote
spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p><em>Advanced use; not recommended.</em> By default, we recommend
leaving this <code>NULL</code> for genomes that are supported on Ensembl. In this
case, the row annotations (<code>rowRanges()</code>) will be obtained automatically
from Ensembl by passing the <code>organism</code>, <code>genomeBuild</code>, and <code>ensemblRelease</code>
arguments to AnnotationHub and ensembldb. For a genome that is not
supported on Ensembl and/or AnnotationHub, a GFF/GTF (General Feature
Format) file is required. Generally, we recommend using a GTF (GFFv2) file
here over a GFF3 file if possible, although all GFF formats are supported.
The function will internally generate a <code>TxDb</code> containing
transcript-to-gene mappings and construct a <code>GRanges</code> object containing the
genomic ranges (<code>rowRanges()</code>).</p></td>
    </tr>
    <tr>
      <th>transformationLimit</th>
      <td><p>Maximum sample limit allowed to calculate
<code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/rlog'>DESeq2::rlog()</a></code> and <code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/varianceStabilizingTransformation'>DESeq2::varianceStabilizingTransformation()</a></code>.
For large datasets, DESeq2 is slow to apply variance stabilization. In this
case, we recommend loading up the dataset in a high-performance computing
environment. Use <code>Inf</code> to always apply and <code>-Inf</code> to always skip.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments, slotted into the <code>metadata()</code> accessor.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>bcbioRNASeq</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Simply point to the final upload directory generated by
<a href='https://bcbio-nextgen.readthedocs.io/'>bcbio</a>, and this constructor function
will take care of the rest. It automatically imports RNA-seq counts,
metadata, and the program versions used.</p>
<p>This class contains raw read counts and length-scaled
transcripts per million (TPM) generated by <code><a href='http://www.rdocumentation.org/packages/tximport/topics/tximport'>tximport::tximport()</a></code>. Counts can
be loaded at gene or transcript level.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><code>bcbioRNASeq</code> extended <code>SummarizedExperiment</code> prior to v0.2.0, where we
migrated to <code>RangedSummarizedExperiment</code>.</p>
    
    <h2 class="hasAnchor" id="metadata"><a class="anchor" href="#metadata"></a>Metadata</h2>

    
    <p>The <code>metadata()</code> accessor contains:</p><ul>
<li><p>Sample quality control metrics.</p></li>
<li><p>Ensembl annotations.</p></li>
<li><p>Server run paths.</p></li>
<li><p>R session information (e.g. <code><a href='http://www.rdocumentation.org/packages/utils/topics/sessionInfo'>utils::sessionInfo()</a></code>).</p></li>
</ul>
    
    <h2 class="hasAnchor" id="deseq-"><a class="anchor" href="#deseq-"></a>DESeq2</h2>

    
    <p>DESeq2 is run automatically when <code>bcbioRNASeq()</code> is called, and variance
stabilized counts are slotted into <code>assays()</code>. If the number of samples is
bigger than the <code>transformationLimit</code> argument, <code>rlog</code> and <code>vst</code> counts will
not be slotted into <code>assays()</code>. In this case, we recommend visualization
using <code>tmm()</code> counts, which are automatically calculated using edgeR.</p>
    
    <h2 class="hasAnchor" id="remote-data"><a class="anchor" href="#remote-data"></a>Remote data</h2>

    
    <p>When working in RStudio, we recommend connecting to the bcbio run directory
as a remote connection over
<a href='https://github.com/osxfuse/osxfuse/wiki/SSHFS'>sshfs</a>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>SummarizedExperiment::SummarizedExperiment()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/methods/topics/new'>methods::initialize()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/methods/topics/validObject'>methods::validObject()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/BiocGenerics/topics/updateObject'>BiocGenerics::updateObject()</a></code>.</p></li>
<li><p><code>.S4methods(class = "bcbioRNASeq")</code>.</p></li>
</ul></div>
    <div class='dont-index'><p>Other S4 Object: <code><a href='coerce.html'>coerce</a></code>,
  <code><a href='extract.html'>extract</a></code>, <code><a href='show.html'>show</a></code>,
  <code><a href='updateObject.html'>updateObject</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>uploadDir</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata/bcbio"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>)

<span class='co'># Gene level</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioRNASeq</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"genes"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"salmon"</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Mus musculus"</span>,
    <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='fl'>87L</span>
)</div><div class='output co'>#&gt; <span class='message'>2017-05-23_rnaseq</span></div><div class='output co'>#&gt; <span class='message'>4 samples detected:</span>
#&gt; <span class='message'>group1_1, group1_2, group2_1, group2_2</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml</span></div><div class='output co'>#&gt; <span class='message'>Reading sample metrics</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml</span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  txID = col_character(),</span>
#&gt; <span class='message'>  geneID = col_character()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>Reading salmon counts using tximport</span></div><div class='output co'>#&gt; <span class='message'>1 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>2 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>3 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>4 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>summarizing abundance</span></div><div class='output co'>#&gt; <span class='message'>summarizing counts</span></div><div class='output co'>#&gt; <span class='message'>summarizing length</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.12.0 (2018-04-30)</span></div><div class='output co'>#&gt; <span class='message'>Ensembl 87 EnsDb for Mus Musculus</span></div><div class='output co'>#&gt; <span class='message'>id: "AH53222"</span>
#&gt; <span class='message'>organism: "Mus musculus"</span>
#&gt; <span class='message'>build: "GRCm38"</span>
#&gt; <span class='message'>release: "87"</span>
#&gt; <span class='message'>format: "genes"</span></div><div class='output co'>#&gt; <span class='message'>Standardizing the metadata columns</span></div><div class='output co'>#&gt; <span class='message'>Using geneName instead of symbol</span></div><div class='output co'>#&gt; <span class='message'>Setting geneBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting seqCoordSystem as factor</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Generating DESeqDataSet with DESeq2 1.20.0</span></div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Applying rlog transformation</span></div><div class='output co'>#&gt; <span class='message'>Applying variance stabilizing transformation</span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.2.6
#&gt; Samples: 4
#&gt; Genes: 502
#&gt; Organism: Mus musculus
#&gt; AnnotationHub: AH53222
#&gt; Ensembl Release: 87
#&gt; Genome Build: GRCm38
#&gt; Upload Dir: /Users/mike/GitHub/hbc/bcbioRNASeq/inst/extdata/bcbio
#&gt; Upload Date: 2017-05-23
#&gt; R Load Date: 2018-07-09</div><div class='input'><span class='fu'>is</span>(<span class='no'>x</span>, <span class='st'>"RangedSummarizedExperiment"</span>)</div><div class='output co'>#&gt; [1] TRUE</div><div class='input'><span class='fu'>validObject</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] TRUE</div><div class='input'>
<span class='co'># Transcript level</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioRNASeq</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"transcripts"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"salmon"</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Mus musculus"</span>,
    <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='fl'>87L</span>
)</div><div class='output co'>#&gt; <span class='message'>2017-05-23_rnaseq</span></div><div class='output co'>#&gt; <span class='message'>4 samples detected:</span>
#&gt; <span class='message'>group1_1, group1_2, group2_1, group2_2</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml</span></div><div class='output co'>#&gt; <span class='message'>Reading sample metrics</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml</span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  txID = col_character(),</span>
#&gt; <span class='message'>  geneID = col_character()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>Reading salmon counts using tximport</span></div><div class='output co'>#&gt; <span class='message'>1 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>2 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>3 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>4 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.12.0 (2018-04-30)</span></div><div class='output co'>#&gt; <span class='message'>Ensembl 87 EnsDb for Mus Musculus</span></div><div class='output co'>#&gt; <span class='message'>id: "AH53222"</span>
#&gt; <span class='message'>organism: "Mus musculus"</span>
#&gt; <span class='message'>build: "GRCm38"</span>
#&gt; <span class='message'>release: "87"</span>
#&gt; <span class='message'>format: "transcripts"</span></div><div class='output co'>#&gt; <span class='message'>Standardizing the metadata columns</span></div><div class='output co'>#&gt; <span class='message'>Using geneName instead of symbol</span></div><div class='output co'>#&gt; <span class='message'>Setting geneBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting seqCoordSystem as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting txBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting txSupportLevel as factor</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, txBiotype, seqnames</span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.2.6
#&gt; Samples: 4
#&gt; Transcripts: 996
#&gt; Organism: Mus musculus
#&gt; AnnotationHub: AH53222
#&gt; Ensembl Release: 87
#&gt; Genome Build: GRCm38
#&gt; Upload Dir: /Users/mike/GitHub/hbc/bcbioRNASeq/inst/extdata/bcbio
#&gt; Upload Date: 2017-05-23
#&gt; R Load Date: 2018-07-09</div><div class='input'><span class='fu'>validObject</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] TRUE</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#metadata">Metadata</a></li>

      <li><a href="#deseq-">DESeq2</a></li>

      <li><a href="#remote-data">Remote data</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Lorena Pantano, Rory Kirchner, Victor Barrera

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

