<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>bcbio RNA-Seq data set — bcbioRNASeq • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="bcbio RNA-Seq data set — bcbioRNASeq" />

<meta property="og:description" content="Simply point to the final upload directory generated by bcbio,
and this generator function will take care of the rest." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>bcbio RNA-Seq data set</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/bcbioRNASeq.R'><code>R/bcbioRNASeq.R</code></a></small>
    <div class="hidden name"><code>bcbioRNASeq.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Simply point to the final upload directory generated by bcbio,
and this generator function will take care of the rest.</p>
    
    </div>

    <pre class="usage"><span class='fu'>bcbioRNASeq</span>(<span class='no'>uploadDir</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>caller</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"salmon"</span>, <span class='st'>"kallisto"</span>, <span class='st'>"sailfish"</span>, <span class='st'>"star"</span>, <span class='st'>"hisat2"</span>),
  <span class='kw'>samples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>censorSamples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gffFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>spikeNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>countsFromAbundance</span> <span class='kw'>=</span> <span class='st'>"lengthScaledTPM"</span>, <span class='kw'>vst</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>rlog</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uploadDir</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Path to final upload directory.
This path is set when running "<code>bcbio_nextgen -w template</code>".</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Import counts at gene level ("<code>genes</code>"; <em>default</em>) or transcript level
("<code>transcripts</code>"; <em>advanced use</em>). Only tximport-compatible callers (e.g.
salmon, kallisto, sailfish) can be loaded at transcript level. Aligned
counts from featureCounts-compatible callers (e.g. STAR, HISAT2) can only
be loaded at gene level.</p></td>
    </tr>
    <tr>
      <th>caller</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Expression caller:</p><pre>- "`salmon`" (*default*): [Salmon][] alignment-free, quasi-mapped counts.
- "`kallisto`": [Kallisto][] alignment-free, pseudo-aligned counts.
- "`sailfish`": [Sailfish][] alignment-free, lightweight counts.
- "`star`": [STAR][] (Spliced Transcripts Alignment to a Reference)
  aligned counts.
- "`hisat2`": [HISAT2][] (Hierarchical Indexing for Spliced Alignment of
  Transcripts) graph-based aligned counts.
</pre></td>
    </tr>
    <tr>
      <th>samples</th>
      <td><p><code>character</code>.
Specify a subset of samples to include.</p></td>
    </tr>
    <tr>
      <th>censorSamples</th>
      <td><p><code>character</code>.
Specify a subset of samples to censor.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Sample metadata file path. CSV or TSV is preferred, but Excel worksheets
are also supported. Check the documentation for conventions and required
columns.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Full Latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Ensembl genome build assembly name (e.g. <code>"GRCh38"</code>). If set <code>NULL</code>,
defaults to the most recent build available. Note: don't pass in UCSC build
IDs (e.g. <code>"hg38"</code>).</p></td>
    </tr>
    <tr>
      <th>ensemblRelease</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code>.
Ensembl release version (e.g. <code>90</code>). We recommend setting this value if
possible, for improved reproducibility. When left unset, the latest release
available via AnnotationHub/ensembldb is used. Note that the latest version
available can vary, depending on the versions of AnnotationHub and
ensembldb in use.</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
GFF/GTF (General Feature Format) file. Generally, we recommend using a GTF
(GFFv2) instead of a GFFv3 file if possible.</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code>.
Vector indicating which assay rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code>.
Vector indicating which assay rows denote spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    <tr>
      <th>countsFromAbundance</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Whether to generate estimated counts using abundance estimates
(<em>recommended by default</em>). <code>lengthScaledTPM</code> is a suitable default, and
counts are scaled using the average transcript length over samples and then
the library size. Refer to <code><a href='https://www.rdocumentation.org/packages/tximport/topics/tximport'>tximport::tximport()</a></code> for more information on
this parameter, but it should only ever be changed when loading some
datasets at transcript level (e.g. for DTU analsyis).</p></td>
    </tr>
    <tr>
      <th>vst</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Calculate variance-stabilizing transformation using
<code><a href='https://www.rdocumentation.org/packages/DESeq2/topics/varianceStabilizingTransformation'>DESeq2::varianceStabilizingTransformation()</a></code>. Recommended by default for
visualization.</p></td>
    </tr>
    <tr>
      <th>rlog</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Calcualte regularized log transformation using <code><a href='https://www.rdocumentation.org/packages/DESeq2/topics/rlog'>DESeq2::rlog()</a></code>. This
calculation is slow for large datasets and now discouraged by default for
visualization.</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code>.
Groups of interest that define the samples. If left unset, defaults to
<code>sampleName</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>bcbioRNASeq</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>bcbioRNASeq</code> generator function automatically imports RNA-seq counts,
metadata, and the program versions used from a <a href='https://bcbio-nextgen.readthedocs.io'>bcbio</a> RNA-seq run.</p>
    
    <h2 class="hasAnchor" id="sample-metadata"><a class="anchor" href="#sample-metadata"></a>Sample metadata</h2>

    
    <p>When loading a bcbio RNA-seq run, the sample metadata will be imported
automatically from the <code>project-summary.yaml</code> file in the final upload
directory. If you notice any typos in your metadata after completing the run,
these can be corrected by editing the YAML file.</p>
<p>Alternatively, you can pass in a sample metadata file into the
<code>bcbioRNASeq</code> function call using the <code>sampleMetadataFile</code> argument. This
requires either a CSV or Excel spreadsheet.</p>
<p>The samples in the bcbio run must map to the <code>description</code> column. The values
provided in <code>description</code> must be unique. These values will be sanitized into
syntactically valid names (see <code><a href='https://www.rdocumentation.org/packages/basejump/topics/reexports'>basejump::makeNames</a></code> for more information),
and assigned as the column names of the <code>bcbioRNASeq</code> object. The original
values are stored as the <code>sampleName</code> column in <code>colData</code>, and are used for
all plotting functions. Do not attempt to set a <code>sampleID</code> column, as this is
used internally by the package.</p>
<p>Here is a minimal example of a properly formatted sample metadata file:</p>
<table class='table'>
<tr><td></td><td>description</td></tr>
<tr><td>genotype</td><td></td></tr>
<tr><td>sample1</td><td>wildtype</td></tr>
<tr><td></td><td>sample2</td></tr>
<tr><td>knockout</td><td></td></tr>
<tr><td>sample3</td><td>wildtype</td></tr>
<tr><td></td><td>sample4</td></tr>
<tr><td>knockout</td><td></td></tr>
</table>
    
    <h2 class="hasAnchor" id="valid-names"><a class="anchor" href="#valid-names"></a>Valid names</h2>

    
    <p>R is strict about values that are considered valid for use in <code>names</code> and
<code>dimnames</code> (i.e. <code>rownames</code> and <code>colnames</code>). Non-alphanumeric
characters, spaces, and <strong>dashes</strong> are not valid. Use either underscores or
periods in place of dashes when working in R. Also note that names should
<strong>not begin with a number</strong>, and will be prefixed with an <code>X</code> when sanitized.
Consult the documentation in the <code>make.names</code> function for more
information. We strongly recommend adhering to these conventions when
labeling samples, to help avoid unexpected downstream behavior in R due to
<code>dimnames</code> mismatches.</p>
    
    <h2 class="hasAnchor" id="genome-annotations"><a class="anchor" href="#genome-annotations"></a>Genome annotations</h2>

    
    <p><code>bcbioRNASeq</code> provides support for automatic import of genome annotations,
which internally get processed into genomic ranges (<code>GRanges</code>) and are
slotted into the <code>rowRanges</code> of the S4 object. Currently, we offer support
for (1) <a href='https://ensembl.org/'>Ensembl</a> genome annotations from <a href='https://bioconductor.org/packages/AnnotationHub/'>AnnotationHub</a> via
<a href='https://bioconductor.org/packages/ensembldb/'>ensembldb</a> (<em>recommended</em>); or (2) direct import from a GTF/GFF file using
<a href='https://bioconductor.org/packages/rtracklayer/'>rtracklayer</a>.</p>
<p><a href='https://bioconductor.org/packages/ensembldb/'>ensembldb</a> requires the <code>organism</code> and <code>ensemblRelease</code> arguments to be
defined. When both of these are set, <code>bcbioRNASeq</code> will attempt to
download and use the pre-built <a href='https://ensembl.org/'>Ensembl</a> genome annotations from
<a href='https://bioconductor.org/packages/AnnotationHub/'>AnnotationHub</a>. This method is preferred over direct loading of a GTF/GFF
file because the <a href='https://bioconductor.org/packages/AnnotationHub/'>AnnotationHub</a> annotations contain additional rich
metadata not defined in a GFF file, specifically <code>description</code> and <code>entrezID</code>
values.</p>
<p>Alternatively, if you are working with a non-standard or poorly annotated
genome that isn't available on <a href='https://bioconductor.org/packages/AnnotationHub/'>AnnotationHub</a>, we provide fall back
support for loading the genome annotations directly from the GTF file used by
the bcbio RNA-seq pipeline. This should be fully automatic for an R session
active on the same server used to run <a href='https://bcbio-nextgen.readthedocs.io/'>bcbio</a>.</p>
<p>Example bcbio GTF path: <code>genomes/Hsapiens/hg38/rnaseq/ref-transcripts.gtf</code>.</p>
<p>In the event that you are working from a remote environment that doesn't
have file system access to the <a href='https://bcbio-nextgen.readthedocs.io/'>bcbio</a> <code>genomes</code> directory, we provide
additional fall back support for importing genome annotations from a GTF/GFF
directly with the <code>gffFile</code> argument.</p>
<p>Internally, genome annotations are imported via the <a href='https://steinbaugh.com/basejump/'>basejump</a> package,
specifically with either of these functions:</p><ul>
<li><p><code><a href='https://www.rdocumentation.org/packages/basejump/topics/makeGRanges'>basejump::makeGRangesFromEnsembl()</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/basejump/topics/makeGRanges'>basejump::makeGRangesFromGFF()</a></code>.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="genome-build"><a class="anchor" href="#genome-build"></a>Genome build</h2>

    
    <p>Ensure that the organism and genome build used with bcio match correctly here
in the function call. In particular, for the legacy <em>Homo sapiens</em>
GRCh37/hg19 genome build, ensure that <code>genomeBuild = "GRCh37"</code>. Otherwise,
the genomic ranges set in <code>rowRanges</code> will mismatch.
It is recommended for current projects that GRCh38/hg38 is used in place of
GRCh37/hg19 if possible.</p>
    
    <h2 class="hasAnchor" id="deseq-"><a class="anchor" href="#deseq-"></a>DESeq2</h2>

    
    <p>DESeq2 is run automatically when <code>bcbioRNASeq()</code> is called. Internally, this
automatically slots normalized counts into
<code><a href='https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>assays()</a></code>, and optionally generates
variance-stabilized counts (e.g. <code>vst</code>, <code>rlog</code>), depending on the call. When
loading a dataset with a large number of samples (i.e. &gt; 50), we recommend
disabling the <code>rlog</code> transformation, since it can take a long time to
compute.</p>
    
    <h2 class="hasAnchor" id="remote-connections"><a class="anchor" href="#remote-connections"></a>Remote connections</h2>

    
    <p>When working on a local machine, it is possible to load bcbio run data over a
remote connection using <a href='https://github.com/osxfuse/osxfuse/wiki/SSHFS'>sshfs</a>. When loading a large number of samples, it
is preferable to call <code>bcbioRNASeq()</code> directly in R on the remote server, if
possible.</p>
    
    <h2 class="hasAnchor" id="automatic-metadata"><a class="anchor" href="#automatic-metadata"></a>Automatic metadata</h2>

    
    <p>The <code>metadata</code> slot always contains:</p><ul>
<li><p>Object version.</p></li>
<li><p>bcbio data provenance information.</p></li>
<li><p>File paths and timestamps.</p></li>
<li><p>R session information.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code>.S4methods(class = "bcbioRNASeq")</code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>SummarizedExperiment::SummarizedExperiment()</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/methods/topics/new'>methods::initialize()</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/methods/topics/validObject'>methods::validObject()</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/BiocGenerics/topics/updateObject'>BiocGenerics::updateObject()</a></code>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>uploadDir</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>"extdata/bcbio"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>)

<span class='co'>## Gene level.</span>
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioRNASeq</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"genes"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"salmon"</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Mus musculus"</span>,
    <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='fl'>87L</span>
)</div><div class='output co'>#&gt; <span class='message'>Dated project directory: 2018-03-18_GSE65267-merged</span></div><div class='output co'>#&gt; <span class='message'>6 sample(s) detected:</span>
#&gt; <span class='message'>control_rep1, control_rep2, control_rep3, fa_day7_rep1, fa_day7_rep2, fa_day7...</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Data versions are missing.</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>bcbio-nextgen.log file is empty.</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen-commands.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>bcbio-nextgen-commands.log file is empty.</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metadata from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metrics from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Reading salmon counts using tximport 1.10.1.</span></div><div class='output co'>#&gt; <span class='message'>Reading from quant.sf files.</span></div><div class='output co'>#&gt; <span class='message'>Scaling using lengthScaledTPM.</span></div><div class='output co'>#&gt; <span class='message'>1 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = <span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>2 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>3 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>4 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>5 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>6 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>summarizing abundance</span></div><div class='output co'>#&gt; <span class='message'>summarizing counts</span></div><div class='output co'>#&gt; <span class='message'>summarizing length</span></div><div class='output co'>#&gt; <span class='message'>Using makeGRangesFromEnsembl() for annotations.</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>Matching EnsDb from AnnotationHub 2.14.2 (2018-10-24).</span></div><div class='output co'>#&gt; <span class='message'>AH53222: Ensembl 87 EnsDb for Mus Musculus</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Mus musculus</span>
#&gt; <span class='message'>  - Genome Build: GRCm38</span>
#&gt; <span class='message'>  - Ensembl Release: 87</span>
#&gt; <span class='message'>  - Level: genes</span></div><div class='output co'>#&gt; <span class='message'>Using `geneName` column instead of `symbol`.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>Generating DESeqDataSet with DESeq2 1.22.2.</span></div><div class='output co'>#&gt; <span class='message'>Calculating normalized counts.</span></div><div class='output co'>#&gt; <span class='message'>Calculating transformations.</span></div><div class='output co'>#&gt; <span class='message'>using pre-existing size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Applying variance-stabilizing transformation.</span></div><div class='output co'>#&gt; <span class='message'>Calculating FPKM.</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.3.7
#&gt; uploadDir: /Users/mike/git/bcbioRNASeq/inst/extdata/bcbio
#&gt; dates(2): [bcbio] 2018-03-18; [R] 2019-01-13
#&gt; level: genes
#&gt; caller: salmon
#&gt; organism: Mus musculus
#&gt; interestingGroups: sampleName
#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 100 6 
#&gt; metadata(26): version level ... wd sessionInfo
#&gt; assays(6): counts tpm ... vst fpkm
#&gt; rownames(100): ENSMUSG00000000001 ENSMUSG00000000003 ...
#&gt;   ENSMUSG00000062661 ENSMUSG00000074340
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(26): averageInsertSize averageReadLength ... tissue
#&gt;   treatment</div><div class='input'>
<span class='co'>## Transcript level.</span>
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioRNASeq</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"transcripts"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"salmon"</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Mus musculus"</span>,
    <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='fl'>87L</span>
)</div><div class='output co'>#&gt; <span class='message'>Dated project directory: 2018-03-18_GSE65267-merged</span></div><div class='output co'>#&gt; <span class='message'>6 sample(s) detected:</span>
#&gt; <span class='message'>control_rep1, control_rep2, control_rep3, fa_day7_rep1, fa_day7_rep2, fa_day7...</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Data versions are missing.</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>bcbio-nextgen.log file is empty.</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen-commands.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>bcbio-nextgen-commands.log file is empty.</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metadata from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metrics from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Reading salmon counts using tximport 1.10.1.</span></div><div class='output co'>#&gt; <span class='message'>Reading from quant.sf files.</span></div><div class='output co'>#&gt; <span class='message'>Scaling using lengthScaledTPM.</span></div><div class='output co'>#&gt; <span class='message'>1 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>2 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>3 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>4 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>5 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>6 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>Using makeGRangesFromEnsembl() for annotations.</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>Matching EnsDb from AnnotationHub 2.14.2 (2018-10-24).</span></div><div class='output co'>#&gt; <span class='message'>AH53222: Ensembl 87 EnsDb for Mus Musculus</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Mus musculus</span>
#&gt; <span class='message'>  - Genome Build: GRCm38</span>
#&gt; <span class='message'>  - Ensembl Release: 87</span>
#&gt; <span class='message'>  - Level: transcripts</span></div><div class='output co'>#&gt; <span class='message'>Using `geneName` column instead of `symbol`.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, transcriptBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by transcriptID.</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.3.7
#&gt; uploadDir: /Users/mike/git/bcbioRNASeq/inst/extdata/bcbio
#&gt; dates(2): [bcbio] 2018-03-18; [R] 2019-01-13
#&gt; level: transcripts
#&gt; caller: salmon
#&gt; organism: Mus musculus
#&gt; interestingGroups: sampleName
#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 100 6 
#&gt; metadata(26): version level ... wd sessionInfo
#&gt; assays(3): counts tpm avgTxLength
#&gt; rownames(100): ENSMUST00000000001 ENSMUST00000000003 ...
#&gt;   ENSMUST00000000674 ENSMUST00000000687
#&gt; rowData names(13): broadClass description ... transcriptName
#&gt;   transcriptSupportLevel
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(26): averageInsertSize averageReadLength ... tissue
#&gt;   treatment</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#sample-metadata">Sample metadata</a></li>

      <li><a href="#valid-names">Valid names</a></li>

      <li><a href="#genome-annotations">Genome annotations</a></li>

      <li><a href="#genome-build">Genome build</a></li>

      <li><a href="#deseq-">DESeq2</a></li>

      <li><a href="#remote-connections">Remote connections</a></li>

      <li><a href="#automatic-metadata">Automatic metadata</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Michael Steinbaugh, Lorena Pantano, Rory Kirchner, Victor Barrera</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io/'>Lorena Pantano</a>, <a href='https://github.com/roryk/'>Rory Kirchner</a>, <a href='https://github.com/vbarrera/'>Victor Barrera</a>, <a href='https://github.com/chapmanb/'>Brad Chapman</a>, <a href='https://github.com/marypiper/'>Mary Piper</a>, <a href='https://github.com/mistrm82/'>Meeta Mistry</a>, <a href='https://github.com/rkhetani/'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford/'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann/'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson/'>John Hutchinson</a>, <a href='https://github.com/sjhosui/'>Shannan Ho Sui</a>, <a href='https://bioinformatics.sph.harvard.edu/'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

