<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title><code>bcbioRNASeq</code> object and constructor — bcbioRNASeq • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="<code>bcbioRNASeq</code> object and constructor — bcbioRNASeq" />

<meta property="og:description" content="bcbioRNASeq is an S4 class that extends RangedSummarizedExperiment, and
is designed to store a bcbio RNA-seq
analysis." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1><code>bcbioRNASeq</code> object and constructor</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/AllClasses.R'><code>R/AllClasses.R</code></a>, <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/bcbioRNASeq.R'><code>R/bcbioRNASeq.R</code></a></small>
    <div class="hidden name"><code>bcbioRNASeq.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p><code>bcbioRNASeq</code> is an S4 class that extends <code>RangedSummarizedExperiment</code>, and
is designed to store a <a href='https://bcbio-nextgen.readthedocs.org'>bcbio</a> RNA-seq
analysis.</p>
    
    </div>

    <pre class="usage"><span class='fu'>bcbioRNASeq</span>(<span class='no'>uploadDir</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>caller</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"salmon"</span>, <span class='st'>"kallisto"</span>, <span class='st'>"sailfish"</span>, <span class='st'>"star"</span>, <span class='st'>"hisat2"</span>),
  <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>, <span class='kw'>samples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>censorSamples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>gffFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>spikeNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>vst</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>rlog</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uploadDir</th>
      <td><p><code>string</code>. Path to final upload directory. This path is set
when running "<code>bcbio_nextgen -w template</code>".</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p><code>string</code>. Import counts at gene level ("<code>genes</code>"; <em>default</em>) or
transcript level ("<code>transcripts</code>"; <em>advanced use</em>).</p></td>
    </tr>
    <tr>
      <th>caller</th>
      <td><p><code>string</code>. Expression caller:</p><ul>
<li><p>"<code>salmon</code>" (<em>default</em>): <a href='https://combine-lab.github.io/salmon'>Salmon</a>
alignment-free, quasi-mapped counts.</p></li>
<li><p>"<code>kallisto</code>". <a href='https://pachterlab.github.io/kallisto'>Kallisto</a>
alignment-free, pseudo-aligned counts.</p></li>
<li><p>"<code>sailfish</code>".
<a href='http://www.cs.cmu.edu/~ckingsf/software/sailfish'>Sailfish</a>
alignment-free, lightweight counts.</p></li>
<li><p>"<code>star</code>": <a href='https://github.com/alexdobin/STAR'>STAR</a>
(Spliced Transcripts Alignment to a Reference) aligned counts.</p></li>
<li><p>"<code>hisat2</code>": <a href='https://ccb.jhu.edu/software/hisat2'>HISAT2</a>
(Hierarchical Indexing for Spliced Alignment of Transcripts) graph-based
aligned counts.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code>. Groups of interest that define the
samples. If left unset, defaults to <code>sampleName</code>.</p></td>
    </tr>
    <tr>
      <th>samples</th>
      <td><p><code>character</code> or <code>NULL</code>. Specify a subset of samples to load.
The names must match the <code>description</code> specified in the bcbio YAML
metadata. If a <code>sampleMetadataFile</code> is provided, that will take priority
for sample selection. Typically this can be left unset.</p></td>
    </tr>
    <tr>
      <th>censorSamples</th>
      <td><p><code>character</code> or <code>NULL</code>. Samples to exclude from the
analysis.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p><code>string</code> or <code>NULL</code>. Custom metadata file containing
sample information. Otherwise defaults to sample metadata saved in the YAML
file. Remote URLs are supported. Typically this can be left unset.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code>string</code> or <code>NULL</code>. Organism name. Use the full Latin name
(e.g. "Homo sapiens"), since this will be input downstream to AnnotationHub
and ensembldb, unless <code>gffFile</code> is set. If left <code>NULL</code> (<em>not recommended</em>),
the function call will skip loading gene/transcript-level annotations into
<code>rowRanges()</code>. This can be useful for poorly annotation genomes or
experiments involving multiple genomes.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code>string</code> or <code>NULL</code>. Ensembl genome build name (e.g.
"GRCh38"). This will be passed to AnnotationHub for <code>EnsDb</code> annotation
matching, unless <code>gffFile</code> is set.</p></td>
    </tr>
    <tr>
      <th>ensemblRelease</th>
      <td><p><code>scalar integer</code> or <code>NULL</code>. Ensembl release version. If
unset, defaults to current release, and does not typically need to be
user-defined. Passed to AnnotationHub for <code>EnsDb</code> annotation matching,
unless <code>gffFile</code> is set.</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p><code>string</code> or <code>NULL</code>. By default, we recommend leaving this
<code>NULL</code> for genomes that are supported on Ensembl. In this case, the row
annotations (<code>rowRanges()</code>) will be obtained automatically from Ensembl by
passing the <code>organism</code>, <code>genomeBuild</code>, and <code>ensemblRelease</code> arguments to
AnnotationHub and ensembldb. For a genome that is not supported on Ensembl
and/or AnnotationHub, a GFF/GTF (General Feature Format) file is required.
Generally, we recommend using a GTF (GFFv2) file here over a GFF3 file if
possible, although all GFF formats are supported. The function will
internally generate a <code>TxDb</code> containing transcript-to-gene mappings and
construct a <code>GRanges</code> object containing the genomic ranges (<code>rowRanges()</code>).</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code>.
Vector indicating which assay rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code>.
Vector indicating which assay rows denote spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    <tr>
      <th>vst</th>
      <td><p><code>boolean</code>. Calculate variance-stabilizing transformation using
<code><a href='https://rdrr.io/pkg/DESeq2/man/varianceStabilizingTransformation.html'>DESeq2::varianceStabilizingTransformation()</a></code>. Recommended by default
for visualization.</p></td>
    </tr>
    <tr>
      <th>rlog</th>
      <td><p><code>boolean</code>. Calcualte regularized log transformation using
<code><a href='https://rdrr.io/pkg/DESeq2/man/rlog.html'>DESeq2::rlog()</a></code>. This calculation is slow for large datasets and now
discouraged by default for visualization.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments, slotted into the <code>metadata()</code> accessor.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>bcbioRNASeq</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Simply point to the final upload directory generated by
<a href='https://bcbio-nextgen.readthedocs.io/'>bcbio</a>, and this constructor function
will take care of the rest. It automatically imports RNA-seq counts,
metadata, and the program versions used.</p>
<p>This class contains raw read counts and length-scaled
transcripts per million (TPM) generated by <code><a href='https://rdrr.io/pkg/tximport/man/tximport.html'>tximport::tximport()</a></code>. Counts can
be loaded at gene or transcript level.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><code>bcbioRNASeq</code> extended <code>SummarizedExperiment</code> prior to v0.2.0, where we
migrated to <code>RangedSummarizedExperiment</code>.</p>
    
    <h2 class="hasAnchor" id="genome-build"><a class="anchor" href="#genome-build"></a>Genome build</h2>

    
    <p>Ensure that the organism and genome build used with bcio match correctly
here in the function call. In particular, for the legacy <em>Homo sapiens</em>
GRCh37/hg19 genome build, ensure that <code>genomeBuild = "GRCh37"</code>. Otherwise,
the genomic ranges set in <code>rowRanges()</code> will mismatch. It is recommended
for current projects that GRCh38/hg38 is used in place of GRCh37/hg19
if possible.</p>
    
    <h2 class="hasAnchor" id="metadata"><a class="anchor" href="#metadata"></a>Metadata</h2>

    
    <p>The <code>metadata()</code> slot contains:</p><ul>
<li><p>Sample quality control metrics.</p></li>
<li><p>Ensembl annotations.</p></li>
<li><p>Server run paths.</p></li>
<li><p>R session information (e.g. <code><a href='https://rdrr.io/r/utils/sessionInfo.html'>utils::sessionInfo()</a></code>).</p></li>
</ul>
    
    <h2 class="hasAnchor" id="deseq-"><a class="anchor" href="#deseq-"></a>DESeq2</h2>

    
    <p>DESeq2 is run automatically when <code>bcbioRNASeq()</code> is called, and variance
stabilized counts are slotted into <code>assays()</code>. If the number of samples is
bigger than the <code>transformationLimit</code> argument, <code>rlog</code> and <code>vst</code> counts will
not be slotted into <code>assays()</code>. In this case, we recommend visualization
using TMM counts, which are automatically calculated using edgeR.</p>
    
    <h2 class="hasAnchor" id="remote-data"><a class="anchor" href="#remote-data"></a>Remote data</h2>

    
    <p>When working in RStudio, we recommend connecting to the bcbio run directory
as a remote connection over
<a href='https://github.com/osxfuse/osxfuse/wiki/SSHFS'>sshfs</a>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>SummarizedExperiment::SummarizedExperiment()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/r/methods/new.html'>methods::initialize()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/r/methods/validObject.html'>methods::validObject()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/pkg/BiocGenerics/man/updateObject.html'>BiocGenerics::updateObject()</a></code>.</p></li>
<li><p><code>.S4methods(class = "bcbioRNASeq")</code>.</p></li>
</ul></div>
    <div class='dont-index'><p>Other S4 Object: <code><a href='coerce.html'>coerce</a></code>,
  <code><a href='extract.html'>extract</a></code>, <code><a href='show.html'>show</a></code>,
  <code><a href='updateObject.html'>updateObject</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>upload_dir</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata/bcbio"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>)

<span class='co'># Gene level</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioRNASeq</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>upload_dir</span>,
    <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"genes"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"salmon"</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Mus musculus"</span>,
    <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='fl'>87L</span>
)</div><div class='output co'>#&gt; <span class='message'>2017-05-23_rnaseq</span></div><div class='output co'>#&gt; <span class='message'>4 samples detected:</span>
#&gt; <span class='message'>[1] "group1_1" "group1_2" "group2_1" "group2_2"</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Data versions are missing.</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen-commands.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metadata from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metrics from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Reading salmon counts using tximport 1.12.0</span></div><div class='output co'>#&gt; <span class='message'>1 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = <span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>2 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>3 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>4 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>summarizing abundance</span></div><div class='output co'>#&gt; <span class='message'>summarizing counts</span></div><div class='output co'>#&gt; <span class='message'>summarizing length</span></div><div class='output co'>#&gt; <span class='message'>Using `makeGRangesFromEnsembl()` for annotations</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>Matching EnsDb from AnnotationHub 2.16.0 (2019-05-01).</span></div><div class='output co'>#&gt; <span class='message'>AH53222: Ensembl 87 EnsDb for Mus Musculus</span>
#&gt; <span class='message'>Run this code to download EnsDb manually:</span>
#&gt; <span class='message'>  &gt; library(AnnotationHub)</span>
#&gt; <span class='message'>  &gt; ah &lt;- AnnotationHub()</span>
#&gt; <span class='message'>  &gt; edb &lt;- ah[["AH53222"]]</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Mus musculus</span>
#&gt; <span class='message'>  - Genome Build: GRCm38</span>
#&gt; <span class='message'>  - Ensembl Release: 87</span>
#&gt; <span class='message'>  - Level: genes</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneBiotype, geneName, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>50143 genes detected.</span></div><div class='output co'>#&gt; <span class='message'>Generating DESeqDataSet with DESeq2 1.24.0</span></div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Applying variance-stabilizing transformation</span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.2.10
#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 502 4 
#&gt; metadata(28): version level ... wd sessionInfo
#&gt; assays(5): counts tpm length normalized vst
#&gt; rownames(502): ENSMUSG00000002459 ENSMUSG00000004768 ...
#&gt;   ENSMUSG00000104523 ENSMUSG00000105982
#&gt; rowData names(7): broadClass description ... geneName seqCoordSystem
#&gt; colnames(4): group1_1 group1_2 group2_1 group2_2
#&gt; colData names(21): sampleName description ... totalReads x5x3Bias
#&gt; Upload Dir: "/Users/mike/git/hbc/bcbioRNASeq/inst/extdata/bcbio"
#&gt; Upload Date: 2017-05-23
#&gt; R Load Date: 2019-05-13
#&gt; Level: "genes"
#&gt; Caller: "salmon"
#&gt; Organism: "Mus musculus"
#&gt; Interesting Groups: "sampleName"
#&gt; AnnotationHub: character(0)
#&gt; Ensembl Release: "87"
#&gt; Genome Build: "GRCm38"</div><div class='input'><span class='fu'>is</span>(<span class='no'>x</span>, <span class='st'>"RangedSummarizedExperiment"</span>)</div><div class='output co'>#&gt; [1] TRUE</div><div class='input'><span class='fu'>validObject</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] TRUE</div><div class='input'>
<span class='co'># Transcript level</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioRNASeq</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>upload_dir</span>,
    <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"transcripts"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"salmon"</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Mus musculus"</span>,
    <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='fl'>87L</span>
)</div><div class='output co'>#&gt; <span class='message'>2017-05-23_rnaseq</span></div><div class='output co'>#&gt; <span class='message'>4 samples detected:</span>
#&gt; <span class='message'>[1] "group1_1" "group1_2" "group2_1" "group2_2"</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Data versions are missing.</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen-commands.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metadata from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metrics from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Reading salmon counts using tximport 1.12.0</span></div><div class='output co'>#&gt; <span class='message'>1 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>2 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>3 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>4 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = </span><span style='color: #BB0000;'>col_character()</span><span>,</span>
#&gt; <span class='message'>  Length = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  EffectiveLength = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  TPM = </span><span style='color: #00BB00;'>col_double()</span><span>,</span>
#&gt; <span class='message'>  NumReads = </span><span style='color: #00BB00;'>col_double()</span><span></span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>Using `makeGRangesFromEnsembl()` for annotations</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>Matching EnsDb from AnnotationHub 2.16.0 (2019-05-01).</span></div><div class='output co'>#&gt; <span class='message'>AH53222: Ensembl 87 EnsDb for Mus Musculus</span>
#&gt; <span class='message'>Run this code to download EnsDb manually:</span>
#&gt; <span class='message'>  &gt; library(AnnotationHub)</span>
#&gt; <span class='message'>  &gt; ah &lt;- AnnotationHub()</span>
#&gt; <span class='message'>  &gt; edb &lt;- ah[["AH53222"]]</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Mus musculus</span>
#&gt; <span class='message'>  - Genome Build: GRCm38</span>
#&gt; <span class='message'>  - Ensembl Release: 87</span>
#&gt; <span class='message'>  - Level: transcripts</span></div><div class='output co'>#&gt; <span class='message'>Merging gene-level annotations: gene_name, gene_biotype, seq_coord_system, description, symbol, entrezid</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, seqnames, transcriptBiotype</span></div><div class='output co'>#&gt; <span class='message'>Arranging by transcriptID.</span></div><div class='output co'>#&gt; <span class='message'>124168 transcripts detected.</span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.2.10
#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 996 4 
#&gt; metadata(28): version level ... wd sessionInfo
#&gt; assays(3): counts tpm length
#&gt; rownames(996): ENSMUST00000002533 ENSMUST00000006037 ...
#&gt;   ENSMUST00000208793 ENSMUST00000209020
#&gt; rowData names(13): broadClass description ... transcriptName
#&gt;   transcriptSupportLevel
#&gt; colnames(4): group1_1 group1_2 group2_1 group2_2
#&gt; colData names(21): sampleName description ... totalReads x5x3Bias
#&gt; Upload Dir: "/Users/mike/git/hbc/bcbioRNASeq/inst/extdata/bcbio"
#&gt; Upload Date: 2017-05-23
#&gt; R Load Date: 2019-05-13
#&gt; Level: "transcripts"
#&gt; Caller: "salmon"
#&gt; Organism: "Mus musculus"
#&gt; Interesting Groups: "sampleName"
#&gt; AnnotationHub: character(0)
#&gt; Ensembl Release: "87"
#&gt; Genome Build: "GRCm38"</div><div class='input'><span class='fu'>validObject</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] TRUE</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#genome-build">Genome build</a></li>

      <li><a href="#metadata">Metadata</a></li>

      <li><a href="#deseq-">DESeq2</a></li>

      <li><a href="#remote-data">Remote data</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Michael Steinbaugh, Lorena Pantano, Rory Kirchner, Victor Barrera</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

