<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Methods for coercing an object to a class — as.DESeqDataSet • bcbioRNASeq</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Methods for coercing an object to a class — as.DESeqDataSet" />
<meta property="og:description" content="Force an object to belong to a class." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.34</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Methods for coercing an object to a class</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/coerce-methods.R'><code>R/coerce-methods.R</code></a></small>
    <div class="hidden name"><code>coerce.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Force an object to belong to a class.</p>
    </div>

    <pre class="usage"><span class='fu'>as.DESeqDataSet</span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='fu'>as.DESeqTransform</span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='fu'>as.DGEList</span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='co'># S4 method for bcbioRNASeq</span>
<span class='fu'>as.DESeqDataSet</span>(<span class='no'>x</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for bcbioRNASeq</span>
<span class='fu'>as.DESeqTransform</span>(<span class='no'>x</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for bcbioRNASeq</span>
<span class='fu'>as.DGEList</span>(<span class='no'>x</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code>.
Perform command quietly, suppressing messages.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Modified object, of desired coercion type.</p>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Updated 2020-01-20.</p>
    <h2 class="hasAnchor" id="bcbiornaseq-to-deseqdataset"><a class="anchor" href="#bcbiornaseq-to-deseqdataset"></a>bcbioRNASeq to DESeqDataSet</h2>

    

<ol>
<li><p>Coerce to <code>RangedSummarizedExperiment</code>.</p></li>
<li><p>Round raw counts to <code>integer matrix</code>.</p></li>
<li><p>Subset <code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>colData()</a></code> to include only clean
factor columns. See <code><a href='https://rdrr.io/pkg/basejump/man/sampleData.html'>sampleData()</a></code> for details.</p></li>
<li><p>Simplify <code><a href='https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html'>metadata()</a></code> to include only relevant
information and updates <code>sessionInfo</code>.</p></li>
</ol>

<p>Note that gene-level counts are required. Alternatively,
<code><a href='https://rdrr.io/pkg/tximport/man/summarizeToGene.html'>summarizeToGene()</a></code> can be called to convert
transcript-level counts to gene-level. By default, we're using length-scaled
TPM, so a corresponding average transcript length matrix isn't necessary. The
average transcript length matrix is only necessary when raw counts matrix
isn't scaled during tximport call (see <code>countsFromAbundance</code> in
<code><a href='https://rdrr.io/pkg/tximport/man/tximport.html'>tximport()</a></code> documentation).</p>
    <h2 class="hasAnchor" id="bcbiornaseq-to-deseqtransform"><a class="anchor" href="#bcbiornaseq-to-deseqtransform"></a>bcbioRNASeq to DESeqTransform</h2>

    

<ol>
<li><p>Coerce to <code>DESeqDataSet</code>.</p></li>
<li><p>Call <code><a href='https://rdrr.io/pkg/DESeq2/man/DESeq.html'>DESeq2::DESeq()</a></code>.</p></li>
<li><p>Call <code><a href='https://rdrr.io/pkg/DESeq2/man/varianceStabilizingTransformation.html'>DESeq2::varianceStabilizingTransformation()</a></code>.</p></li>
</ol>

    <h2 class="hasAnchor" id="bcbiornaseq-to-dgelist"><a class="anchor" href="#bcbiornaseq-to-dgelist"></a>bcbioRNASeq to DGEList</h2>

    


<p>When <code>countsFromAbundance = "lengthScaledTPM"</code> (default):</p><ol>
<li><p>Call <code><a href='https://rdrr.io/pkg/edgeR/man/DGEList.html'>edgeR::DGEList()</a></code>.</p></li>
</ol>

<p>When <code>countsFromAbundance = "no"</code>:</p><ol>
<li><p>Call <code><a href='https://rdrr.io/pkg/edgeR/man/DGEList.html'>edgeR::DGEList()</a></code>.</p></li>
<li><p>Obtain per-observation scaling factors for length, adjusted to avoid
changing the magnitude of the counts.</p></li>
<li><p>Computing effective library sizes from scaled counts, to account for
composition biases between samples.</p></li>
<li><p>Combine effective library sizes with the length factors, and calculate
offsets for a log-link GLM.</p></li>
<li><p>Apply offset matrix using <code><a href='https://rdrr.io/pkg/edgeR/man/scaleOffset.html'>edgeR::scaleOffset()</a></code>.</p></li>
</ol>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<ul>
<li><p><code><a href='https://rdrr.io/pkg/tximport/man/tximport.html'>tximport::tximport()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html'>DESeq2::DESeqDataSetFromTximport()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/pkg/edgeR/man/DGEList.html'>edgeR::DGEList()</a></code>.</p></li>
</ul>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>bcb</span>)

<span class='co'>## bcbioRNASeq to DESeqDataSet ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>as</span>(<span class='no'>bcb</span>, <span class='st'>"DESeqDataSet"</span>)
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span>(<span class='kw pkg'>S4Vectors</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/S4Vectors/man/Vector-class.html'>mcols</a></span>(<span class='no'>x</span>))</div><div class='output co'>#&gt; [1] "broadClass"     "description"    "entrezID"       "geneBiotype"   
#&gt; [5] "geneID"         "geneIDVersion"  "geneName"       "seqCoordSystem"</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/class.html'>class</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "DESeqDataSet"
#&gt; attr(,"package")
#&gt; [1] "DESeq2"</div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: DESeqDataSet 
#&gt; dim: 100 6 
#&gt; metadata(11): caller countsFromAbundance ... date sessionInfo
#&gt; assays(1): counts
#&gt; rownames(100): ENSMUSG00000000001 ENSMUSG00000000003 ...
#&gt;   ENSMUSG00000062661 ENSMUSG00000074340
#&gt; rowData names(8): broadClass description ... geneName seqCoordSystem
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(26): averageInsertSize averageReadLength ... treatment
#&gt;   x5x3Bias</div><div class='input'>
<span class='co'>## bcbioRNASeq to RangedSummarizedExperiment ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>as</span>(<span class='no'>bcb</span>, <span class='st'>"RangedSummarizedExperiment"</span>)
<span class='fu'>slotNames</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "rowRanges"       "colData"         "assays"          "NAMES"          
#&gt; [5] "elementMetadata" "metadata"       </div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 100 6 
#&gt; metadata(27): allSamples bcbioCommandsLog ... wd yaml
#&gt; assays(7): counts aligned ... vst fpkm
#&gt; rownames(100): ENSMUSG00000000001 ENSMUSG00000000003 ...
#&gt;   ENSMUSG00000062661 ENSMUSG00000074340
#&gt; rowData names(8): broadClass description ... geneName seqCoordSystem
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(26): averageInsertSize averageReadLength ... treatment
#&gt;   x5x3Bias</div><div class='input'>
<span class='co'>## bcbioRNASeq to SummarizedExperiment ====</span>
<span class='co'>## Coerce to RangedSummarizedExperiment first.</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>as</span>(<span class='no'>bcb</span>, <span class='st'>"RangedSummarizedExperiment"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>as</span>(<span class='no'>x</span>, <span class='st'>"SummarizedExperiment"</span>)
<span class='fu'><a href='https://rdrr.io/r/base/class.html'>class</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "SummarizedExperiment"
#&gt; attr(,"package")
#&gt; [1] "SummarizedExperiment"</div><div class='input'><span class='fu'>slotNames</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "colData"         "assays"          "NAMES"           "elementMetadata"
#&gt; [5] "metadata"       </div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: SummarizedExperiment 
#&gt; dim: 100 6 
#&gt; metadata(27): allSamples bcbioCommandsLog ... wd yaml
#&gt; assays(7): counts aligned ... vst fpkm
#&gt; rownames(100): ENSMUSG00000000001 ENSMUSG00000000003 ...
#&gt;   ENSMUSG00000062661 ENSMUSG00000074340
#&gt; rowData names(8): broadClass description ... geneName seqCoordSystem
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(26): averageInsertSize averageReadLength ... treatment
#&gt;   x5x3Bias</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io/'>Lorena Pantano</a>, <a href='https://github.com/roryk/'>Rory Kirchner</a>, <a href='https://github.com/vbarrera/'>Victor Barrera</a>, <a href='https://github.com/chapmanb/'>Brad Chapman</a>, <a href='https://github.com/marypiper/'>Mary Piper</a>, <a href='https://github.com/mistrm82/'>Meeta Mistry</a>, <a href='https://github.com/rkhetani/'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford/'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann/'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson/'>John Hutchinson</a>, <a href='https://github.com/sjhosui/'>Shannan Ho Sui</a>, <a href='https://bioinformatics.sph.harvard.edu/'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


