<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Export — export • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Export — export" />

<meta property="og:description" content="Export data out of R and write to disk." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Export</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/export-methods.R'><code>R/export-methods.R</code></a></small>
    <div class="hidden name"><code>export.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Export data out of R and write to disk.</p>
    
    </div>

    <pre class="usage"><span class='co'># S4 method for DESeqResultsTables</span>
<span class='fu'>export</span>(<span class='no'>x</span>, <span class='kw'>dir</span> <span class='kw'>=</span> <span class='st'>"."</span>, <span class='kw'>compress</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>dropbox</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>rdsToken</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p><code>string</code>. Directory path.</p></td>
    </tr>
    <tr>
      <th>compress</th>
      <td><p><code>boolean</code>. Apply gzip compression to all files.</p></td>
    </tr>
    <tr>
      <th>dropbox</th>
      <td><p><code>boolean</code>. Export results to <a href='https://dropbox.com'>Dropbox</a> instead of local
storage. Recommended method by <a href='http://bioinformatics.sph.harvard.edu'>HBC</a> for permanent storage (e.g. <a href='https://stemcellcommons.org'>Stem Cell Commons</a>). When this option is enabled, unique links per file are
generated internally with <code><a href='http://www.rdocumentation.org/packages/bcbioBase/topics/copyToDropbox'>bcbioBase::copyToDropbox()</a></code>, which relies on the
<a href='https://cran.r-project.org/package=rdrop2'>rdrop2</a> package. Note that local files are written to <code><a href='http://www.rdocumentation.org/packages/base/topics/tempfile'>base::tempdir()</a></code>
when this option is enabled.</p></td>
    </tr>
    <tr>
      <th>rdsToken</th>
      <td><p><code>string</code> or <code>NULL</code>. RDS file token to use for Dropbox
authentication. If set <code>NULL</code> and <code>dropbox = TRUE</code> then an interactive
prompt will appear requesting authorization.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Invisible <code>character</code>. File path(s).</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This is a wrapper for <code><a href='http://www.rdocumentation.org/packages/rio/topics/export'>rio::export()</a></code> that adds support for additional S4
classes in Bioconductor.</p>
    
    <h2 class="hasAnchor" id="row-names"><a class="anchor" href="#row-names"></a>Row names</h2>

    
    <p>The standard <code><a href='http://www.rdocumentation.org/packages/rio/topics/export'>rio::export()</a></code> call will drop rownames when exporting to CSV.
The <code>readr::write_*()</code> family of functions also never write rownames. This is
a <em>really poor</em> default setting for handling genomic data, which often
contain gene identifiers in the rownames. Here we're performing any internal
tibble coercion step to ensure rownames are always moved to a "<code>rowname</code>"
column in the CSV export.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='http://www.rdocumentation.org/packages/rio/topics/export'>rio::export()</a></code>.</p>
<p>Other Import/Export Functions: <code>assignAndSaveData</code>,
  <code>import</code>, <code>loadDataAsName</code>,
  <code>loadData</code>, <code>loadRemoteData</code>,
  <code>saveData</code>, <code>transmit</code>,
  <code>writeCounts</code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>deseq_small</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='DESeqResultsTables.html'>DESeqResultsTables</a></span>(<span class='no'>deseq_small</span>)</div><div class='output co'>#&gt; <span class='message'>DESeqResults: treatment folic acid vs control (shrunken LFC)</span></div><div class='output co'>#&gt; <span class='message'>Adding `rowData()` annotations (atomic columns only).</span>
#&gt; <span class='message'>[1] "broadClass"     "description"    "geneBiotype"    "geneID"        </span>
#&gt; <span class='message'>[5] "geneName"       "seqCoordSystem"</span></div><div class='input'><span class='fu'>export</span>(<span class='no'>x</span>, <span class='kw'>dir</span> <span class='kw'>=</span> <span class='st'>"example"</span>)</div><div class='output co'>#&gt; <span class='message'>Writing treatment_folic_acid_vs_control_all.csv.gz, treatment_folic_acid_vs_control_up.csv.gz, treatment_folic_acid_vs_control_down.csv.gz, treatment_folic_acid_vs_control_both.csv.gz to /Users/mike/git/bcbioRNASeq-private/docs/reference/example.</span></div><div class='output co'>#&gt; <span class='message'>Exported treatment_folic_acid_vs_control_all.csv.gz.</span></div><div class='output co'>#&gt; <span class='message'>Exported treatment_folic_acid_vs_control_up.csv.gz.</span></div><div class='output co'>#&gt; <span class='message'>Exported treatment_folic_acid_vs_control_down.csv.gz.</span></div><div class='output co'>#&gt; <span class='message'>Exported treatment_folic_acid_vs_control_both.csv.gz.</span></div><div class='output co'>#&gt; DESeqResultsTables 1.21.22
#&gt; http://bioinformatics.sph.harvard.edu/bcbioRNASeq
#&gt; citation("bcbioRNASeq")
#&gt; contrast: treatment folic acid vs control
#&gt; alpha: 0.1
#&gt; lfcThreshold: 0
#&gt; dir: /Users/mike/git/bcbioRNASeq-private/docs/reference/example
#&gt; out of 500 with nonzero total read count
#&gt; adjusted p-value &lt; 0.1
#&gt; LFC &gt; 0 (up)       : 115, 23%
#&gt; LFC &lt; 0 (down)     : 139, 28%
#&gt; outliers [1]       : 6, 1.2%
#&gt; low counts [2]     : 0, 0%
#&gt; (mean count &lt; 33)</div><div class='input'><span class='fu'>list.files</span>(<span class='st'>"example"</span>)</div><div class='output co'>#&gt; [1] "treatment_folic_acid_vs_control_all.csv.gz" 
#&gt; [2] "treatment_folic_acid_vs_control_both.csv.gz"
#&gt; [3] "treatment_folic_acid_vs_control_down.csv.gz"
#&gt; [4] "treatment_folic_acid_vs_control_up.csv.gz"  </div><div class='input'>
<span class='co'>## Clean up.</span>
<span class='fu'>unlink</span>(<span class='st'>"example"</span>, <span class='kw'>recursive</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#row-names">Row names</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

