<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Extract or Replace Parts of an Object — extract • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Extract or Replace Parts of an Object — extract" />

<meta property="og:description" content="Extract genes by row and samples by column from a bcbioRNASeq object.
Internal count transformations are rescaled automatically, if defined." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract or Replace Parts of an Object</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/extract-methods.R'><code>R/extract-methods.R</code></a></small>
    <div class="hidden name"><code>extract.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Extract genes by row and samples by column from a <code>bcbioRNASeq</code> object.
Internal count transformations are rescaled automatically, if defined.</p>
    
    </div>

    <pre class="usage"># S4 method for bcbioRNASeq,ANY,ANY,ANY
[(x, i, j, drop = FALSE,
  transform = TRUE)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>object from which to extract element(s) or in which to replace element(s).</p></td>
    </tr>
    <tr>
      <th>i</th>
      <td><p>indices specifying elements to extract or replace.  Indices are
    <code>numeric</code> or <code>character</code> vectors or empty (missing) or
    <code>NULL</code>.  Numeric values are coerced to integer as by
    <code>as.integer</code> (and hence truncated towards zero).
    Character vectors will be matched to the <code>names</code> of the
    object (or for matrices/arrays, the <code>dimnames</code>):
    see &#8216;Character indices&#8217; below for further details.</p>
<p>For <code>[</code>-indexing only: <code>i</code>, <code>j</code>, <code>&#8230;</code> can be
    logical vectors, indicating elements/slices to select.  Such vectors
    are recycled if necessary to match the corresponding extent.
    <code>i</code>, <code>j</code>, <code>&#8230;</code> can also be negative integers,
    indicating elements/slices to leave out of the selection.</p>
<p>When indexing arrays by <code>[</code> a single argument <code>i</code> can be a
    matrix with as many columns as there are dimensions of <code>x</code>; the
    result is then a vector with elements corresponding to the sets of
    indices in each row of <code>i</code>.</p>
<p>An index value of <code>NULL</code> is treated as if it were <code>integer(0)</code>.</p></td>
    </tr>
    <tr>
      <th>j</th>
      <td><p>indices specifying elements to extract or replace.  Indices are
    <code>numeric</code> or <code>character</code> vectors or empty (missing) or
    <code>NULL</code>.  Numeric values are coerced to integer as by
    <code>as.integer</code> (and hence truncated towards zero).
    Character vectors will be matched to the <code>names</code> of the
    object (or for matrices/arrays, the <code>dimnames</code>):
    see &#8216;Character indices&#8217; below for further details.</p>
<p>For <code>[</code>-indexing only: <code>i</code>, <code>j</code>, <code>&#8230;</code> can be
    logical vectors, indicating elements/slices to select.  Such vectors
    are recycled if necessary to match the corresponding extent.
    <code>i</code>, <code>j</code>, <code>&#8230;</code> can also be negative integers,
    indicating elements/slices to leave out of the selection.</p>
<p>When indexing arrays by <code>[</code> a single argument <code>i</code> can be a
    matrix with as many columns as there are dimensions of <code>x</code>; the
    result is then a vector with elements corresponding to the sets of
    indices in each row of <code>i</code>.</p>
<p>An index value of <code>NULL</code> is treated as if it were <code>integer(0)</code>.</p></td>
    </tr>
    <tr>
      <th>drop</th>
      <td><p>For matrices and arrays.  If <code>TRUE</code> the result is
    coerced to the lowest possible dimension (see the examples).  This
    only works for extracting elements, not for the replacement.  See
    <code>drop</code> for further details.</p></td>
    </tr>
    <tr>
      <th>transform</th>
      <td><p><code>boolean</code>. Recalculate slotted DESeq2 transformations.
Recommended by default but can take a while for large datasets.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>bcbioRNASeq</code>.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>DESeq2 transformations will only be calculated when <code>transform = TRUE</code>
and either <code>rlog</code> or <code>vst</code> counts are defined in <code>assays()</code>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>help("[", "base")</code>.</p>
<p>Other S4 Object: <code><a href='bcbioRNASeq.html'>bcbioRNASeq</a></code>,
  <code><a href='coerce.html'>coerce</a></code>, <code><a href='show.html'>show</a></code>,
  <code><a href='updateObject.html'>updateObject</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>bcb_small</span>

<span class='co'># Minimum of 100 genes, 2 samples</span>
<span class='no'>genes</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='fu'>rownames</span>(<span class='no'>object</span>), <span class='fl'>100L</span>)
<span class='fu'>head</span>(<span class='no'>genes</span>)</div><div class='output co'>#&gt; [1] "ENSMUSG00000000088" "ENSMUSG00000000326" "ENSMUSG00000000346"
#&gt; [4] "ENSMUSG00000000740" "ENSMUSG00000001175" "ENSMUSG00000001666"</div><div class='input'><span class='no'>samples</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='fu'>colnames</span>(<span class='no'>object</span>), <span class='fl'>2L</span>)
<span class='fu'>head</span>(<span class='no'>samples</span>)</div><div class='output co'>#&gt; [1] "control_rep1" "control_rep2"</div><div class='input'>
<span class='co'># Extract by sample name</span>
<span class='no'>object</span>[, <span class='no'>samples</span>]</div><div class='output co'>#&gt; <span class='message'>Updating normalized counts</span></div><div class='output co'>#&gt; <span class='message'>Generating DESeqDataSet with DESeq2 1.20.0</span></div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Recalculating DESeq2 variance stabilizations (transform = TRUE)</span></div><div class='output co'>#&gt; <span class='message'>Applying variance-stabilizing transformation</span></div><div class='output co'>#&gt; <span class='message'>Applying rlog transformation</span></div><div class='output co'>#&gt; bcbioRNASeq 0.2.6
#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 500 2 
#&gt; metadata(31): version level ... subset originalVersion
#&gt; assays(6): counts tpm ... rlog vst
#&gt; rownames(500): ENSMUSG00000000088 ENSMUSG00000000326 ...
#&gt;   ENSMUSG00000114367 ENSMUSG00000114458
#&gt; rowData names(7): geneID geneName ... entrezID broadClass
#&gt; colnames(2): control_rep1 control_rep2
#&gt; colData names(26): sampleName day ... x5x3Bias xGC
#&gt; ====================================================================================================
#&gt; Upload Dir: "/mnt/O2/bcbio/bcbioRNASeq/F1000v2/GSE65267-merged/final"
#&gt; Upload Date: 2018-03-18
#&gt; R Load Date: 2018-08-09
#&gt; Level: "genes"
#&gt; Caller: "salmon"
#&gt; Organism: "Mus musculus"
#&gt; Interesting Groups: "treatment"
#&gt; AnnotationHub: "AH57770"
#&gt; Ensembl Release: "90"
#&gt; Genome Build: "GRCm38"</div><div class='input'>
<span class='co'># Extract by gene list</span>
<span class='no'>object</span>[<span class='no'>genes</span>, ]</div><div class='output co'>#&gt; <span class='message'>Updating normalized counts</span></div><div class='output co'>#&gt; <span class='message'>Generating DESeqDataSet with DESeq2 1.20.0</span></div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Recalculating DESeq2 variance stabilizations (transform = TRUE)</span></div><div class='output co'>#&gt; <span class='message'>Applying variance-stabilizing transformation</span></div><div class='output co'>#&gt; <span class='message'>Applying rlog transformation</span></div><div class='output co'>#&gt; bcbioRNASeq 0.2.6
#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 100 6 
#&gt; metadata(31): version level ... subset originalVersion
#&gt; assays(6): counts tpm ... rlog vst
#&gt; rownames(100): ENSMUSG00000000088 ENSMUSG00000000326 ...
#&gt;   ENSMUSG00000024503 ENSMUSG00000024610
#&gt; rowData names(7): geneID geneName ... entrezID broadClass
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(26): sampleName day ... x5x3Bias xGC
#&gt; ====================================================================================================
#&gt; Upload Dir: "/mnt/O2/bcbio/bcbioRNASeq/F1000v2/GSE65267-merged/final"
#&gt; Upload Date: 2018-03-18
#&gt; R Load Date: 2018-08-09
#&gt; Level: "genes"
#&gt; Caller: "salmon"
#&gt; Organism: "Mus musculus"
#&gt; Interesting Groups: "treatment"
#&gt; AnnotationHub: "AH57770"
#&gt; Ensembl Release: "90"
#&gt; Genome Build: "GRCm38"</div><div class='input'>
<span class='co'># Extract by both genes and samples</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='no'>object</span>[<span class='no'>genes</span>, <span class='no'>samples</span>]</div><div class='output co'>#&gt; <span class='message'>Updating normalized counts</span></div><div class='output co'>#&gt; <span class='message'>Generating DESeqDataSet with DESeq2 1.20.0</span></div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Recalculating DESeq2 variance stabilizations (transform = TRUE)</span></div><div class='output co'>#&gt; <span class='message'>Applying variance-stabilizing transformation</span></div><div class='output co'>#&gt; <span class='message'>Applying rlog transformation</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.2.6
#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 100 2 
#&gt; metadata(31): version level ... subset originalVersion
#&gt; assays(6): counts tpm ... rlog vst
#&gt; rownames(100): ENSMUSG00000000088 ENSMUSG00000000326 ...
#&gt;   ENSMUSG00000024503 ENSMUSG00000024610
#&gt; rowData names(7): geneID geneName ... entrezID broadClass
#&gt; colnames(2): control_rep1 control_rep2
#&gt; colData names(26): sampleName day ... x5x3Bias xGC
#&gt; ====================================================================================================
#&gt; Upload Dir: "/mnt/O2/bcbio/bcbioRNASeq/F1000v2/GSE65267-merged/final"
#&gt; Upload Date: 2018-03-18
#&gt; R Load Date: 2018-08-09
#&gt; Level: "genes"
#&gt; Caller: "salmon"
#&gt; Organism: "Mus musculus"
#&gt; Interesting Groups: "treatment"
#&gt; AnnotationHub: "AH57770"
#&gt; Ensembl Release: "90"
#&gt; Genome Build: "GRCm38"</div><div class='input'><span class='fu'>assayNames</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "counts"     "tpm"        "length"     "normalized" "rlog"      
#&gt; [6] "vst"       </div><div class='input'>
<span class='co'># Fast subsetting, by skipping DESeq2 transformations</span>
<span class='co'># Note that rlog and vst counts will not be defined</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='no'>object</span>[, <span class='no'>samples</span>, <span class='kw'>transform</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>]</div><div class='output co'>#&gt; <span class='message'>Updating normalized counts</span></div><div class='output co'>#&gt; <span class='message'>Generating DESeqDataSet with DESeq2 1.20.0</span></div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Skipping DESeq2 variance stabilizations (transform = FALSE)</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.2.6
#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 500 2 
#&gt; metadata(31): version level ... subset originalVersion
#&gt; assays(4): counts tpm length normalized
#&gt; rownames(500): ENSMUSG00000000088 ENSMUSG00000000326 ...
#&gt;   ENSMUSG00000114367 ENSMUSG00000114458
#&gt; rowData names(7): geneID geneName ... entrezID broadClass
#&gt; colnames(2): control_rep1 control_rep2
#&gt; colData names(26): sampleName day ... x5x3Bias xGC
#&gt; ====================================================================================================
#&gt; Upload Dir: "/mnt/O2/bcbio/bcbioRNASeq/F1000v2/GSE65267-merged/final"
#&gt; Upload Date: 2018-03-18
#&gt; R Load Date: 2018-08-09
#&gt; Level: "genes"
#&gt; Caller: "salmon"
#&gt; Organism: "Mus musculus"
#&gt; Interesting Groups: "treatment"
#&gt; AnnotationHub: "AH57770"
#&gt; Ensembl Release: "90"
#&gt; Genome Build: "GRCm38"</div><div class='input'><span class='fu'>names</span>(<span class='fu'>assays</span>(<span class='no'>x</span>))</div><div class='output co'>#&gt; [1] "counts"     "tpm"        "length"     "normalized"</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Lorena Pantano, Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

