<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Heatmaps — plot_correlation_heatmap • bcbioRnaseq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRnaseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRnaseq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Heatmaps</h1>
    </div>

    
    <p>These functions are wrappers to <code><a href='reexports.html'>pheatmap()</a></code> that have been optimized to work
with a <a href='bcbioRnaDataSet.html'>bcbioRnaDataSet</a> run object. They also depend on a <a href='reexports.html'>DESeqTransform</a>
object containing transformed counts, which are ideal for heatmap plots.</p>
    

    <pre class="usage"><span class='fu'>plot_correlation_heatmap</span>(<span class='no'>run</span>, <span class='no'>dt</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>intgroup</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>title</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"pearson"</span>,
  <span class='kw'>clustering_method</span> <span class='kw'>=</span> <span class='st'>"ward.D2"</span>, <span class='no'>...</span>)

<span class='fu'>plot_gene_heatmap</span>(<span class='no'>run</span>, <span class='no'>dt</span>, <span class='no'>genes</span>, <span class='kw'>intgroup</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>title</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>clustering_method</span> <span class='kw'>=</span> <span class='st'>"ward.D2"</span>, <span class='kw'>cluster_rows</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>cluster_cols</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>scale</span> <span class='kw'>=</span> <span class='st'>"row"</span>, <span class='no'>...</span>)

<span class='fu'>plot_deg_heatmap</span>(<span class='no'>run</span>, <span class='no'>dt</span>, <span class='no'>res</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>run</th>
      <td><p><a href='bcbioRnaDataSet.html'>bcbioRnaDataSet</a>.</p></td>
    </tr>
    <tr>
      <th>dt</th>
      <td><p><a href='reexports.html'>DESeqTransform</a>. Transformed counts generated by <code><a href='reexports.html'>rlog()</a></code>
(<strong>recommended</strong>) or <code><a href='reexports.html'>vst()</a></code> on a <a href='reexports.html'>DESeqDataSet</a> are supported.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>Character vector of specific gene identifiers to plot.</p></td>
    </tr>
    <tr>
      <th>intgroup</th>
      <td><p>Interesting groups to label with bars above heatmap. If
<code>NULL</code>, defaults to <code>intgroup</code> defined in <a href='bcbioRnaDataSet.html'>bcbioRnaDataSet</a>.</p></td>
    </tr>
    <tr>
      <th>annotation</th>
      <td><p>Optional alternative annotation to use. Useful when
plotting more than one column.</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p>Optional text to include in plot title.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>Correlation coefficient (or covariance) to be computed.
Defaults to <code>pearson</code> but <code>spearman</code> can also be used.</p></td>
    </tr>
    <tr>
      <th>clustering_method</th>
      <td><p>Clustering method. Accepts the same values as
<code>hclust()</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Passthrough parameters to <code><a href='reexports.html'>pheatmap()</a></code>.</p></td>
    </tr>
    <tr>
      <th>cluster_rows</th>
      <td><p>Whether rows should be clustered using <code>hclust()</code>.</p></td>
    </tr>
    <tr>
      <th>cluster_cols</th>
      <td><p>Whether columns should be clustered using <code>hclust()</code>.</p></td>
    </tr>
    <tr>
      <th>scale</th>
      <td><p>Character indicating if the values should be centered and scaled
in either the <code>row</code> direction, <code>column</code> direction, or <code>none</code>.</p></td>
    </tr>
    <tr>
      <th>res</th>
      <td><p><a href='reexports.html'>DESeqResults</a>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code><a href='reexports.html'>pheatmap()</a></code> object.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>plot_correlation_heatmap()</code> calculates a correlation matrix based on gene
expression per sample. By default, this function processes all gene counts
per sample to calculate the corrlation matrix. This behavior can be overrided
with the input of <code>gene</code> identifier vector. In this case, only the expression
of the desired genes will be used to calculate the correlation matrix.
    <code>plot_gene_heatmap()</code> facilities heatmap plotting of a specified set of
genes. The function by default performs row- and column-wise hierarhical
clustering using the Ward method, but this behavior can be overrided by
setting <code>cluster_rows</code> or <code>cluster_cols</code> to <code>FALSE</code>. When column clustering
is disabled, the function sorts columns by the interesting groups
(<code>intgroup</code>) specified in the <a href='bcbioRnaDataSet.html'>bcbioRnaDataSet</a> and then the sample names.
    <code>plot_deg_heatmap()</code> is a simplified version of <code>plot_gene_heatmap()</code> that is
optimized for handling a <a href='reexports.html'>DESeqResults</a> object rather a gene vector. All of
the optional parameters for <code>plot_gene_heatmap()</code> are also available to this
function.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <ul>
<li><p><code>cor()</code>.</p></li>
<li><p><code><a href='reexports.html'>pheatmap()</a></code>.</p></li>
</ul>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
          </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michael J. Steinbaugh, Rory Kirchner, Lorena Pantano, Victor Barrera, Harvard Chan Bioinformatics Core.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
