<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot Mapped Reads — plot_mapped_reads • bcbioRnaseq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRnaseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRnaseq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot Mapped Reads</h1>
    </div>

    
    <p>Plot Mapped Reads</p>
    

    <pre class="usage"><span class='fu'>plot_mapped_reads</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>.plot_mapped_reads</span>(<span class='no'>metrics</span>, <span class='no'>interesting_group</span>, <span class='kw'>pass_limit</span> <span class='kw'>=</span> <span class='fl'>20L</span>,
  <span class='kw'>warn_limit</span> <span class='kw'>=</span> <span class='fl'>10L</span>, <span class='kw'>flip</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># S4 method for bcbioRNADataSet</span>
<span class='fu'>plot_mapped_reads</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for data.frame</span>
<span class='fu'>plot_mapped_reads</span>(<span class='no'>object</span>,
  <span class='kw'>interesting_group</span> <span class='kw'>=</span> <span class='st'>"sample_name"</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>interesting_group</th>
      <td><p><em>Optional</em>. Category to use to group samples (color
and shape). If unset, this is automatically determined by the metadata set
inside the <a href='bcbioRNADataSet-class.html'>bcbioRNADataSet</a>.</p></td>
    </tr>
    <tr>
      <th>pass_limit</th>
      <td><p>Threshold to plot pass color marker.</p></td>
    </tr>
    <tr>
      <th>warn_limit</th>
      <td><p>Threshold to plot warning color marker.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><a href='http://www.rdocumentation.org/packages/ggplot2/topics/ggplot'>ggplot</a>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p>Other Quality Control Plots: <code><a href='plot_count_density.html'>plot_count_density</a></code>,
  <code><a href='plot_counts_per_gene.html'>plot_counts_per_gene</a></code>,
  <code><a href='plot_exonic_mapping_rate.html'>plot_exonic_mapping_rate</a></code>,
  <code><a href='plot_gene_detection_saturation.html'>plot_gene_detection_saturation</a></code>,
  <code><a href='plot_genes_detected.html'>plot_genes_detected</a></code>,
  <code><a href='plot_mapping_rate.html'>plot_mapping_rate</a></code>,
  <code><a href='plot_total_reads.html'>plot_total_reads</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>bcb</span>)

<span class='co'># bcbioRNADataSet</span>
<span class='fu'>plot_mapped_reads</span>(<span class='no'>bcb</span>)</div><div class='output co'>#&gt; <span class='error'>Error in layer(data = data, mapping = mapping, stat = StatIdentity, geom = GeomHline,     position = PositionIdentity, show.legend = show.legend, inherit.aes = FALSE,     params = list(na.rm = na.rm, ...)): object 'pass_color' not found</span></div><div class='input'>
<span class='co'># data.frame</span>
<span class='no'>metrics</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='metrics.html'>metrics</a></span>(<span class='no'>bcb</span>)
<span class='fu'>plot_mapped_reads</span>(<span class='no'>metrics</span>)</div><div class='output co'>#&gt; <span class='error'>Error in layer(data = data, mapping = mapping, stat = StatIdentity, geom = GeomHline,     position = PositionIdentity, show.legend = show.legend, inherit.aes = FALSE,     params = list(na.rm = na.rm, ...)): object 'pass_color' not found</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Rory Kirchner, Victor Barrera

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michael J. Steinbaugh, Lorena Pantano, Rory Kirchner, Victor Barrera, John Hutchinson, Harvard Chan Bioinformatics Core.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
