<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>MA-plot: plot differences versus averages for high-throughput data — plotMA • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="MA-plot: plot differences versus averages for high-throughput data — plotMA" />

<meta property="og:description" content="A generic function which produces an MA-plot for an object containing microarray, RNA-Seq
  or other data." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MA-plot: plot differences versus averages for high-throughput data</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/plotMA-methods.R'><code>R/plotMA-methods.R</code></a></small>
    <div class="hidden name"><code>plotMA.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>A generic function which produces an MA-plot for an object containing microarray, RNA-Seq
  or other data.</p>
    
    </div>

    <pre class="usage"><span class='fu'>plotMA</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for DESeqResults</span>
<span class='fu'>plotMA</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gene2symbol</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ntop</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>direction</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"both"</span>, <span class='st'>"up"</span>, <span class='st'>"down"</span>),
  <span class='kw'>pointColor</span> <span class='kw'>=</span> <span class='fu'>getOption</span>(<span class='st'>"basejump.point.color"</span>, <span class='st'>"gray50"</span>),
  <span class='kw'>sigPointColor</span> <span class='kw'>=</span> <span class='fu'>getOption</span>(<span class='st'>"basejump.sig.point.color"</span>, <span class='fu'>c</span>(<span class='kw'>upregulated</span> <span class='kw'>=</span>
  <span class='st'>"purple"</span>, <span class='kw'>downregulated</span> <span class='kw'>=</span> <span class='st'>"orange"</span>)), <span class='kw'>return</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"ggplot"</span>,
  <span class='st'>"DataFrame"</span>))

<span class='co'># S4 method for DESeqAnalysis</span>
<span class='fu'>plotMA</span>(<span class='no'>object</span>, <span class='kw'>results</span> <span class='kw'>=</span> <span class='fl'>1L</span>,
  <span class='kw'>lfcShrink</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ntop</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>direction</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"both"</span>,
  <span class='st'>"up"</span>, <span class='st'>"down"</span>), <span class='kw'>pointColor</span> <span class='kw'>=</span> <span class='fu'>getOption</span>(<span class='st'>"basejump.point.color"</span>,
  <span class='st'>"gray50"</span>), <span class='kw'>sigPointColor</span> <span class='kw'>=</span> <span class='fu'>getOption</span>(<span class='st'>"basejump.sig.point.color"</span>,
  <span class='fu'>c</span>(<span class='kw'>upregulated</span> <span class='kw'>=</span> <span class='st'>"purple"</span>, <span class='kw'>downregulated</span> <span class='kw'>=</span> <span class='st'>"orange"</span>)),
  <span class='kw'>return</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"ggplot"</span>, <span class='st'>"DataFrame"</span>))

<span class='fu'>plotMeanAverage</span>(<span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>A data object, typically containing count values from an RNA-Seq experiment or microarray intensity values.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments, for use in specific methods.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p><code>character</code>. Genes to include. These must match the rownames of
the object. It is best practice to use the stable gene identifiers from
Ensembl (e.g. "ENSG00000000003") and not the gene symbols.</p></td>
    </tr>
    <tr>
      <th>gene2symbol</th>
      <td><p><code>gene2symbol</code>. Gene-to-symbol mappings. Must contain
<code>geneID</code> and <code>geneName</code> columns, with rownames defined.</p></td>
    </tr>
    <tr>
      <th>ntop</th>
      <td><p><code>scalar integer</code>. Number of top genes to label.</p></td>
    </tr>
    <tr>
      <th>direction</th>
      <td><p><code>string</code>. Plot "<code>both</code>", "<code>up</code>", or "<code>down</code>" directions.</p></td>
    </tr>
    <tr>
      <th>pointColor</th>
      <td><p><code>string</code>. Default point color for the plot.</p></td>
    </tr>
    <tr>
      <th>sigPointColor</th>
      <td><p><code>character</code>. Color names for labeling upregulated and
downregulated genes. Also supports a character string for labeling DEGs
with the same color, regardless of direction.</p></td>
    </tr>
    <tr>
      <th>return</th>
      <td><p><code>string</code>. Object class to return. Uses <code>match.arg()</code> internally
and picks the first item in the vector by default.</p></td>
    </tr>
    <tr>
      <th>results</th>
      <td><p><code>scalar</code>. Position or name of <code>DESeqResults</code>.</p></td>
    </tr>
    <tr>
      <th>lfcShrink</th>
      <td><p><code>boolean</code>. Use shrunken log2 fold change (LFC) values.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>ggplot</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>An MA plot is an application of a Bland–Altman plot for visual representation
of genomic data. The plot visualizes the differences between measurements
taken in two samples, by transforming the data onto M (log ratio) and A
(mean average) scales, then plotting these values.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/plotMA'>DESeq2::plotMA()</a></code>.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>bcbioRNASeq</span><span class='kw ns'>::</span><span class='no'><a href='http://www.rdocumentation.org/packages/bcbioRNASeq/topics/deseq_small'>deseq_small</a></span>
<span class='fu'>print</span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; DESeqAnalysis 1.21.22
#&gt; http://bioinformatics.sph.harvard.edu/bcbioRNASeq
#&gt; citation("bcbioRNASeq")
#&gt; transform: vst
#&gt; contrastNames: treatment folic acid vs control
#&gt; class: DESeqDataSet 
#&gt; dim: 500 6 
#&gt; metadata(11): version caller ... date sessionInfo
#&gt; assays(4): counts mu H cooks
#&gt; rownames(500): ENSMUSG00000000088 ENSMUSG00000000326 ...
#&gt;   ENSMUSG00000114367 ENSMUSG00000114458
#&gt; rowData names(29): broadClass description ... deviance maxCooks
#&gt; colnames(6): control_rep1 control_rep2 ... fa_day7_rep2 fa_day7_rep3
#&gt; colData names(8): sampleName day ... interestingGroups sizeFactor</div><div class='input'>
<span class='co'>## DESeqAnalysis ====</span>
<span class='co'>## This is the current recommended method.</span>

<span class='no'>dds</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='as.html'>as</a></span>(<span class='no'>object</span>, <span class='st'>"DESeqDataSet"</span>)
<span class='no'>g2s</span> <span class='kw'>&lt;-</span> <span class='fu'>Gene2Symbol</span>(<span class='no'>dds</span>)
<span class='no'>geneIDs</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='no'>g2s</span><span class='kw'>[[</span><span class='st'>"geneID"</span>]])
<span class='fu'>print</span>(<span class='no'>geneIDs</span>)</div><div class='output co'>#&gt; [1] "ENSMUSG00000000088" "ENSMUSG00000000326" "ENSMUSG00000000346"
#&gt; [4] "ENSMUSG00000000740" "ENSMUSG00000001175" "ENSMUSG00000001666"</div><div class='input'><span class='no'>geneNames</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='no'>g2s</span><span class='kw'>[[</span><span class='st'>"geneName"</span>]])
<span class='fu'>print</span>(<span class='no'>geneNames</span>)</div><div class='output co'>#&gt; [1] "Cox5a"  "Comt"   "Dazap2" "Rpl13"  "Calm1"  "Ddt"   </div><div class='input'>
<span class='fu'>plotMA</span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; <span class='message'>DESeqResults: treatment folic acid vs control (shrunken LFC)</span></div><div class='img'><img src='plotMA-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'>## Customize the colors.</span>
<span class='fu'>plotMA</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
    <span class='kw'>pointColor</span> <span class='kw'>=</span> <span class='st'>"black"</span>,
    <span class='kw'>sigPointColor</span> <span class='kw'>=</span> <span class='st'>"purple"</span>
)</div><div class='output co'>#&gt; <span class='message'>DESeqResults: treatment folic acid vs control (shrunken LFC)</span></div><div class='img'><img src='plotMA-2.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>plotMA</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
    <span class='kw'>sigPointColor</span> <span class='kw'>=</span> <span class='fu'>c</span>(
        <span class='kw'>upregulated</span> <span class='kw'>=</span> <span class='st'>"green"</span>,
        <span class='kw'>downregulated</span> <span class='kw'>=</span> <span class='st'>"red"</span>
    )
)</div><div class='output co'>#&gt; <span class='message'>DESeqResults: treatment folic acid vs control (shrunken LFC)</span></div><div class='img'><img src='plotMA-3.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'>## Directional support (up or down).</span>
<span class='fu'>plotMA</span>(<span class='no'>object</span>, <span class='kw'>direction</span> <span class='kw'>=</span> <span class='st'>"up"</span>, <span class='kw'>ntop</span> <span class='kw'>=</span> <span class='fl'>5L</span>)</div><div class='output co'>#&gt; <span class='message'>DESeqResults: treatment folic acid vs control (shrunken LFC)</span></div><div class='img'><img src='plotMA-4.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>plotMA</span>(<span class='no'>object</span>, <span class='kw'>direction</span> <span class='kw'>=</span> <span class='st'>"down"</span>, <span class='kw'>ntop</span> <span class='kw'>=</span> <span class='fl'>5L</span>)</div><div class='output co'>#&gt; <span class='message'>DESeqResults: treatment folic acid vs control (shrunken LFC)</span></div><div class='img'><img src='plotMA-5.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'>## Label genes manually.</span>
<span class='co'>## Note that either gene IDs or names (symbols) are supported.</span>
<span class='fu'>plotMA</span>(<span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneIDs</span>)</div><div class='output co'>#&gt; <span class='message'>DESeqResults: treatment folic acid vs control (shrunken LFC)</span></div><div class='img'><img src='plotMA-6.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>plotMA</span>(<span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneNames</span>)</div><div class='output co'>#&gt; <span class='message'>DESeqResults: treatment folic acid vs control (shrunken LFC)</span></div><div class='img'><img src='plotMA-7.png' alt='' width='700' height='433' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Rory Kirchner

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

