<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Differential Expression Results Tables — resultsTables • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Differential Expression Results Tables — resultsTables" />

<meta property="og:description" content="Differential Expression Results Tables" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Differential Expression Results Tables</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/resultsTables-methods.R'><code>R/resultsTables-methods.R</code></a></small>
    <div class="hidden name"><code>resultsTables.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Differential Expression Results Tables</p>
    
    </div>

    <pre class="usage"><span class='fu'>resultsTables</span>(<span class='no'>results</span>, <span class='no'>counts</span>, <span class='no'>...</span>)

<span class='co'># S4 method for DESeqResults,DESeqDataSet</span>
<span class='fu'>resultsTables</span>(<span class='no'>results</span>, <span class='no'>counts</span>, <span class='no'>alpha</span>,
  <span class='kw'>lfcThreshold</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>summary</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>write</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>headerLevel</span> <span class='kw'>=</span> <span class='fl'>2L</span>, <span class='kw'>dir</span> <span class='kw'>=</span> <span class='st'>"."</span>, <span class='kw'>dropboxDir</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>rdsToken</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>results</th>
      <td><p><code>DESeqResults</code>.</p></td>
    </tr>
    <tr>
      <th>counts</th>
      <td><p><code>DESeqDataSet</code>. Corresponding object containing the counts that
were used to generate the <code>DESeqResults</code> object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p><code>scalar numeric</code>. Adjusted P value ("alpha") cutoff.</p></td>
    </tr>
    <tr>
      <th>lfcThreshold</th>
      <td><p><code>scalar numeric</code>. Log fold change ratio (base 2) cutoff
threshold.</p></td>
    </tr>
    <tr>
      <th>summary</th>
      <td><p><code>boolean</code>. Show summary statistics.</p></td>
    </tr>
    <tr>
      <th>write</th>
      <td><p><code>boolean</code>. Write CSV files to disk.</p></td>
    </tr>
    <tr>
      <th>headerLevel</th>
      <td><p><code>scalar integer</code> (<code>1</code>-<code>7</code>). R Markdown header level.</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p><code>string</code>. Local directory path.</p></td>
    </tr>
    <tr>
      <th>dropboxDir</th>
      <td><p><code>string</code> or <code>NULL</code>. Dropbox directory path where to archive
the results tables for permanent storage (e.g. Stem Cell Commons). When
this option is enabled, unique links per file are generated internally with
the rdrop2 package. Note that local files are written to <code><a href='http://www.rdocumentation.org/packages/base/topics/tempfile'>base::tempdir()</a></code>
and the <code>dir</code> argument is ignored, if this is enabled.</p></td>
    </tr>
    <tr>
      <th>rdsToken</th>
      <td><p><code>string</code> or <code>NULL</code>. RDS file token to use for Dropbox
authentication. If set <code>NULL</code> and <code>dropboxDir</code> is defined, then an
interactive prompt will appear requesting authorization.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>list</code> containing modified <code>DESeqResults</code> return, including
additional gene-level metadata and normalized counts.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Log fold change cutoff threshold ("<code>lfcThreshold</code>") does not apply to
statistical hypothesis testing, only gene filtering in the results tables.
See <code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/results'>DESeq2::results()</a></code> for additional information about using
<code>lfcThreshold</code> and <code>altHypothesis</code> to set an alternative hypothesis based
on expected fold changes.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Differential Expression Functions: <code><a href='alphaSummary.html'>alphaSummary</a></code>,
  <code><a href='contrastName.html'>contrastName</a></code>, <code><a href='plotDEGHeatmap.html'>plotDEGHeatmap</a></code>,
  <code><a href='plotDEGPCA.html'>plotDEGPCA</a></code>, <code><a href='plotMA.html'>plotMeanAverage</a></code>,
  <code><a href='plotVolcano.html'>plotVolcano</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># DESeqResults, DESeqDataSet ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>resultsTables</span>(<span class='kw'>results</span> <span class='kw'>=</span> <span class='no'>res_small</span>, <span class='kw'>counts</span> <span class='kw'>=</span> <span class='no'>dds_small</span>)</div><div class='output co'>#&gt; 
#&gt; 
#&gt; ## treatment folic acid vs control
#&gt; 
#&gt; 
#&gt; 
#&gt; ### Summary statistics
#&gt; 
#&gt; 
#&gt; - 500 genes in counts matrix
#&gt; - Base mean &gt; 0: 500 genes (non-zero)
#&gt; - Base mean &gt; 1: 500 genes
#&gt; - Alpha: 0.1
#&gt; - LFC threshold: 0
#&gt; - DEG pass alpha: 254 genes
#&gt; - DEG LFC up: 115 genes
#&gt; - DEG LFC down: 139 genes
#&gt; </div><div class='input'><span class='fu'>names</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "deg"          "degLFC"       "degLFCUp"     "degLFCDown"   "all"         
#&gt; [6] "contrast"     "alpha"        "lfcThreshold"</div><div class='input'><span class='fu'>glimpse</span>(<span class='no'>x</span><span class='kw'>[[</span><span class='st'>"deg"</span>]])</div><div class='output co'>#&gt; Observations: 254
#&gt; Variables: 44
#&gt; $ baseMean                                      &lt;dbl&gt; 12562.2432, 69339.510...
#&gt; $ log2FoldChange                                &lt;dbl&gt; 8.271501, 5.355191, 4...
#&gt; $ lfcSE                                         &lt;dbl&gt; 0.2736464, 0.2861448,...
#&gt; $ stat                                          &lt;dbl&gt; 29.930599, 18.716176,...
#&gt; $ pvalue                                        &lt;dbl&gt; 7.870297e-197, 3.6546...
#&gt; $ padj                                          &lt;dbl&gt; 3.887927e-194, 9.0269...
#&gt; $ geneID                                        &lt;chr&gt; "ENSMUSG00000024164",...
#&gt; $ seqnames                                      &lt;fct&gt; 17, 5, 3, 14, 11, 2, ...
#&gt; $ start                                         &lt;int&gt; 57203970, 104435118, ...
#&gt; $ end                                           &lt;int&gt; 57228136, 104441050, ...
#&gt; $ width                                         &lt;int&gt; 24167, 5933, 52092, 1...
#&gt; $ strand                                        &lt;fct&gt; -, +, +, +, +, -, -, ...
#&gt; $ geneName                                      &lt;chr&gt; "C3", "Spp1", "Cp", "...
#&gt; $ geneBiotype                                   &lt;fct&gt; protein_coding, prote...
#&gt; $ description                                   &lt;chr&gt; "complement component...
#&gt; $ seqCoordSystem                                &lt;fct&gt; chromosome, chromosom...
#&gt; $ broadClass                                    &lt;fct&gt; coding, coding, codin...
#&gt; $ baseVar                                       &lt;dbl&gt; 188041599.0, 60020548...
#&gt; $ allZero                                       &lt;lgl&gt; FALSE, FALSE, FALSE, ...
#&gt; $ dispGeneEst                                   &lt;dbl&gt; 0.012368440, 0.053292...
#&gt; $ dispGeneIter                                  &lt;dbl&gt; 14, 3, 7, 8, 7, 8, 3,...
#&gt; $ dispFit                                       &lt;dbl&gt; 0.08561209, 0.0839086...
#&gt; $ dispersion                                    &lt;dbl&gt; 0.05265557, 0.0658051...
#&gt; $ dispIter                                      &lt;int&gt; 10, 9, 7, 11, 7, 10, ...
#&gt; $ dispOutlier                                   &lt;lgl&gt; FALSE, FALSE, FALSE, ...
#&gt; $ dispMAP                                       &lt;dbl&gt; 0.05265557, 0.0658051...
#&gt; $ Intercept                                     &lt;dbl&gt; 5.898032, 11.389176, ...
#&gt; $ treatment_folic_acid_vs_control               &lt;dbl&gt; 8.715299, 5.664043, 4...
#&gt; $ SE_Intercept                                  &lt;dbl&gt; 0.2196067, 0.2142979,...
#&gt; $ SE_treatment_folic_acid_vs_control            &lt;dbl&gt; 0.2911836, 0.3026282,...
#&gt; $ WaldStatistic_Intercept                       &lt;dbl&gt; 26.85725, 53.14645, 3...
#&gt; $ WaldStatistic_treatment_folic_acid_vs_control &lt;dbl&gt; 29.930599, 18.716176,...
#&gt; $ WaldPvalue_Intercept                          &lt;dbl&gt; 6.940570e-159, 0.0000...
#&gt; $ WaldPvalue_treatment_folic_acid_vs_control    &lt;dbl&gt; 7.870297e-197, 3.6546...
#&gt; $ betaConv                                      &lt;lgl&gt; TRUE, TRUE, TRUE, TRU...
#&gt; $ betaIter                                      &lt;dbl&gt; 2, 3, 3, 2, 3, 4, 3, ...
#&gt; $ deviance                                      &lt;dbl&gt; 81.60880, 115.99167, ...
#&gt; $ maxCooks                                      &lt;dbl&gt; 0.54053191, 0.3631777...
#&gt; $ control_rep1                                  &lt;dbl&gt; 7.813753e+01, 2.31360...
#&gt; $ control_rep2                                  &lt;dbl&gt; 46.84188, 2998.40658,...
#&gt; $ control_rep3                                  &lt;dbl&gt; 5.622846e+01, 2.73325...
#&gt; $ fa_day7_rep1                                  &lt;dbl&gt; 25465.2621, 182255.32...
#&gt; $ fa_day7_rep2                                  &lt;dbl&gt; 25909.906, 121258.268...
#&gt; $ fa_day7_rep3                                  &lt;dbl&gt; 23817.0835, 104478.21...</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

