<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Sample Data — sampleData • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Sample Data — sampleData" />

<meta property="og:description" content="Return the sample metadata. Columns are always sanitized to factor." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Sample Data</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/sampleData-methods.R'><code>R/sampleData-methods.R</code></a></small>
    <div class="hidden name"><code>sampleData.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Return the sample metadata. Columns are always sanitized to factor.</p>
    
    </div>

    <pre class="usage"><span class='co'># S4 method for bcbioRNASeq</span>
<span class='fu'>sampleData</span>(<span class='no'>object</span>, <span class='no'>interestingGroups</span>,
  <span class='kw'>clean</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for bcbioRNASeq,DataFrame</span>
<span class='fu'>sampleData</span>(<span class='no'>object</span>) <span class='kw'>&lt;-</span> <span class='no'>value</span></pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code>. Groups of interest that define the
samples. If left unset, defaults to <code>sampleName</code>.</p></td>
    </tr>
    <tr>
      <th>clean</th>
      <td><p><code>boolean</code>. Only return <code>factor</code> columns not defined in
<a href='http://www.rdocumentation.org/packages/bcbioBase/topics/metadataBlacklist'>bcbioBase::metadataBlacklist</a>.</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>Value to assign.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>DataFrame</code>.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>This is a complement to the standard <code>colData()</code> function, but improves
support for accessing sample metadata for datasets where multiple items in
the columns map to a single sample (e.g. cells for a single-cell RNA-seq
experiment).</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='http://www.rdocumentation.org/packages/bcbioBase/topics/metadataBlacklist'>bcbioBase::metadataBlacklist</a>.</p>
<p>Other Data Functions: <code><a href='aggregateReplicates.html'>aggregateReplicates</a></code>,
  <code><a href='counts.html'>counts</a></code>, <code><a href='metrics.html'>metrics</a></code>,
  <code><a href='tmm.html'>tmm</a></code>, <code><a href='tpm.html'>tpm</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># SummarizedExperiment ====</span>
<span class='fu'>sampleData</span>(<span class='no'>bcb_small</span>, <span class='kw'>clean</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>) <span class='kw'>%&gt;%</span> <span class='fu'>glimpse</span>()</div><div class='output co'>#&gt; Formal class 'DataFrame' [package "S4Vectors"] with 6 slots
#&gt;   ..@ rownames       : chr [1:6] "control_rep1" "control_rep2" "control_rep3" "fa_day7_rep1" ...
#&gt;   ..@ nrows          : int 6
#&gt;   ..@ listData       :List of 6
#&gt;   .. ..$ sampleName: Factor w/ 6 levels "control_rep1",..: 1 2 3 4 5 6
#&gt;   .. ..$ day       : Factor w/ 2 levels "0","7": 1 1 1 2 2 2
#&gt;   .. ..$ replicate : Factor w/ 3 levels "1","2","3": 1 2 3 1 2 3
#&gt;   .. ..$ strain    : Factor w/ 1 level "Balb/c": 1 1 1 1 1 1
#&gt;   .. ..$ tissue    : Factor w/ 1 level "kidney": 1 1 1 1 1 1
#&gt;   .. ..$ treatment : Factor w/ 2 levels "control","folic_acid": 1 1 1 2 2 2
#&gt;   ..@ elementType    : chr "ANY"
#&gt;   ..@ elementMetadata: NULL
#&gt;   ..@ metadata       : list()</div><div class='input'><span class='fu'>sampleData</span>(<span class='no'>bcb_small</span>, <span class='kw'>clean</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>) <span class='kw'>%&gt;%</span> <span class='fu'>glimpse</span>()</div><div class='output co'>#&gt; Formal class 'DataFrame' [package "S4Vectors"] with 6 slots
#&gt;   ..@ rownames       : chr [1:6] "control_rep1" "control_rep2" "control_rep3" "fa_day7_rep1" ...
#&gt;   ..@ nrows          : int 6
#&gt;   ..@ listData       :List of 27
#&gt;   .. ..$ sampleName                   : Factor w/ 6 levels "control_rep1",..: 1 2 3 4 5 6
#&gt;   .. ..$ day                          : Factor w/ 2 levels "0","7": 1 1 1 2 2 2
#&gt;   .. ..$ description                  : Factor w/ 6 levels "control_rep1",..: 1 2 3 4 5 6
#&gt;   .. ..$ genomeBuild                  : Factor w/ 1 level "GRCm38_90": 1 1 1 1 1 1
#&gt;   .. ..$ replicate                    : Factor w/ 3 levels "1","2","3": 1 2 3 1 2 3
#&gt;   .. ..$ samRef                       : Factor w/ 1 level "/n/app/bcbio/dev/genomes/Mmusculus/GRCm38_90/seq/GRCm38_90.fa": 1 1 1 1 1 1
#&gt;   .. ..$ strain                       : Factor w/ 1 level "Balb/c": 1 1 1 1 1 1
#&gt;   .. ..$ tissue                       : Factor w/ 1 level "kidney": 1 1 1 1 1 1
#&gt;   .. ..$ treatment                    : Factor w/ 2 levels "control","folic_acid": 1 1 1 2 2 2
#&gt;   .. ..$ averageInsertSize            : num [1:6] 243 254 238 235 248 ...
#&gt;   .. ..$ averageReadLength            : num [1:6] 50 50 50 50 50 50
#&gt;   .. ..$ duplicates                   : num [1:6] 0 0 0 0 0 0
#&gt;   .. ..$ duplicationRateOfMapped      : num [1:6] 0.539 0.542 0.566 0.453 0.442 ...
#&gt;   .. ..$ exonicRate                   : num [1:6] 0.734 0.718 0.787 0.655 0.646 ...
#&gt;   .. ..$ intergenicRate               : num [1:6] 0.0634 0.0655 0.0461 0.0487 0.0586 0.0702
#&gt;   .. ..$ intronicRate                 : num [1:6] 0.203 0.217 0.167 0.296 0.296 ...
#&gt;   .. ..$ mappedPairedReads            : num [1:6] 5.94e+07 1.23e+08 5.32e+07 6.79e+07 5.94e+07 ...
#&gt;   .. ..$ mappedReads                  : num [1:6] 5.94e+07 1.23e+08 5.32e+07 6.79e+07 5.94e+07 ...
#&gt;   .. ..$ qualityFormat                : Factor w/ 1 level "standard": 1 1 1 1 1 1
#&gt;   .. ..$ rrna                         : num [1:6] 16412535 31779476 15967974 14933148 13229240 ...
#&gt;   .. ..$ rrnaRate                     : num [1:6] 0.614 0.585 0.645 0.555 0.558 ...
#&gt;   .. ..$ sequenceLength               : num [1:6] 50 50 50 50 50 50
#&gt;   .. ..$ sequencesFlaggedAsPoorQuality: num [1:6] 0 0 0 0 0 0
#&gt;   .. ..$ totalReads                   : num [1:6] 9.72e+07 1.79e+08 7.68e+07 8.11e+07 7.41e+07 ...
#&gt;   .. ..$ x5x3Bias                     : num [1:6] 0.88 0.77 0.91 0.45 0.76 0.65
#&gt;   .. ..$ xGC                          : num [1:6] 46 45 43 40 41 46
#&gt;   .. ..$ interestingGroups            : Factor w/ 2 levels "control","folic_acid": 1 1 1 2 2 2
#&gt;   .. .. ..- attr(*, "names")= chr [1:6] "control_rep1" "control_rep2" "control_rep3" "fa_day7_rep1" ...
#&gt;   ..@ elementType    : chr "ANY"
#&gt;   ..@ elementMetadata: NULL
#&gt;   ..@ metadata       : list()</div><div class='input'>
<span class='co'># Assignment support</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='no'>bcb_small</span>
<span class='fu'>sampleData</span>(<span class='no'>x</span>)<span class='kw'>[[</span><span class='st'>"test"</span>]] <span class='kw'>&lt;-</span> <span class='fu'>seq_len</span>(<span class='fu'>ncol</span>(<span class='no'>x</span>))
<span class='co'># `test` column should be now defined</span>
<span class='st'>"test"</span> <span class='kw'>%in%</span> <span class='fu'>colnames</span>(<span class='fu'>sampleData</span>(<span class='no'>x</span>))</div><div class='output co'>#&gt; [1] TRUE</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

