setClassUnion("missingOrNULL", c("missing", "NULL"))



#' bcbioRNADataSet
#'
#' S4 class containing RNA-seq counts and metadata generated by
#' [bcbio-nextgen](https://bcbio-nextgen.readthedocs.org/).
#'
#' [bcbioRNADataSet] is a subclass of [SummarizedExperiment] designed to store
#' an RNA-seq analysis. This class contains raw read counts and transcripts per
#' million (TPM) generated by [tximport()].
#'
# Additional slots can be defined by setting the representation, if necessary.
# This is useful for including alternate count matrices (e.g. featureCounts).
#'
#' The `metadata` accessor contains:
#'
#' - Sample summary statistics (a.k.a. metrics).
#' - Ensembl annotations.
#' - Server run paths.
#' - R local environment information, including [sessionInfo()].
#'
#' @author Lorena Patano, Michael Steinbaugh
#' @export
#'
#' @examples
#' path <- system.file("extra", package = "bcbioRnaseq")
#' bcb <- load_run(file.path(path, "bcbio"))
bcbioRNADataSet <- setClass(  # nolint
    "bcbioRNADataSet",
    contains = "SummarizedExperiment",
    slots = c(callers = "SimpleList"))
setValidity("bcbioRNADataSet", function(object) TRUE)



#' bcbioRNAResults
#'
#' S4 class containing contrast-level results of a differential expression
#' analysis.
#'
#' @author Michael Steinbaugh
#'
#' @export
bcbioRNAResults <- setClass(  # nolint
    "bcbioRNAResults",
    contains = "SummarizedExperiment",
    slots = c(dds = "DESeqDataSet",
              res = "DESeqResults"))
setValidity("bcbioRNAResults", function(object) TRUE)
