#' @rdname bcbioRNASeq-class
#' @export
#'
#' @details
#' The [bcbioRNASeq()] generator function automatically imports RNA-seq counts,
#' metadata, and the program versions used from a [bcbio][] RNA-seq run. Simply
#' point to the final upload directory generated by bcbio, and this generator
#' function will take care of the rest.
#'
#' [bcbio]: https://bcbio-nextgen.readthedocs.io
#'
#' @section Sample metadata:
#'
#' When loading a bcbio RNA-seq run, the sample metadata will be imported
#' automatically from the `project-summary.yaml` file in the final upload
#' directory. If you notice any typos in your metadata after completing the run,
#' these can be corrected by editing the YAML file.
#'
#' Alternatively, you can pass in a sample metadata file into the
#' [bcbioRNASeq()] function call using the `sampleMetadataFile` argument. This
#' requires either a CSV or Excel spreadsheet.
#'
#' The samples in the bcbio run must map to the `description` column. The values
#' provided in `description` must be unique. These values will be sanitized into
#' syntactically valid names (see [`make.names`][base::make.names] for more
#' information), and assigned as the column names of the `bcbioRNASeq` object.
#' The original values are stored as the `sampleName` column in `colData`, and
#' are used for all plotting functions. Do not attempt to set a `sampleID`
#' column, as this is used internally by the package.
#'
#' Here is a minimal example of a properly formatted sample metadata file:
#'
#' \tabular{ll}{
#'     description \tab genotype\cr
#'     sample1 \tab wildtype\cr
#'     sample2 \tab knockout\cr
#'     sample3 \tab wildtype\cr
#'     sample4 \tab knockout
#' }
#'
#' @section Valid names:
#'
#' R is strict about values that are considered valid for use in
#' [`names()`][base::names] and [`dimnames()`][base::dimnames] (i.e.
#' [`rownames()`][base::rownames] and [`colnames()`][base::colnames]).
#' Non-alphanumeric characters, spaces, and **dashes** are not valid. Use either
#' underscores or periods in place of dashes when working in R. Also note that
#' names should **not begin with a number**, and will be prefixed with an `X`
#' when sanitized. Consult the documentation in the
#' [`make.names()`][base::make.names] function for more information. We strongly
#' recommend adhering to these conventions when labeling samples, to help avoid
#' unexpected downstream behavior in R due to [`dimnames()`][base::dimnames]
#' mismatches.
#'
#' @section Genome annotations:
#'
#' [bcbioRNASeq()] provides support for automatic import of genome annotations,
#' which internally get processed into genomic ranges (`GRanges`) and are
#' slotted into the [`rowRanges()`][SummarizedExperiment::rowRanges] of the
#' object. Currently, we offer support for (1) [Ensembl][] genome annotations
#' from [AnnotationHub][] via [ensembldb][] (*recommended*); or (2) direct
#' import from a GTF/GFF file using [rtracklayer][].
#'
#' [ensembldb][] requires the `organism` and `ensemblRelease` arguments to be
#' defined. When both of these are set, `bcbioRNASeq` will attempt to
#' download and use the pre-built [Ensembl][] genome annotations from
#' [AnnotationHub][]. This method is preferred over direct loading of a GTF/GFF
#' file because the [AnnotationHub][] annotations contain additional rich
#' metadata not defined in a GFF file, specifically `description` and `entrezID`
#' values.
#'
#' Alternatively, if you are working with a non-standard or poorly annotated
#' genome that isn't available on [AnnotationHub][], we provide fall back
#' support for loading the genome annotations directly from the GTF file used by
#' the bcbio RNA-seq pipeline. This should be fully automatic for an R session
#' active on the same server used to run [bcbio][].
#'
#' Example bcbio GTF path: `genomes/Hsapiens/hg38/rnaseq/ref-transcripts.gtf`.
#'
#' In the event that you are working from a remote environment that doesn't
#' have file system access to the [bcbio][] `genomes` directory, we provide
#' additional fall back support for importing genome annotations from a GTF/GFF
#' directly with the `gffFile` argument.
#'
#' Internally, genome annotations are imported via the [basejump][] package,
#' specifically with either of these functions:
#'
#' - [basejump::makeGRangesFromEnsembl()].
#' - [basejump::makeGRangesFromGFF()].
#'
#' [AnnotationHub]: https://bioconductor.org/packages/AnnotationHub/
#' [basejump]: https://steinbaugh.com/basejump/
#' [bcbio]: https://bcbio-nextgen.readthedocs.io/
#' [Ensembl]: https://ensembl.org/
#' [ensembldb]: https://bioconductor.org/packages/ensembldb/
#' [rtracklayer]: https://bioconductor.org/packages/rtracklayer/
#'
#' @section Genome build:
#'
#' Ensure that the organism and genome build used with bcio match correctly here
#' in the function call. In particular, for the legacy *Homo sapiens*
#' GRCh37/hg19 genome build, ensure that `genomeBuild = "GRCh37"`. Otherwise,
#' the genomic ranges set in [`rowRanges()`][SummarizedExperiment::rowRanges]
#' will mismatch. It is recommended for current projects that GRCh38/hg38 is
#' used in place of GRCh37/hg19 if possible.
#'
#' @section DESeq2:
#'
#' DESeq2 is run automatically when [bcbioRNASeq()] is called. Internally, this
#' automatically slots normalized counts into
#' [`assays()`][SummarizedExperiment::assays], and optionally generates
#' variance-stabilized counts (e.g. `vst`, `rlog`), depending on the call. When
#' loading a dataset with a large number of samples (i.e. > 50), we recommend
#' disabling the `rlog` transformation, since it can take a long time to
#' compute.
#'
#' @section Remote connections:
#'
#' When working on a local machine, it is possible to load bcbio run data over a
#' remote connection using [sshfs][]. When loading a large number of samples, it
#' is preferable to call [bcbioRNASeq()] directly in R on the remote server, if
#' possible.
#'
#' [sshfs]: https://github.com/osxfuse/osxfuse/wiki/SSHFS
#'
#' @inheritParams basejump::makeSummarizedExperiment
#' @inheritParams basejump::params
#' @inheritParams bcbioBase::params
#' @inheritParams params
#'
#' @param level `character(1)`.
#'   Import counts at gene level ("`genes`"; *default*) or transcript level
#'   ("`transcripts`"; *advanced use*). Only tximport-compatible callers (e.g.
#'   salmon, kallisto, sailfish) can be loaded at transcript level. Aligned
#'   counts from featureCounts-compatible callers (e.g. STAR, HISAT2) can only
#'   be loaded at gene level.
#' @param caller `character(1)`.
#'   Expression caller:
#'
#'     - "`salmon`" (*default*): [Salmon][] alignment-free, quasi-mapped counts.
#'     - "`kallisto`": [Kallisto][] alignment-free, pseudo-aligned counts.
#'     - "`sailfish`": [Sailfish][] alignment-free, lightweight counts.
#'     - "`star`": [STAR][] (Spliced Transcripts Alignment to a Reference)
#'       aligned counts.
#'     - "`hisat2`": [HISAT2][] (Hierarchical Indexing for Spliced Alignment of
#'       Transcripts) graph-based aligned counts.
#'
#'   [HISAT2]: https://ccb.jhu.edu/software/hisat2/
#'   [Kallisto]: https://pachterlab.github.io/kallisto/
#'   [Sailfish]: http://www.cs.cmu.edu/~ckingsf/software/sailfish/
#'   [Salmon]: https://combine-lab.github.io/salmon/
#'   [STAR]: https://github.com/alexdobin/STAR/
#' @param countsFromAbundance `character(1)`.
#'   Whether to generate estimated counts using abundance estimates
#'   (*recommended by default*). `lengthScaledTPM` is a suitable default, and
#'   counts are scaled using the average transcript length over samples and then
#'   the library size. Refer to [tximport::tximport()] for more information on
#'   this parameter, but it should only ever be changed when loading some
#'   datasets at transcript level (e.g. for DTU analsyis).
#' @param vst `logical(1)`.
#'   Calculate variance-stabilizing transformation using
#'   [DESeq2::varianceStabilizingTransformation()]. Recommended by default for
#'   visualization.
#' @param rlog `logical(1)`.
#'   Calcualte regularized log transformation using [DESeq2::rlog()]. This
#'   calculation is slow for large datasets and now discouraged by default for
#'   visualization.
#'
#' @return `bcbioRNASeq`.
#'
#' @seealso
#' - `.S4methods(class = "bcbioRNASeq")`.
#' - `SummarizedExperiment::SummarizedExperiment()`.
#' - `methods::initialize()`.
#' - `methods::validObject()`.
#' - `BiocGenerics::updateObject()`.
#'
#' @examples
#' uploadDir <- system.file("extdata/bcbio", package = "bcbioRNASeq")
#'
#' ## Gene level.
#' object <- bcbioRNASeq(
#'     uploadDir = uploadDir,
#'     level = "genes",
#'     caller = "salmon",
#'     organism = "Mus musculus",
#'     ensemblRelease = 87L
#' )
#' print(object)
#'
#' ## Transcript level.
#' object <- bcbioRNASeq(
#'     uploadDir = uploadDir,
#'     level = "transcripts",
#'     caller = "salmon",
#'     organism = "Mus musculus",
#'     ensemblRelease = 87L
#' )
#' print(object)
bcbioRNASeq <- function(
    uploadDir,
    level = c("genes", "transcripts"),
    caller = c("salmon", "kallisto", "sailfish", "star", "hisat2"),
    samples = NULL,
    censorSamples = NULL,
    sampleMetadataFile = NULL,
    organism = NULL,
    genomeBuild = NULL,
    ensemblRelease = NULL,
    gffFile = NULL,
    transgeneNames = NULL,
    spikeNames = NULL,
    countsFromAbundance = "lengthScaledTPM",
    vst = TRUE,
    rlog = FALSE,
    interestingGroups = "sampleName",
    ...
) {
    # Legacy arguments ---------------------------------------------------------
    # nocov start
    call <- match.call()
    # annotable
    if ("annotable" %in% names(call)) {
        stop("`annotable` is defunct. Consider using `gffFile` instead.")
    }
    # ensemblVersion
    if ("ensemblVersion" %in% names(call)) {
        stop("Use `ensemblRelease` instead of `ensemblVersion`.")
    }
    # transformationLimit
    if ("transformationLimit" %in% names(call)) {
        stop(paste(
            "`transformationLimit` is deprecated in favor of",
            "separate `vst` and `rlog` arguments."
        ))
    }
    rm(call)
    # nocov end

    # Assert checks ------------------------------------------------------------
    assert(isADirectory(uploadDir))
    level <- match.arg(level)
    caller <- match.arg(caller)
    if (level == "transcripts") {
        assert(isSubset(caller, tximportCallers))
    }
    assert(
        isAny(samples, classes = c("character", "NULL")),
        isAny(censorSamples, classes = c("character", "NULL")),
        isString(sampleMetadataFile, nullOK = TRUE),
        isString(organism, nullOK = TRUE),
        isString(genomeBuild, nullOK = TRUE),
        isInt(ensemblRelease, nullOK = TRUE),
        isAny(transgeneNames, classes = c("character", "NULL")),
        isAny(spikeNames, classes = c("character", "NULL")),
        isString(gffFile, nullOK = TRUE)
    )

    if (isString(gffFile)) {
        assert(isAFile(gffFile) || containsAURL(gffFile))
    }
    # Don't allow AnnotationHub formals when specifying GFF file.
    if (!is.null(gffFile)) {
        assert(
            is.null(genomeBuild),
            is.null(ensemblRelease)
        )
    }
    # Organism is required when we're defining the genome.
    if (
        !is.null(genomeBuild) ||
        !is.null(ensemblRelease) ||
        !is.null(gffFile)
    ) {
        assert(isString(organism))
    }
    match.arg(
        arg = countsFromAbundance,
        choices = eval(formals(tximport)[["countsFromAbundance"]])
    )
    assert(
        isFlag(vst),
        isFlag(rlog),
        isCharacter(interestingGroups)
    )

    # Directory paths ----------------------------------------------------------
    uploadDir <- realpath(uploadDir)
    projectDir <- projectDir(uploadDir)
    sampleDirs <- sampleDirs(uploadDir)

    # Project summary YAML -----------------------------------------------------
    yamlFile <- file.path(projectDir, "project-summary.yaml")
    yaml <- import(yamlFile)
    assert(is.list(yaml))

    # bcbio run information ----------------------------------------------------
    dataVersions <- readDataVersions(file.path(projectDir, "data_versions.csv"))
    assert(is(dataVersions, "DataFrame"))

    programVersions <-
        readProgramVersions(file.path(projectDir, "programs.txt"))
    assert(is(programVersions, "DataFrame"))

    log <- import(file.path(projectDir, "bcbio-nextgen.log"))
    # This step enables our minimal dataset inside the package to pass checks.
    tryCatch(
        expr = assert(isCharacter(log)),
        error = function(e) {
            message("bcbio-nextgen.log file is empty.")
        }
    )

    commandsLog <- import(file.path(projectDir, "bcbio-nextgen-commands.log"))
    # This step enables our minimal dataset inside the package to pass checks.
    tryCatch(
        expr = assert(isCharacter(commandsLog)),
        error = function(e) {
            message("bcbio-nextgen-commands.log file is empty.")
        }
    )

    # Transcript-to-gene mappings ----------------------------------------------
    tx2gene <- readTx2Gene(
        file = file.path(projectDir, "tx2gene.csv"),
        organism = organism,
        genomeBuild = genomeBuild,
        ensemblRelease = ensemblRelease
    )
    assert(is(tx2gene, "Tx2Gene"))

    # Sequencing lanes ---------------------------------------------------------
    lanes <- detectLanes(sampleDirs)
    assert(
        isInt(lanes) ||
        identical(lanes, integer())
    )

    # Samples ------------------------------------------------------------------
    # Get the sample data.
    if (isString(sampleMetadataFile)) {
        # Normalize path of local file.
        if (file.exists(sampleMetadataFile)) {
            sampleMetadataFile <- realpath(sampleMetadataFile)
        }
        # User-defined external file.
        # Note that `readSampleData` also supports URLs.
        sampleData <- readSampleData(file = sampleMetadataFile, lanes = lanes)
    } else {
        # Automatic metadata from YAML file.
        sampleData <- getSampleDataFromYAML(yaml)
    }
    assert(isSubset(rownames(sampleData), names(sampleDirs)))

    # Subset the sample directories, if necessary.
    if (is.character(samples) || is.character(censorSamples)) {
        # Matching against the YAML "description" input here.
        description <- as.character(sampleData[["description"]])
        assert(isNonEmpty(description))
        if (is.character(samples)) {
            assert(isSubset(samples, description))
        } else {
            samples <- description
        }
        if (is.character(censorSamples)) {
            assert(isSubset(censorSamples, samples))
            samples <- setdiff(samples, censorSamples)
        }
        assert(isCharacter(samples))
        sampleData <- sampleData %>%
            as_tibble() %>%
            filter(!!sym("description") %in% !!samples) %>%
            as("DataFrame")
    }
    samples <- rownames(sampleData)
    assert(
        # Requiring at least 2 samples.
        length(samples) >= 2L,
        isSubset(samples, names(sampleDirs)),
        # Check that name sanitization worked.
        validNames(samples)
    )
    if (length(samples) < length(sampleDirs)) {
        message(paste(
            "Loading a subset of samples:",
            str_trunc(toString(samples), width = 80L),
            sep = "\n"
        ))
        allSamples <- FALSE
    } else {
        allSamples <- TRUE
    }
    sampleDirs <- sampleDirs[samples]

    # Column data --------------------------------------------------------------
    # Sample metrics. Note that sample metrics used for QC plots are not
    # currently generated when using fast RNA-seq workflow. This depends upon
    # MultiQC and aligned counts generated with STAR.
    colData <- getMetricsFromYAML(yaml)
    if (hasLength(colData)) {
        assert(
            areDisjointSets(colnames(colData), colnames(sampleData)),
            isSubset(rownames(sampleData), rownames(colData))
        )
        colData <- colData[rownames(sampleData), , drop = FALSE]
        colData <- cbind(colData, sampleData)
    } else {
        message("Fast mode detected. No metrics were calculated.")
        colData <- sampleData
    }
    assert(
        is(colData, "DataFrame"),
        identical(samples, rownames(colData))
    )

    # Assays -------------------------------------------------------------------
    assays <- list()
    # Use tximport by default for transcript-aware callers.
    # Otherwise, resort to loading the featureCounts aligned counts data.
    if (caller %in% tximportCallers) {
        if (level == "transcripts") {
            txOut <- TRUE
        } else {
            txOut <- FALSE
        }
        txi <- .tximport(
            sampleDirs = sampleDirs,
            type = caller,
            txOut = txOut,
            countsFromAbundance = countsFromAbundance,
            tx2gene = tx2gene
        )
        # Raw counts. Length scaled by default (see `countsFromAbundance`).
        # These counts are expected to be non-integer.
        assays[["counts"]] <- txi[["counts"]]
        # Transcripts per million.
        assays[["tpm"]] <- txi[["abundance"]]
        # Average transcript lengths.
        assays[["avgTxLength"]] <- txi[["length"]]
    } else if (caller %in% featureCountsCallers) {
        txi <- NULL
        countsFromAbundance <- "no"
        assert(identical(level, "genes"))
        # Load up the featureCounts aligned counts matrix.
        counts <- import(file = file.path(projectDir, "combined.counts"))
        assert(is.matrix(counts))
        colnames(counts) <- makeNames(colnames(counts))
        # Subset the combined matrix to match the samples.
        assert(isSubset(samples, colnames(counts)))
        counts <- counts[, samples, drop = FALSE]
        # Raw counts. These counts are expected to be integer.
        assays[["counts"]] <- counts
    }

    assert(
        identical(names(assays)[[1L]], "counts"),
        identical(colnames(assays[[1L]]), rownames(colData))
    )

    # Row data -----------------------------------------------------------------
    # Annotation priority:
    # 1. AnnotationHub.
    #    - Requires `organism` to be declared.
    #    - Ensure that Ensembl release and genome build match.
    # 2. GTF/GFF file. Use the bcbio GTF if possible.
    # 3. Fall back to slotting empty ranges. This is offered as support for
    #    complex datasets (e.g. multiple organisms).
    if (isString(organism) && is.numeric(ensemblRelease)) {
        # AnnotationHub (ensembldb).
        message("Using makeGRangesFromEnsembl() for annotations.")
        rowRanges <- makeGRangesFromEnsembl(
            organism = organism,
            level = level,
            genomeBuild = genomeBuild,
            release = ensemblRelease
        )
    } else {
        # GTF/GFF file.
        if (is.null(gffFile)) {
            # Attempt to use bcbio GTF automatically.
            gffFile <- getGTFFileFromYAML(yaml)
        }
        if (!is.null(gffFile)) {
            message("Using makeGRangesFromGFF() for annotations.")
            gffFile <- realpath(gffFile)
            rowRanges <- makeGRangesFromGFF(file = gffFile, level = level)
        } else {
            message("Slotting empty ranges into rowRanges().")
            rowRanges <- emptyRanges(rownames(assays[[1L]]))
        }
    }
    assert(is(rowRanges, "GRanges"))

    # Attempt to get genome build and Ensembl release if not declared.
    # Note that these will remain NULL when using GTF file (see above).
    if (is.null(genomeBuild)) {
        genomeBuild <- metadata(rowRanges)[["genomeBuild"]]
    }
    if (is.null(ensemblRelease)) {
        ensemblRelease <- metadata(rowRanges)[["ensemblRelease"]]
    }

    # Metadata -----------------------------------------------------------------
    # Interesting groups.
    interestingGroups <- camel(interestingGroups)
    assert(isSubset(interestingGroups, colnames(colData)))

    # Organism.
    # Attempt to detect automatically if not declared by user.
    if (is.null(organism)) {
        organism <- tryCatch(
            expr = detectOrganism(rownames(assays[[1L]])),
            error = function(e) {
                warning(paste(
                    "Failed to detect organism automatically.",
                    "Specify with `organism` argument."
                ))
            }
        )
    }

    metadata <- list(
        version = packageVersion,
        level = level,
        caller = caller,
        countsFromAbundance = countsFromAbundance,
        uploadDir = uploadDir,
        sampleDirs = sampleDirs,
        sampleMetadataFile = as.character(sampleMetadataFile),
        projectDir = projectDir,
        runDate = runDate(projectDir),
        interestingGroups = interestingGroups,
        organism = as.character(organism),
        genomeBuild = as.character(genomeBuild),
        ensemblRelease = as.integer(ensemblRelease),
        gffFile = as.character(gffFile),
        tx2gene = tx2gene,
        lanes = lanes,
        yaml = yaml,
        dataVersions = dataVersions,
        programVersions = programVersions,
        bcbioLog = log,
        bcbioCommandsLog = commandsLog,
        allSamples = allSamples,
        call = match.call()
    )

    # Generate bcbioRNASeq object ----------------------------------------------
    bcb <- .new.bcbioRNASeq(
        assays = assays,
        rowRanges = rowRanges,
        colData = colData,
        metadata = metadata,
        transgeneNames = transgeneNames,
        spikeNames = spikeNames
    )

    # DESeq2 normalizations ----------------------------------------------------
    if (level == "genes") {
        dds <- as(bcb, "DESeqDataSet")

        # Calculate size factors for normalized counts.
        message("Calculating normalized counts.")
        dds <- estimateSizeFactors(dds)
        assays(bcb)[["normalized"]] <- counts(dds, normalized = TRUE)

        # Skip full DESeq2 calculations (for internal bcbio test data).
        if (.dataHasVariation(dds)) {
            message("Calculating transformations.")
            # Expect warning about the empty design formula.
            dds <- suppressWarnings(DESeq(dds))
            if (isTRUE(vst)) {
                message("Applying variance-stabilizing transformation.")
                assays(bcb)[["vst"]] <-
                    assay(varianceStabilizingTransformation(dds))
            }
            if (isTRUE(rlog)) {
                message("Applying regularized log transformation.")
                assays(bcb)[["rlog"]] <- assay(rlog(dds))
            }
        } else {
            # nocov start
            # This step is covered by bcbio pipeline tests.
            message("Data has no variation. Skipping transformations.")
            # nocov end
        }

        # Calculate FPKM.
        # Skip this step if we've slotted empty ranges.
        if (length(unique(width(rowRanges(dds)))) > 1L) {
            message("Calculating FPKM.")
            assays(bcb)[["fpkm"]] <- fpkm(dds)
        } else {
            message(paste(
                "rowRanges() contains empty ranges.",
                "Skipping FPKM calculation."
            ))
        }
    }

    # Return -------------------------------------------------------------------
    assert(hasValidDimnames(bcb))
    validObject(bcb)
    bcb
}



.new.bcbioRNASeq <-  # nolint
    function(...) {
        new(Class = "bcbioRNASeq", makeSummarizedExperiment(...))
    }
