#' Package RNA-Seq counts into [SummarizedExperiment]
#'
#' Contains counts generated by [tximport()], along with
#' [bcbio-nextgen](https://bcbio-nextgen.readthedocs.io/) run metadata.
#'
#' @rdname SummarizedExperiment
#' @keywords internal
#'
#' @author Michael Steinbaugh
#'
#' @param tximport [tximport] list.
#' @param colData Sample metadata.
#' @param rowData [Ensembl](http://www.ensembl.org/) gene annotations.
#' @param metadata Custom metadata.
#'
#' @return [SummarizedExperiment].
.SummarizedExperiment <- function(
    tximport,
    colData,
    rowData,
    metadata = NULL) {
    message("Packaging SummarizedExperiment")

    # tximport ====
    counts <- tximport[["counts"]]
    # [TODO] Add check for `lengthScaledTPM`, otherwise return NULL?
    tpm <- tximport[["abundance"]]

    # TMM normalization (edgeR) ====
    tmm <- .tmm(counts)

    # Metadata ====
    # Coerce to [SimpleList], if necessary
    if (!is.null(metadata) & class(metadata) != "SimpleList") {
        metadata <- as(metadata, "SimpleList")
    }

    # Prepare [colData] and [rowData] ====
    colData <- colData[colnames(counts), ]
    rownames(colData) <- colnames(counts)

    rowData <- rowData[rownames(counts), ]
    rownames(rowData) <- rownames(counts)
    # [TODO] How to handle deprecated Ensembl identifiers?
    # This is due to mismatches between genome build and annotables build.
    # Examples: ENSMUSG00000029333, ENSMUSG00000035349, ENSMUSG00000042402

    if (!identical(rownames(counts), rownames(rowData))) {
        stop("Gene identifier mismatch in counts and rowData")
    }

    # Return [SummarizedExperiment] ====
    SummarizedExperiment(
        assays = SimpleList(
            counts = counts,  # raw counts first
            tpm = tpm,
            tmm = tmm),
        colData = colData,
        rowData = rowData,
        metadata = metadata)
}
