#' Package RNA-Seq counts into [SummarizedExperiment]
#'
#' Contains counts generated by [tximport()], along with
#' [bcbio-nextgen](https://bcbio-nextgen.readthedocs.io/) run metadata.
#'
#' @rdname summarized_experiment
#' @keywords internal
#'
#' @author Michael Steinbaugh
#'
#' @param assays assays.
#' @param col_data Sample metadata.
#' @param row_data [Ensembl](http://www.ensembl.org/) gene annotations.
#' @param metadata Custom metadata.
#'
#' @return [SummarizedExperiment].
.summarized_experiment <- function(
    assays,
    col_data,
    row_data,
    metadata = NULL) {
    message("Packaging SummarizedExperiment")
    assays <- as(assays, "SimpleList")
    counts <- assays[[1]]

    # colData
    col_data <- as(col_data, "DataFrame") %>%
        .[colnames(counts), , drop = FALSE]
    rownames(col_data) <- colnames(counts)

    # rowData
    row_data <- as(row_data, "DataFrame") %>%
        .[rownames(counts), , drop = FALSE]
    rownames(row_data) <- rownames(counts)
    if (!identical(rownames(counts), rownames(row_data))) {
        stop("Gene identifier mismatch")
    }

    # Metadata
    if (is.null(metadata)) {
        metadata <- SimpleList()
    } else {
        metadata <- as(metadata, "SimpleList")
    }
    metadata[["date"]] <- Sys.Date()
    metadata[["wd"]] <- getwd()
    metadata[["hpc"]] <- detect_hpc()
    metadata[["session_info"]] <- sessionInfo()
    # Check for retired Ensembl identifiers, which can happen when a more recent
    # annotable build is used than the genome build. If present, store these
    # identifiers in the metadata.
    if (any(is.na(row_data[["ensgene"]]))) {
        metadata[["retired_ensgene"]] <- row_data %>%
            as_tibble %>%
            filter(is.na(.data[["ensgene"]])) %>%
            pull("rowname") %>%
            sort
    }

    SummarizedExperiment(
        assays,
        colData = col_data,
        rowData = row_data,
        metadata = metadata)
}
