#' Package RNA-Seq counts into [SummarizedExperiment]
#'
#' Contains counts generated by [tximport()], along with
#' [bcbio-nextgen](https://bcbio-nextgen.readthedocs.io/) run metadata.
#'
#' @rdname summarized_experiment
#' @keywords internal
#'
#' @author Michael Steinbaugh
#'
#' @param tximport [tximport] list.
#' @param col_data Sample metadata.
#' @param row_data [Ensembl](http://www.ensembl.org/) gene annotations.
#' @param metadata Custom metadata.
#'
#' @return [SummarizedExperiment].
.summarized_experiment <- function(
    tximport,
    col_data,
    row_data,
    metadata = NULL) {
    message("Packaging SummarizedExperiment")

    # tximport ====
    counts <- tximport[["counts"]]
    # TODO Add check for `lengthScaledTPM`, otherwise return NULL?
    tpm <- tximport[["abundance"]]

    # TMM normalization (edgeR) ====
    tmm <- .tmm(counts)

    # Metadata ====
    # Coerce to [SimpleList], if necessary
    if (!is.null(metadata) & class(metadata) != "SimpleList") {
        metadata <- as(metadata, "SimpleList")
    }

    # col_data ====
    col_data <- col_data[colnames(counts), , drop = FALSE]
    rownames(col_data) <- colnames(counts)

    row_data <- row_data[rownames(counts), , drop = FALSE]
    rownames(row_data) <- rownames(counts)
    # TODO How to handle deprecated Ensembl identifiers?
    # This is due to mismatches between genome build and annotables build.
    # Examples: ENSMUSG00000029333, ENSMUSG00000035349, ENSMUSG00000042402

    if (!identical(rownames(counts), rownames(row_data))) {
        stop("Gene identifier mismatch")
    }

    # Return [SummarizedExperiment] ====
    SummarizedExperiment(
        assays = SimpleList(
            counts = counts,  # raw counts first
            tpm = tpm,
            tmm = tmm),
        colData = col_data,
        rowData = row_data,
        metadata = metadata)
}
