#' Top Tables of Differential Expression Results
#'
#' @name topTables
#' @author Michael Steinbaugh
#'
#' @inheritParams general
#'
#' @param object DESeq2 results tables list generated by [resultsTables()].
#' @param n Number genes to report.
#' @param coding Whether to only return coding genes.
#'
#' @return `kable`.
#'
#' @examples
#' load(system.file("extdata/bcb.rda", package = "bcbioRNASeq"))
#' load(system.file("extdata/res.rda", package = "bcbioRNASeq"))
#'
#' rowData <- rowData(bcb)
#' resTbl <- resultsTables(
#'     res,
#'     rowData = rowData,
#'     summary = FALSE,
#'     write = FALSE
#' )
#' topTables(resTbl)
NULL



# Constructors =================================================================
#' @importFrom basejump fixNA
#' @importFrom dplyr filter mutate rename
#' @importFrom tibble remove_rownames
.subsetTop <- function(
    object,
    n = 50L,
    coding = FALSE
) {
    assert_is_data.frame(object)
    assert_has_colnames(object)
    assert_has_rows(object)
    assertIsImplicitInteger(n)
    assert_is_a_bool(coding)
    # Note that `symbol` and `description` columns are optional
    requiredCols <- c("ensgene", "baseMean", "log2FoldChange", "padj")
    assert_is_subset(requiredCols, colnames(object))

    if (isTRUE(coding)) {
        assert_is_subset("broadClass", colnames(object))
        object <- object %>%
            .[.[["broadClass"]] == "coding", , drop = FALSE]
    }

    if (!nrow(object)) {
        return(NULL)
    }

    keepCols <- c(requiredCols, c("symbol", "description", "biotype"))
    return <- object %>%
        as_tibble() %>%
        remove_rownames() %>%
        head(n = n) %>%
        mutate(
            baseMean = round(.data[["baseMean"]]),
            log2FoldChange = format(.data[["log2FoldChange"]], digits = 3L),
            padj = format(.data[["padj"]], digits = 3L, scientific = TRUE)
        ) %>%
        .[, which(colnames(.) %in% keepCols)]

    # Sanitize the description, if necessary
    if ("description" %in% colnames(return)) {
        # Remove symbol information in description, if present
        return[["description"]] <- gsub(
            pattern = " \\[.+\\]$",
            replacement = "",
            x = return[["description"]]
        )
    }

    return %>%
        # Shorten `log2FoldChange` to `lfc`
        rename(lfc = .data[["log2FoldChange"]])
}



#' @importFrom knitr kable
.topTables.list <- function(  # nolint
    object,
    n = 50L,
    coding = FALSE
) {
    # Passthrough: n, coding
    assert_is_list(object)
    assert_is_subset(
        c("all", "deg", "degLFCDown", "degLFCUp"),
        names(object)
    )

    up <- .subsetTop(
        object[["degLFCUp"]],
        n = n,
        coding = coding
    )
    down <- .subsetTop(
        object[["degLFCDown"]],
        n = n,
        coding = coding
    )
    contrast <- object[["contrast"]]
    if (!is.null(up)) {
        show(kable(
            up,
            caption = paste(contrast, "(upregulated)")
        ))
    }
    if (!is.null(down)) {
        show(kable(
            down,
            caption = paste(contrast, "(downregulated)")
        ))
    }
}



# Methods ======================================================================
#' @rdname topTables
#' @export
setMethod(
    "topTables",
    signature("list"),
    .topTables.list
)
