#' Sample PCA Plot for Transformed Data
#'
#' Wrapper for [DESeq2::plotPCA()] that improves principal component analysis
#' (PCA) sample coloring and labeling.
#'
#' @rdname plotPCA
#' @name plotPCA
#' @author Michael Steinbaugh
#'
#' @importFrom BiocGenerics plotPCA
#'
#' @inheritParams AllGenerics
#'
#' @param transform String specifying [rlog] (**recommended**) or [vst]
#'   [DESeqTransform] slotted inside the [bcbioRNASeq] object.
#' @param genes *Optional*. Character vector of gene identifiers to use.
#' @param censorSamples *Optional*. Censors to exclude from PCA plot.
#' @param interestingGroups *Optional*. Interesting groups to use for point
#'   appearance. If `NULL`, color defaults to all `interestingGroups`
#'   parameters set in the [bcbioRNASeq] object.
#' @param label *Optional*. Superimpose sample text labels on the plot.
#' @param shape *Optional*. Make points easier to inspect with differing shapes.
#'   Generally this isn't recommended for PCA and works poorly for more
#'   than 6 discrete factors, as defined by the `interestingGroups` argument.
#'   We recommend typically leaving this to `FALSE`.
#' @param returnData Return PCA loadings data instead of plotting.
#'
#' @seealso [DESeq2::plotPCA()].
#'
#' @return [ggplot].
#'
#' @examples
#' plotPCA(bcb, label = TRUE)
#' plotPCA(bcb, label = FALSE)
#'
#' # Manually set interesting groups
#' # Note these columns aren't present in our example data
#' \dontrun{
#' plotPCA(bcb, interestingGroups = c("genotype", "treatment"))
#' }
NULL



# Constructors ====
#' @importFrom basejump camel
#' @importFrom ggplot2 aes_string coord_fixed geom_point ggplot guides labs
#' @importFrom ggrepel geom_text_repel
#' @importFrom grid arrow unit
#' @importFrom S4Vectors metadata
#' @importFrom viridis scale_color_viridis
.plotPCA <- function(
    object,
    transform = "rlog",
    genes = NULL,
    censorSamples = NULL,
    interestingGroups = NULL,
    label = FALSE,
    shape = FALSE,
    returnData = FALSE) {
    if (!transform %in% c("rlog", "vst")) {
        stop("DESeqTransform must be rlog or vst", call. = FALSE)
    }
    dt <- assays(object)[[transform]]
    .checkInterestingGroups(colData(dt), interestingGroups)

    # Subset genes, if desired
    if (!is.null(genes)) {
        dt <- dt[genes, , drop = FALSE]
        # Set ntop to the number of genes requested
        ntop <- length(genes)
    } else {
        # Recommended DESeq default
        ntop <- 500
    }

    # Censor samples, if desired
    if (!is.null(censorSamples)) {
        dt <- dt %>%
            .[, setdiff(colnames(.), censorSamples), drop = FALSE]
    }

    # Interesting groups
    if (is.null(interestingGroups)) {
        interestingGroups <- metadata(object)[["interestingGroups"]]
    }
    interestingGroupsName <- paste(interestingGroups, collapse = ":\n")

    # `group` column is generated by `plotPCA()` using `interestingGroups`
    data <- plotPCA(
        dt,
        intgroup = interestingGroups,
        returnData = TRUE,
        ntop = ntop) %>%
        camel(strict = FALSE)

    percentVar <- round(100 * attr(data, "percentVar"))

    # Always define color by `interestingGroups`
    data[["color"]] <- data[["group"]]

    if (isTRUE(shape)) {
        data[["shape"]] <- data[["group"]]
    } else {
        data[["shape"]] <- "default"
    }

    # Use sampleName for plot labels
    data[["label"]] <- colData(object)[, "sampleName"]

    p <- ggplot(
        data,
        mapping = aes_string(
            x = "pc1",
            y = "pc2",
            color = "color",
            shape = "shape")
    ) +
        geom_point(size = 4) +
        coord_fixed() +
        labs(title = "pca",
             x = paste0("pc1: ", percentVar[[1]], "% variance"),
             y = paste0("pc2: ", percentVar[[2]], "% variance"),
             color = interestingGroupsName,
             shape = interestingGroupsName) +
        scale_color_viridis(discrete = TRUE)

    if (!isTRUE(shape)) {
        p <- p +
            guides(shape = FALSE)
    }

    if (isTRUE(label)) {
        p <- p +
            geom_text_repel(
                # Color the labels to match the points
                aes_string(label = "label"),
                color = "black",
                fontface = "bold",
                # Add extra padding around each text label
                box.padding = unit(0.5, "lines"),
                # Add extra padding around each data point
                point.padding = unit(1.5, "lines"),
                # Color of the line segments
                segment.color = "gray",
                # Width of the line segments
                segment.size = 0.5,
                # Draw an arrow from the label to the data point
                arrow = arrow(length = unit(0.01, "npc")),
                # Strength of the repulsion force
                force = 1,
                show.legend = FALSE)
    }

    if (isTRUE(returnData)) {
        data
    } else {
        p
    }
}



# Methods ====
#' @rdname plotPCA
#' @export
setMethod(
    "plotPCA",
    signature("bcbioRNASeq"),
    .plotPCA)
